

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Alternative QC using ChIPQC &mdash; Epigenomics Workshop 2020 1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Signal visualisation" href="../visDeepTools/lab-vis-deepTools.html" />
    <link rel="prev" title="Detection of broad peaks from ChIP-seq data" href="../chipseqBroadPeaks/lab-broadpeaks.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Epigenomics Workshop 2020
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../precourse.html">Pre-course Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../online-classroom.html">Online classroom</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../schedule.html">Schedule</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../setup/lab-setup.html">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="../methylation_tutorials.html">DNA Methylation</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../chipseq_tutorials.html">ChIPseq</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../chipseqProc/lab-chipseq-processing.html">ChIPseq: processing and narrow peak detection</a></li>
<li class="toctree-l3"><a class="reference internal" href="../chipseqBroadPeaks/lab-broadpeaks.html">ChIPseq: broad peaks detection</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">ChIPseq: alternative QC</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l4"><a class="reference internal" href="#setting-up">Setting-up</a></li>
<li class="toctree-l4"><a class="reference internal" href="#running-chipqc">Running ChIPQC</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../visDeepTools/lab-vis-deepTools.html">Signal visualisation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../atacseq_tutorials.html">ATACseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quantitative_chip_tutorials.html">Advanced ChIP methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chip_downstream_tutorials.html">Downstream Processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../expdesign_data.html">Experimental Design and Data Considerations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_integration_tutorials.html">Data Integration</a></li>
<li class="toctree-l2"><a class="reference internal" href="../nextflow.html">Nextflow &amp; nf-core</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Epigenomics Workshop 2020</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../tutorials.html">Tutorials</a> &raquo;</li>
        
          <li><a href="../chipseq_tutorials.html">ChIP-seq</a> &raquo;</li>
        
      <li>Alternative QC using ChIPQC</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../../_sources/content/tutorials/chipseqQC/lab-chipqc.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="alternative-qc-using-chipqc">
<h1>Alternative QC using ChIPQC<a class="headerlink" href="#alternative-qc-using-chipqc" title="Permalink to this headline">¶</a></h1>
<p><strong>Learning outcomes</strong></p>
<p>Using <code class="docutils literal notranslate"><span class="pre">ChIPQC</span></code> package
- to generate a summary QC report for experimental sample groups
- to be able to understand and assess QC metrics and plots</p>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#introduction" id="id1">Introduction</a></p></li>
<li><p><a class="reference internal" href="#setting-up" id="id2">Setting-up</a></p></li>
<li><p><a class="reference internal" href="#running-chipqc" id="id3">Running ChIPQC</a></p></li>
</ul>
</div>
<div class="section" id="introduction">
<h2><a class="toc-backref" href="#id1">Introduction</a><a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>Here, we will explore the alternative quality control workflow, using Bioconductor <code class="docutils literal notranslate"><span class="pre">ChIPQC</span></code> package. <code class="docutils literal notranslate"><span class="pre">ChIPQC</span></code> computes quality metrics for aligned data from ChIP-seq experiments. It also provides simple ways to generate a ChIP-seq experiment quality report which can be examined to asses the absolute and relative quality of individual ChIP-seq samples (and their associated controls, as well as overall quality of the experimental data.)</p>
</div>
<div class="section" id="setting-up">
<h2><a class="toc-backref" href="#id2">Setting-up</a><a class="headerlink" href="#setting-up" title="Permalink to this headline">¶</a></h2>
<p>In principle one can run <code class="docutils literal notranslate"><span class="pre">ChIPQC</span></code> both on Uppmax or locally.
The instructions below are
to use the package on Uppmax. For local usage, please make sure the paths to files are correct.</p>
<p>Follow set-up instructions from <a class="reference internal" href="../diffBind/lab-diffBinding-remote.html"><span class="doc">Downstream analysis tutorial</span></a>, differential binding part. We will need the same files and we can work in the same directory.</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>If you start from a different location, you should <code class="docutils literal notranslate"><span class="pre">cd</span> <span class="pre">chipseq/analysis/R</span></code></p>
</div>
<p>You can now load the version of R for which we tested this class along with other dependencies:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> analysis/R

module load bioinfo-tools
module load samtools/1.8
module load conda/latest

conda activate /sw/courses/epigenomics/software/conda/v8

R
</pre></div>
</div>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>We are running
<code class="docutils literal notranslate"><span class="pre">R</span> <span class="pre">version</span> <span class="pre">4.0.3</span> <span class="pre">(2020-10-10)</span> <span class="pre">--</span> <span class="pre">&quot;Bunny-Wunnies</span> <span class="pre">Freak</span> <span class="pre">Out&quot;</span></code></p>
</div>
</div>
<div class="section" id="running-chipqc">
<h2><a class="toc-backref" href="#id3">Running ChIPQC</a><a class="headerlink" href="#running-chipqc" title="Permalink to this headline">¶</a></h2>
<p>While running commands (in <code class="docutils literal notranslate"><span class="pre">R</span></code>) feel free to have a look at <a class="reference external" href="http://bioconductor.org/packages/devel/bioc/vignettes/ChIPQC/inst/doc/ChIPQC.pdf">ChIPQC package documentation</a> to learn more about different steps and/or build upon them. Here we will just show you the very basics.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">#       add the path to the tutorial-specific R libraries</span>
<span class="nf">assign</span><span class="p">(</span><span class="s">&quot;.lib.loc&quot;</span><span class="p">,</span> <span class="s">&quot;/sw/courses/epigenomics/software/R&quot;</span><span class="p">,</span> <span class="n">envir</span> <span class="o">=</span> <span class="nf">environment</span><span class="p">(</span><span class="n">.libPaths</span><span class="p">))</span>

<span class="c1">#       you can see that the tutorial-specific R library path is added</span>
<span class="nf">.libPaths</span><span class="p">()</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;/sw/courses/epigenomics/software/R&quot;</span>

<span class="c1">#</span>
<span class="nf">library</span><span class="p">(</span><span class="n">DiffBind</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ChIPQC</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="p">)</span>


<span class="c1">#       reading in the sample information (metadata)</span>
<span class="n">samples</span> <span class="o">=</span> <span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;samples_REST.txt&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>

<span class="c1">#       inspecting the metadata</span>
<span class="n">samples</span>

<span class="c1">#       creating an object containing data</span>
<span class="n">res</span><span class="o">=</span><span class="nf">dba</span><span class="p">(</span><span class="n">sampleSheet</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span> <span class="n">config</span><span class="o">=</span><span class="nf">data.frame</span><span class="p">(</span><span class="n">RunParallel</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">))</span>

<span class="c1"># inspecting the object</span>
<span class="n">res</span>

<span class="c1">#       performing quality control</span>
<span class="n">resqc</span> <span class="o">=</span> <span class="nf">ChIPQC</span><span class="p">(</span><span class="n">res</span><span class="p">,</span><span class="n">annotation</span><span class="o">=</span><span class="s">&quot;hg19&quot;</span><span class="p">,</span> <span class="n">config</span><span class="o">=</span><span class="nf">data.frame</span><span class="p">(</span><span class="n">RunParallel</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span>

<span class="c1">#       creating the quality control report in html format</span>
<span class="nf">ChIPQCreport</span><span class="p">(</span><span class="n">resqc</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If you run this tutorial on Rackham, you may see an error</p>
<blockquote>
<div><p><code class="docutils literal notranslate"><span class="pre">ChIPQCreport(resqc)</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">Error</span> <span class="pre">in</span> <span class="pre">browseURL...</span></code> : <code class="docutils literal notranslate"><span class="pre">'browser'</span> <span class="pre">must</span> <span class="pre">be</span> <span class="pre">a</span> <span class="pre">non-empty</span> <span class="pre">character</span> <span class="pre">string</span></code></p>
</div></blockquote>
<p>This is because the html report cannot be open in a browser directly from Rackham. You can download it to your computer and view it locally.</p>
</div>
<p>You need to copy the report to your local computer (copy the entire <code class="docutils literal notranslate"><span class="pre">ChIPQCreport</span></code> folder):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>scp -r &lt;USER&gt;@rackham.uppmax.uu.se:path/to/ChIPQCreport .
</pre></div>
</div>
<p>Examine the html report.</p>
<p>What do you think?</p>
<p>Are results in line with the previous quality control workflow?</p>
<hr class="docutils" />
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="../visDeepTools/lab-vis-deepTools.html" class="btn btn-neutral float-right" title="Signal visualisation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../chipseqBroadPeaks/lab-broadpeaks.html" class="btn btn-neutral float-left" title="Detection of broad peaks from ChIP-seq data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, NBIS Epigenomics Workshop Team. All rights reserved..

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>