

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>ChIP-seq down-stream analysis &mdash; Epigenomics Workshop 2020 1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="ChIP-seq down-stream analysis" href="lab-diffBinding-remote.html" />
    <link rel="prev" title="Detection of differential binding sites using csaw" href="../csaw/lab-csaw.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home" alt="Documentation Home"> Epigenomics Workshop 2020
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../precourse.html">Pre-course Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../schedule.html">Schedule</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../setup/lab-setup.html">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="../methylation_tutorials.html">DNA Methylation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chipseq_tutorials.html">ChIPseq</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../chip_downstream_tutorials.html">Downstream Processing</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../chipseeker/lab-ChIPseeker.html">Annotation using ChIPseeker</a></li>
<li class="toctree-l3"><a class="reference internal" href="../csaw/lab-csaw.html">Detection of differential binding sites using csaw</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Detection of differential binding sites using DiffBind local</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l4"><a class="reference internal" href="#data-methods">Data &amp; Methods</a></li>
<li class="toctree-l4"><a class="reference internal" href="#setting-up">Setting-up</a></li>
<li class="toctree-l4"><a class="reference internal" href="#differential-binding">Differential binding</a></li>
<li class="toctree-l4"><a class="reference internal" href="#functional-analysis">Functional analysis</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="lab-diffBinding-remote.html">Detection of differential binding sites using DiffBind remote</a></li>
<li class="toctree-l3"><a class="reference internal" href="../motifs/lab-motifs.html">Identification of sequence motifs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../expdesign_data.html">Experimental Design and Data Considerations</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Epigenomics Workshop 2020</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../tutorials.html">Tutorials</a> &raquo;</li>
        
          <li><a href="../chip_downstream_tutorials.html">Downstream Analyses</a> &raquo;</li>
        
      <li>ChIP-seq down-stream analysis</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../../_sources/content/tutorials/diffBind/lab-diffBinding-local.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="chip-seq-down-stream-analysis">
<h1>ChIP-seq down-stream analysis<a class="headerlink" href="#chip-seq-down-stream-analysis" title="Permalink to this headline">¶</a></h1>
<p><strong>Learning outcomes</strong>
- obtain differentially bound sites with <code class="docutils literal notranslate"><span class="pre">DiffBind</span></code>
- annotate differentially bound sites with nearest genes and genomic features with <code class="docutils literal notranslate"><span class="pre">ChIPpeakAnno</span></code>
- perform functional enrichment analysis to identify predominant biological themes with <code class="docutils literal notranslate"><span class="pre">ChIPpeakAnno</span></code> and <code class="docutils literal notranslate"><span class="pre">reactome.db</span></code></p>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#introduction" id="id19">Introduction</a></p></li>
<li><p><a class="reference internal" href="#data-methods" id="id20">Data &amp; Methods</a></p></li>
<li><p><a class="reference internal" href="#setting-up" id="id21">Setting-up</a></p></li>
<li><p><a class="reference internal" href="#differential-binding" id="id22">Differential binding</a></p>
<ul>
<li><p><a class="reference internal" href="#intro" id="id23">Intro</a></p></li>
<li><p><a class="reference internal" href="#setting-up-diffbind" id="id24">Setting-up <code class="docutils literal notranslate"><span class="pre">DiffBind</span></code></a></p></li>
<li><p><a class="reference internal" href="#running-diffbind" id="id25">Running <code class="docutils literal notranslate"><span class="pre">DiffBind</span></code></a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#functional-analysis" id="id26">Functional analysis</a></p>
<ul>
<li><p><a class="reference internal" href="#setting-up-chippeakanno" id="id27">Setting-up <code class="docutils literal notranslate"><span class="pre">ChIPpeakAnno</span></code></a></p></li>
<li><p><a class="reference internal" href="#running-chippeakanno" id="id28">Running <code class="docutils literal notranslate"><span class="pre">ChIPpeakAnno</span></code></a></p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="introduction">
<h2><a class="toc-backref" href="#id19">Introduction</a><a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>Welcome back to the second part of the tutorial. In the first part we have learnt how to access the quality of ChIP-seq data and we how to derive a consensus peakset for downstream analyses.</p>
<p>In this part we will learn how to place our peaks in a biological context, by identifying differentially bound sites between two sample groups and annotating these sites to find out predominant biological themes separating the groups.</p>
</div>
<div class="section" id="data-methods">
<h2><a class="toc-backref" href="#id20">Data &amp; Methods</a><a class="headerlink" href="#data-methods" title="Permalink to this headline">¶</a></h2>
<p>We will continue using the same data as in the first part of the tutorial. Please note that usually <strong>three biological replicates</strong> are the <strong>minimum requirement</strong> for statistical analysis such as in factor occupancy.</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>The ENCODE data we are using have only two replicates and we are using them to demonstrate the tools and methodologies. No biological conclusions should be drawn from them, or as a matter of fact, from any other dataset with duplicates only. Just because the tool computes the results does not make it right!</p>
</div>
</div>
<div class="section" id="setting-up">
<h2><a class="toc-backref" href="#id21">Setting-up</a><a class="headerlink" href="#setting-up" title="Permalink to this headline">¶</a></h2>
<p>If you have not done it already, install <code class="docutils literal notranslate"><span class="pre">R</span></code> and <code class="docutils literal notranslate"><span class="pre">R-Studio</span></code>. Refer back to <a class="reference internal" href="../../precourse.html"><span class="doc">pre-course</span></a> preparations for instructions.</p>
<p>For information on how to install packages in <code class="docutils literal notranslate"><span class="pre">R</span></code>, please refer to <span class="xref std std-doc">setup &lt;../setup/lab-setup.rst</span>.</p>
</div>
<div class="section" id="differential-binding">
<h2><a class="toc-backref" href="#id22">Differential binding</a><a class="headerlink" href="#differential-binding" title="Permalink to this headline">¶</a></h2>
<div class="section" id="intro">
<h3><a class="toc-backref" href="#id23">Intro</a><a class="headerlink" href="#intro" title="Permalink to this headline">¶</a></h3>
<p>We will usage <code class="docutils literal notranslate"><span class="pre">Bioconductor</span></code> package <a class="reference external" href="http://bioconductor.org/packages/release/bioc/html/DiffBind.html">DiffBind</a> to <strong>identify sites that are differentially bound</strong> between two sample groups.</p>
<p>The package includes <cite>“functions to support the processing of peak sets, including overlapping and merging peak sets, counting sequencing reads overlapping intervals in peak sets, and identifying statistically significantly differentially bound sites based on evidence of binding affinity (measured by differences in read densities). To this end it uses statistical routines developed in an RNA-Seq context (primarily the Bioconductor packages edgeR and DESeq2). Additionally, the package builds on Rgraphics routines to provide a set of standardized plots to aid in binding analysis.”</cite></p>
<blockquote>
<div><p>This means that we will <strong>repeat deriving a consensus peakset</strong> in a more powerful way before identifying differentially bound sites. Actually, identifying the consensus peaks is an important step that takes up entire chapter in the <code class="docutils literal notranslate"><span class="pre">DiffBind</span></code> manual. We recommend reading entire section:  <a class="reference external" href="http://bioconductor.org/packages/devel/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf">6_2 Deriving consensus peaksets</a>.</p>
</div></blockquote>
<p>So how does the differential binding affinity analysis work?</p>
<p><cite>“The core functionality of DiffBind is the differential binding affinity analysis, which enables binding sites to be identified that are statistically significantly differentially bound between sample groups. To accomplish this, first a contrast (or contrasts) is established, dividing the samples into groups to be compared. Next the core analysis routines are executed, by default using DESeq2. This will assign a p-value and FDR to each candidate binding site indicating confidence that they are differentially bound.”</cite></p>
</div>
<div class="section" id="setting-up-diffbind">
<h3><a class="toc-backref" href="#id24">Setting-up <code class="docutils literal notranslate"><span class="pre">DiffBind</span></code></a><a class="headerlink" href="#setting-up-diffbind" title="Permalink to this headline">¶</a></h3>
<p>Create a separate directory on your local computer where you would like to work and name it e.g. <code class="docutils literal notranslate"><span class="pre">diffBind</span></code>.</p>
<p>To be able to run <code class="docutils literal notranslate"><span class="pre">DiffBind</span></code> locally, we will need access to a set of files
- BAM files
- BED files with called peaks regions
- sample sheet information <code class="docutils literal notranslate"><span class="pre">.txt</span></code> file</p>
<p>Download these from Box (link under Alternative Files locations on the tutorials main site) or download these from Uppmax with <a href="#id1"><span class="problematic" id="id2">``</span></a>scp``command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>scp -r &lt;username&gt;@rackham.uppmax.uu.se:~/chipseq/data/bam/* .
scp -r &lt;username&gt;@rackham.uppmax.uu.se:~/chipseq/results/peaks_bed/* .
scp -r &lt;username&gt;@rackham.uppmax.uu.se:~/chipseq/analysis/R/samples_REST.txt .
</pre></div>
</div>
<p>You may want to place the downloaded files in the <code class="docutils literal notranslate"><span class="pre">diffBind</span></code> directory or at least keep a track of their location.</p>
<p>Also we need to modify <code class="docutils literal notranslate"><span class="pre">samples_REST.txt</span></code> so the pathways are pointing to the BAM and BED files on your local computer. Adjust the pathways in any editor your like.</p>
<p>Now, we can open R-Studio and set working directory to working folder e.g. <code class="docutils literal notranslate"><span class="pre">diffBind</span></code> folder by <code class="docutils literal notranslate"><span class="pre">Session</span> <span class="pre">-&gt;</span> <span class="pre">Set</span> <span class="pre">Working</span> <span class="pre">Directory</span> <span class="pre">-&gt;</span> <span class="pre">Choose</span> <span class="pre">Directory</span></code>. Now, all R commands will be in respect to this directory.</p>
<p>You can type commands directly in the Console window. A bit smarter way is to open a new R script under <code class="docutils literal notranslate"><span class="pre">File</span> <span class="pre">-&gt;</span> <span class="pre">New</span> <span class="pre">File</span> <span class="pre">-&gt;</span> <span class="pre">R</span> <span class="pre">Script</span></code> and type commands there, saving it from time to time. This way if you want to go back and repeat commands you can. To execute commands written in script, copy and paste commands to Console window and press Enter, press <code class="docutils literal notranslate"><span class="pre">Run</span></code> button in R-Studio or ask for a demo.</p>
<p>To use <code class="docutils literal notranslate"><span class="pre">DiffBind</span></code> package we need to install it first. To do so:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">requireNamespace</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
    <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;DiffBind&quot;</span><span class="p">,</span> <span class="n">version</span> <span class="o">=</span> <span class="s">&quot;3.8&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>If the above worked, we should be able to load DiffBind library:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">DiffBind</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="running-diffbind">
<h3><a class="toc-backref" href="#id25">Running <code class="docutils literal notranslate"><span class="pre">DiffBind</span></code></a><a class="headerlink" href="#running-diffbind" title="Permalink to this headline">¶</a></h3>
<p>We will now follow <code class="docutils literal notranslate"><span class="pre">DiffBind</span></code> example to obtain differentially bound sites, given our samples. You may want to open <code class="docutils literal notranslate"><span class="pre">DiffBind</span></code> tutorial and read section <a class="reference external" href="http://bioconductor.org/packages/devel/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf">3 Example Obtaining differentially bound sites</a> while typing the command to get more information about each step.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># reading in the sample information (metadata)</span>
<span class="n">samples</span> <span class="o">=</span> <span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;samples_REST.txt&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>

<span class="c1">#       inspecting the metadata</span>
<span class="n">samples</span>

<span class="c1">#       creating an object containing data</span>
<span class="n">res</span><span class="o">=</span><span class="nf">dba</span><span class="p">(</span><span class="n">sampleSheet</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span> <span class="n">config</span><span class="o">=</span><span class="nf">data.frame</span><span class="p">(</span><span class="n">RunParallel</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">))</span>

<span class="c1"># inspecting the object: how many peaks are identified given the default settings?</span>
<span class="n">res</span>

<span class="c1"># counting reads mapping to intervals (peaks)</span>
<span class="c1"># at this step a normalisation is applied by the default set to: score=DBA_SCORE_TMM_MINUS_FULL</span>
<span class="n">res.cnt</span> <span class="o">=</span> <span class="nf">dba.count</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">minOverlap</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">score</span><span class="o">=</span><span class="n">DBA_SCORE_TMM_MINUS_FULL</span><span class="p">,</span> <span class="n">fragmentSize</span><span class="o">=</span><span class="m">130</span><span class="p">)</span>

<span class="c1"># inspecting the object: notice the FRiP values!</span>
<span class="n">res.cnt</span>

<span class="c1"># plotting the correlation of libraries based on normalised counts of reads in peaks</span>
<span class="nf">pdf</span><span class="p">(</span><span class="s">&quot;correlation_libraries_normalised.pdf&quot;</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">res.cnt</span><span class="p">)</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="c1"># PCA scores plot: data overview</span>
<span class="nf">pdf</span><span class="p">(</span><span class="s">&quot;PCA_normalised_libraries.pdf&quot;</span><span class="p">)</span>
<span class="nf">dba.plotPCA</span><span class="p">(</span><span class="n">res.cnt</span><span class="p">,</span><span class="n">DBA_TISSUE</span><span class="p">,</span><span class="n">label</span><span class="o">=</span><span class="n">DBA_TISSUE</span><span class="p">)</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="c1"># setting the contrast</span>
<span class="n">res.cnt2</span> <span class="o">=</span> <span class="nf">dba.contrast</span><span class="p">(</span><span class="n">res.cnt</span><span class="p">,</span> <span class="n">categories</span><span class="o">=</span><span class="n">DBA_TISSUE</span><span class="p">,</span> <span class="n">minMembers</span><span class="o">=</span><span class="m">2</span><span class="p">)</span>

<span class="c1"># inspecting the object: how many contrasts were set in the previous step</span>
<span class="n">res.cnt2</span>

<span class="c1"># performing analysis of differential binding</span>
<span class="n">res.cnt3</span> <span class="o">=</span> <span class="nf">dba.analyze</span><span class="p">(</span><span class="n">res.cnt2</span><span class="p">)</span>

<span class="c1"># inspecting the object: which condition are most alike, which are most different, is this in line with part one of the tutorial?</span>
<span class="nf">dba.show</span><span class="p">(</span><span class="n">res.cnt3</span><span class="p">,</span> <span class="n">bContrasts</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>

<span class="c1"># correlation heatmap  using only significantly differentially bound sites</span>
<span class="c1"># choose the contrast of interest e.g. HeLa vs. neuronal (#1)</span>
<span class="nf">pdf</span><span class="p">(</span><span class="s">&quot;correlation_HeLa_vs_neuronal.pdf&quot;</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">res.cnt3</span><span class="p">,</span> <span class="n">contrast</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="c1"># boxplots to view how read distributions differ between classes of binding sites</span>
<span class="c1"># are reads distributed evenly between those that increase binding affinity HeLa vs. in neuronal?</span>
<span class="nf">pdf</span><span class="p">(</span><span class="s">&quot;Boxplot_HeLa_vs_neuronal.pdf&quot;</span><span class="p">)</span>
<span class="n">pvals</span> <span class="o">&lt;-</span> <span class="nf">dba.plotBox</span><span class="p">(</span><span class="n">res.cnt3</span><span class="p">,</span> <span class="n">contrast</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="c1"># extracting differentially binding sites in GRanges</span>
<span class="n">res.db1</span> <span class="o">=</span> <span class="nf">dba.report</span><span class="p">(</span><span class="n">res.cnt3</span><span class="p">,</span> <span class="n">contrast</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">res.db1</span><span class="p">)</span>

<span class="c1"># plotting overlaps of sites bound by REST in different cell types</span>
<span class="nf">pdf</span><span class="p">(</span><span class="s">&quot;binding_site_overlap.pdf&quot;</span><span class="p">)</span>
<span class="nf">dba.plotVenn</span><span class="p">(</span><span class="n">res.cnt3</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span> <span class="n">label1</span><span class="o">=</span><span class="s">&quot;HeLa&quot;</span><span class="p">,</span><span class="n">label2</span><span class="o">=</span><span class="s">&quot;neuron&quot;</span><span class="p">,</span><span class="n">label3</span><span class="o">=</span><span class="s">&quot;HepG2&quot;</span><span class="p">,</span><span class="n">label4</span><span class="o">=</span><span class="s">&quot;sknsh&quot;</span><span class="p">)</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="c1"># finally, let&#39;s save our R session including the generated data. We will need everything in the next section</span>
<span class="nf">save.image</span><span class="p">(</span><span class="s">&quot;diffBind.RData&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="functional-analysis">
<h2><a class="toc-backref" href="#id26">Functional analysis</a><a class="headerlink" href="#functional-analysis" title="Permalink to this headline">¶</a></h2>
<p>So now we have list of differentially bound sites for comparisons of interest but we do not know much about them besides the genomic location. It is time to them in a biological context. To do so, we will use another <code class="docutils literal notranslate"><span class="pre">Bioconductor</span></code> package <a class="reference external" href="http://bioconductor.org/packages/release/bioc/vignettes/ChIPpeakAnno/inst/doc/pipeline.html">ChIPpeakAnno</a>.</p>
<p>ChIPpeakAnno <cite>“is for facilitating the downstream analysis for ChIP-seq experiments. It includes functions to find the nearest gene, exon, miRNA or custom features such as the most conserved elements and other transcription factor binding sites supplied by users, retrieve the sequences around the peak, obtain enriched Gene Ontology (GO) terms or pathways. Starting 2.0.5, new functions have been added for finding the peaks with bi-directional promoters with summary statistics (peaksNearBDP), for summarizing the occurrence of motifs in peaks (summarizePatternInPeaks) and for adding other IDs to annotated peaks or enrichedGO (addGeneIDs). Starting 3.4, permutation test has been added to determine whether there is a significant overlap between two sets of peaks. In addition, binding patterns of multiple transcription factors (TFs) or distributions of multiple epigenetic markers around genomic features could be visualized and compared easily using a side-by-side heatmap and density plot.</cite></p>
<p>Here, we will annotate deferentially bound sites, summarise them in a genomic feature context and obtain enriched GO terms and pathways.</p>
<div class="section" id="setting-up-chippeakanno">
<h3><a class="toc-backref" href="#id27">Setting-up <code class="docutils literal notranslate"><span class="pre">ChIPpeakAnno</span></code></a><a class="headerlink" href="#setting-up-chippeakanno" title="Permalink to this headline">¶</a></h3>
<p>We will continue our R-Studio session. If you have logged-out or lost connection or simply want to start fresh follow setting up instructions for running DiffBind locally.</p>
<p>To install ChIPpeakAnno</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">requireNamespace</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
    <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;ChIPpeakAnno&quot;</span><span class="p">,</span> <span class="n">version</span> <span class="o">=</span> <span class="s">&quot;3.8&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We will also need to load DiffBind results saved in the differential binding session. We will build on them.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">load</span><span class="p">(</span><span class="s">&quot;diffBind.RData&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="running-chippeakanno">
<h3><a class="toc-backref" href="#id28">Running <code class="docutils literal notranslate"><span class="pre">ChIPpeakAnno</span></code></a><a class="headerlink" href="#running-chippeakanno" title="Permalink to this headline">¶</a></h3>
<p>Like with DiffBind package there is a nice <a class="reference external" href="http://bioconductor.org/packages/release/bioc/vignettes/ChIPpeakAnno/inst/doc/pipeline.html#annotate-peaks">ChIPpeakAnno tutorial</a> that you can view along this exercise to read more about the various steps.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Loading DiffBind library</span>
<span class="c1"># we will need it to extract interesting peaks for down-stream analysis</span>
<span class="nf">library</span><span class="p">(</span><span class="n">DiffBind</span><span class="p">)</span>

<span class="c1"># Loading ChIPpeakAnno library</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ChIPpeakAnno</span><span class="p">)</span>

<span class="c1"># Loading TSS Annotation For Human Sapiens (GRCh37) Obtained From BiomaRt</span>
<span class="nf">data</span><span class="p">(</span><span class="n">TSS.human.GRCh37</span><span class="p">)</span>

<span class="c1"># Choosing the peaks for the interesting comparison, e.g.</span>
<span class="n">data.peaks</span> <span class="o">=</span> <span class="nf">dba.report</span><span class="p">(</span><span class="n">res.cnt3</span><span class="p">,</span> <span class="n">contrast</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">data.peaks</span><span class="p">)</span>

<span class="c1"># Annotate peaks with information on closest TSS using precompiled annotation data</span>
<span class="n">data.peaksAnno</span><span class="o">=</span><span class="nf">annotatePeakInBatch</span><span class="p">(</span><span class="n">data.peaks</span><span class="p">,</span> <span class="n">AnnotationData</span><span class="o">=</span><span class="n">TSS.human.GRCh37</span><span class="p">)</span>

<span class="c1"># View annotated peaks: can you see the added information in comparsition to data.peaks?</span>
<span class="nf">head</span><span class="p">(</span><span class="n">data.peaksAnno</span><span class="p">)</span>

<span class="c1"># Saving results</span>
<span class="nf">write.table</span><span class="p">(</span><span class="n">data.peaksAnno</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s">&quot;peaks_HeLa_vs_neuronal.txt&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> <span class="n">row.names</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
</pre></div>
</div>
<p>### Loading GO and REACTOME database &lt;a name=”FA_go_run”&gt;
Locally, we can install few more R libraries and annotation data to inspect our peaks a bit more. We will need libraries <cite>org.Hs.eg.db</cite>, <cite>TxDb.Hsapiens.UCSC.hg19.knownGene</cite> and <cite>reactome.db</cite>. To install:</p>
<p><a href="#id3"><span class="problematic" id="id4">``</span></a><a href="#id5"><span class="problematic" id="id6">`</span></a>bash</p>
<dl class="simple">
<dt>if (!requireNamespace(“BiocManager”, quietly = TRUE))</dt><dd><p>install.packages(“BiocManager”)</p>
</dd>
</dl>
<p>BiocManager::install(“org.Hs.eg.db”, version = “3.8”)</p>
<dl class="simple">
<dt>if (!requireNamespace(“BiocManager”, quietly = TRUE))</dt><dd><p>install.packages(“BiocManager”)</p>
</dd>
</dl>
<p>BiocManager::install(“reactome.db”, version = “3.8”)</p>
<dl class="simple">
<dt>if (!requireNamespace(“BiocManager”, quietly = TRUE))</dt><dd><p>install.packages(“BiocManager”)</p>
</dd>
</dl>
<p>BiocManager::install(“TxDb.Hsapiens.UCSC.hg19.knownGene”, version = “3.8”)</p>
<p><a href="#id7"><span class="problematic" id="id8">``</span></a><a href="#id9"><span class="problematic" id="id10">`</span></a></p>
<p>### Enriched GO / REACTOME terms  &lt;a name=”FA_go_run”&gt;</p>
<p><a href="#id11"><span class="problematic" id="id12">``</span></a><a href="#id13"><span class="problematic" id="id14">`</span></a>bash</p>
<p>library(org.Hs.eg.db)
library(reactome.db)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)</p>
<p># Peak distribution over genomic features
txdb &lt;- TxDb.Hsapiens.UCSC.hg19.knownGene
peaks.featuresDist&lt;-assignChromosomeRegion(data.peaksAnno, nucleotideLevel=FALSE, precedence=c(“Promoters”, “immediateDownstream”, “fiveUTRs”, “threeUTRs”,”Exons”, “Introns”), TxDb=txdb)</p>
<p>pdf(“peaks_featuresDistr_HeLa_vs_neuronal.pdf”)
par(mar=c(5, 10, 4, 2) + 0.1)
barplot(peaks.featuresDist$percentage, las=1, horiz=T)
dev.off()</p>
<p># GO ontologies
peaks.go &lt;- getEnrichedGO(data.peaksAnno, orgAnn=”org.Hs.eg.db”, maxP=.1, minGOterm=10, multiAdjMethod=”BH”, condense=TRUE)</p>
<p># Preview GO ontologies results
head(peaks.go$bp[, 1:2])
head(peaks.go$mf[, 1:2])
head(peaks.go$cc[, 1:2])</p>
<p># REACTOME pathways
peaks.pathways &lt;- getEnrichedPATH(data.peaksAnno, “org.Hs.eg.db”, “reactome.db”, maxP=.05)</p>
<p># REACTOME pathways: preview data
head(peaks.pathways)</p>
<p># REACTOME pathways: list all pathways
print(unique(peaks.pathways$PATH))</p>
<p><a href="#id15"><span class="problematic" id="id16">``</span></a><a href="#id17"><span class="problematic" id="id18">`</span></a></p>
<p>Feel free to build more on the exercises. Follow the [ChIPpeakAnno tutorial](<a class="reference external" href="http://bioconductor.org/packages/release/bioc/vignettes/ChIPpeakAnno/inst/doc/pipeline.html#annotate-peaks">http://bioconductor.org/packages/release/bioc/vignettes/ChIPpeakAnno/inst/doc/pipeline.html#annotate-peaks</a>) for ideas.</p>
<p>## Concluding remarks and next steps &lt;a name=”Next”&gt;</p>
<p>The workflow presented in the tutorials is quite common and it includes recommended steps for analysis of ChIP-seq data. Naturally, there may be different tools or ways to preform similar tasks. New tools are being developed all the time and no single tool can do it all.</p>
<p>In the extra labs we have prepared you can find for instance an alternative way of quality control of ChIP-seq data with R package called <cite>ChIPQC</cite> as well as alternative differential binding workflow with a packaged called <cite>csaw</cite>. Note, these labs were not extensively tested so you may need to experiment and draw from the knowledge gained in the main labs.</p>
<p>Also, there are more types of analyses one can do beyond the one presented here. A common further analysis, for instance, includes identification of short sequence motifs enriched in regions bound by the assayed factor (peaks). There are several tools available here and we recommend you test one or two with on the tutorial data: [Homer](<a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a>), [GEM](<a class="reference external" href="http://groups.csail.mit.edu/cgs/gem/">http://groups.csail.mit.edu/cgs/gem/</a>), [RSAT](<a class="reference external" href="http://floresta.eead.csic.es/rsat/peak-motifs_form.cgi)m">http://floresta.eead.csic.es/rsat/peak-motifs_form.cgi)m</a> [MEME](<a class="reference external" href="http://meme-suite.org/">http://meme-suite.org/</a>)</p>
<p>Above all, we recommend that you keep trying to analyze your own data. Practice makes perfect :)</p>
<hr class="docutils" />
<p>## Appendix: figures &lt;a name=”Appendix”&gt;</p>
<p>![correlation_librarires_normalised](../figures/lab-diffBinding/correlation_libraries_normalised.pdf)</p>
<p>Fig: Correlation of libraries based on normalised counts of reads in peaks</p>
<hr class="docutils" />
<p>![PCA](../figures/lab-diffBinding/PCA_normalised_libraries.pdf)</p>
<p>Fig: PCA scores plot: data overview using normalised counts of reads in peaks</p>
<hr class="docutils" />
<p>![Heatmap](../figures/lab-diffBinding/correlation_HeLa_vs_neuronal.pdf)</p>
<p>Fig: Correlation heatmap  using only significantly differentially bound sites for HeLa and neuronal</p>
<hr class="docutils" />
<p>![Boxplot](../figures/lab-diffBinding/Boxplot_HeLa_vs_neuronal.pdf)</p>
<p>Fig: Boxplots of reads distributions between HeLa and neuronal</p>
<hr class="docutils" />
<p>![Venn](../figures/lab-diffBinding/binding_site_overlap.pdf)</p>
<p>Fig: Venn diagram of overlapping sites bound by REST in different cell types</p>
<hr class="docutils" />
<p>![Features](../figures/lab-diffBinding/peaks_featuresDistr_HeLa_vs_neuronal.pdf)</p>
<p>Fig: Boxplots of reads distributions between HeLa and neuronal</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="lab-diffBinding-remote.html" class="btn btn-neutral float-right" title="ChIP-seq down-stream analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../csaw/lab-csaw.html" class="btn btn-neutral float-left" title="Detection of differential binding sites using csaw" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, NBIS Epigenomics Workshop Team. All rights reserved.

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>