<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>ChIP-seq with exogenous chromatin spike - Epigenomics 2020</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../../css/theme.css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "ChIP-seq with exogenous chromatin spike";
    var mkdocs_page_input_path = "tutorials/lab-exospike.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js" defer></script>
  <script src="../../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> Epigenomics 2020</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../welcome/">Welcome</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../about/">About</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../precourse/">Precourse Information</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../schedule/">Schedule</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../">Tutorials</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">Epigenomics 2020</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
    
    <li>ChIP-seq with exogenous chromatin spike</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="chip-seq-with-exogenous-chromatin-spike">ChIP-seq with exogenous chromatin spike</h1>
<!-- version 1
still experimental, to be improved
7 xi 2018
 -->

<h2 id="requirements">Requirements</h2>
<ul>
<li>
<p>R 3.5.0 (2018-04-23) or newer</p>
</li>
<li>
<p>Bioconductor packages:</p>
<ul>
<li>ChIPSeqSpike</li>
<li>BSgenome.Hsapiens.UCSC.hg38</li>
</ul>
</li>
</ul>
<h2 id="data">Data</h2>
<p>We will use ChIP-seq of H3K79me2 from Orlando et al, 2014 ("Quantitative ChIP-Seq Normalization Reveals Global Modulation of the Epigenome"). </p>
<p>The histone H3 lysine-79 dimethyl (H3K79me2) modification is catalyzed by the DOT1L protein and is associated with the release of paused RNA Polymerase II and licensing of transcriptional elongation. This modification is typically deposited within the 5′ regions of genes.</p>
<p>The experimental design was as follows:</p>
<ul>
<li>Jurkat cells were treated with selective DOT1L inhibitor EPZ5676 to globally deplete H3K79me2;</li>
<li>chromatin from treated and untreated cells was mixed in different proportions (here we use two conditions: 100 % treated and 50-50% treated + untreated) to mimic a global change in abundance of H3K79me2</li>
<li>chromatin from <em>Drosophila</em> S2 cells was added in a 1:2 ratio (1 S2 cell per 2 Jurkat cells) which provided a constant “reference” amount of H3K79me2 per human cell. </li>
</ul>
<p>GEO accession is <code>GSE60104</code>
ENA accession is <code>PRJNA257491</code></p>
<p>files in the dataset:</p>
<table>
<thead>
<tr>
<th>sample</th>
<th>GEO accession</th>
<th>SRA accession</th>
</tr>
</thead>
<tbody>
<tr>
<td>Jurkat_K79_100%_R1</td>
<td>GSM1465008</td>
<td>SRR1536561</td>
</tr>
<tr>
<td>Jurkat_K79_100%_R2</td>
<td>GSM1464998</td>
<td>SRR1536551</td>
</tr>
<tr>
<td>Jurkat_K79_50%_R1</td>
<td>GSM1465006</td>
<td>SRR1536559</td>
</tr>
<tr>
<td>Jurkat_WCE_100%_R1</td>
<td>GSM1511469</td>
<td>SRR1584493</td>
</tr>
<tr>
<td>Jurkat_WCE_100%_R2</td>
<td>GSM1511474</td>
<td>SRR1584498</td>
</tr>
<tr>
<td>Jurkat_WCE_50%_R1</td>
<td>GSM1511467</td>
<td>SRR1584491</td>
</tr>
</tbody>
</table>
<h2 id="data-preparation">Data preparation</h2>
<p>All data processing steps were already performed.</p>
<p>Raw fastq reads were filtered so that low quality bases and adapters were removed. Reads were mapped to the composite reference which consisted of <strong>hg38</strong> and <strong>dm6</strong> using <code>bowtie</code>. Only reads with one best alignemnt were retained. Alignments were split by reference using <code>samtools</code> and were subset to <strong>chromosome 1</strong> for hg38 and <strong>chromosome 2L</strong> for dm6.</p>
<p>Quality metrics were computed for each bam split by reference genome.</p>
<p>Fixed-step bigWig files were generated as follows:</p>
<ul>
<li>
<p>genome coverage of data in <code>bam</code> files at 1 bp resolution was calculated using <code>bedtools</code></p>
</li>
<li>
<p>covareage was converted to fixed step <code>wig</code> files using https://gist.github.com/svigneau/8846527/bedgraph_to_wig.pl using step size 100</p>
</li>
<li>
<p><code>wig</code> was converted to <code>bigWig</code> using UCSC toolkit using chrom.sizes for <code>hg38</code> downloaded from UCSC genome browser</p>
</li>
</ul>
<pre><code>bedtools genomecov -bga -ibam in.bam &gt;out.bg

bedgraph_to_wig.pl --bedgraph out.bg --wig out.wig --step 100

wigToBigWig out.wig chrom.sizes final.bw

</code></pre>

<p>All files necessary to execute the code in <code>R</code> can be copied from Rackham from:</p>
<p><code>/sw/share/compstore/courses/ngsintro/chipseq/exospike/exospike.tar.gz</code></p>
<p>After copying the files please decompress the archive and note the path to folder <code>/chip_exo_spike</code> on your local system.</p>
<h2 id="fingerprint-plots">Fingerprint plots</h2>
<p>all reads mapped to <code>hg38</code> (i.e. not subset):</p>
<p><img src="../figures/lab-exospike/exo-hg38-fingerprint.png" alt="" style="width: 400px;"/><br></p>
<p>all reads mapped to <code>dm6</code> (i.e. not subset):</p>
<p><img src="../figures/lab-exospike/exo-dm6-fingerprint.png" alt="" style="width: 400px;"/><br></p>
<h2 id="disclaimer">Disclaimer</h2>
<p>Please be aware that this is an experimental code, and as such does not represent any golden standard for analyses of this type. This is my exploration of the topic of using exogenous chromatic spike for ChIP-seq. I will aim to keep updating it with further steps of the analysis, once I get there.</p>
<h2 id="using-chipseqspike-for-chipseq-signal-scaling">Using <code>ChIPSeqSpike</code> for ChIPseq signal scaling</h2>
<p>This workflow is based on https://github.com/descostesn/BiocNYC-ChIPSeqSpike.</p>
<p>The scaling procedure works on computers with <em>non-Windows</em> operating systems. This includes Uppmax, so you can use <code>salloc</code> command to book a node and follow the workflow remotely.</p>
<h3 id="files-and-directories">Files and directories</h3>
<p>In <code>R</code>:</p>
<pre><code>workdir=&quot;/path/to/chip_exo_spike&quot;
setwd(workdir)

bam_path=file.path(workdir,&quot;bam&quot;)
bw_path=file.path(workdir,&quot;tracks&quot;)
exp_data=file.path(workdir,&quot;exp_data.csv&quot;)

#you will have to copy the initial bigwig tracks to the output folder at a later stage
#output_folder=file.path(workdir,&quot;results&quot;)
#dir.create(output_folder)

#so until the package code is fixed:
output_folder=bw_path
</code></pre>

<p>You can inspect the file <code>exp_data.csv</code> to familiarize yourself with the structure:</p>
<pre><code>info=read.table(exp_data, sep=&quot;,&quot;)
head(info)
</code></pre>

<h3 id="scaling-of-signal-to-exogenous-chromatin-spike">Scaling of signal to exogenous chromatin spike</h3>
<p>Load the library and create the object:</p>
<pre><code>library(ChIPSeqSpike)

cs &lt;- spikeDataset(exp_data, bam_path, bw_path)
</code></pre>

<p>Calculate the size factors based on numbers of mapped reads:</p>
<pre><code>cs &lt;- estimateScalingFactors(cs, verbose = TRUE)
</code></pre>

<pre><code>&gt; spikeSummary(cs)
                endoScalFact exoScalFact endoCount exoCount
H3K79me2_0         0.5367522   1.0216143   1863057   978843
input              1.1604563          NA    861730       NA
H3K79me2_50        0.6604427   0.7663511   1514136  1304885
input              2.9039209          NA    344362       NA
H3K79me2_100_r1    1.5994012   0.3687641    625234  2711761
input              2.5008003          NA    399872       NA
H3K79me2_100_r2    2.6171433   0.6153835    382096  1625003
input              7.7456934          NA    129104       NA
</code></pre>

<p><strong>RPM scaling.</strong> The first normalization applied to the data is the ‘Reads Per Million’ (RPM) mapped reads. The method ‘scaling’ is used to achieve this normalization using default parameters.</p>
<pre><code>cs &lt;- scaling(cs, outputFolder = output_folder)
</code></pre>

<p>You are supposed to obtain files <code>*-RPM.bw</code> after this step.</p>
<p><strong>Input subtraction.</strong> This step is to subtract background (from input samples) from signal. The <code>inputSubtraction</code> method simply subtracts scores of the input DNA experiment from the corresponding ones.</p>
<pre><code>cs &lt;- inputSubtraction(cs)
</code></pre>

<p>You are supposed to obtain <code>*-RPM-BGSub.bw</code> after this step.</p>
<p><strong>RPM scaling reversal.</strong> After RPM and input subtraction normalization, the RPM normalization is reversed in order for the data to be normalized by the exogenous scaling factors.</p>
<pre><code>cs&lt;- scaling(cs, reverse = TRUE)
</code></pre>

<p><code>*-RPM-BGSub-reverted.bw</code> files after this step.</p>
<p><strong>Exogenous Scaling.</strong> Finally, exogenous scaling factors are applied to the data.</p>
<pre><code>cs &lt;- scaling(cs, type = &quot;exo&quot;)
</code></pre>

<p>The end result: <code>*-RPM-BGSub-reverted-spiked.bw</code> files after this step.</p>
<p><strong>Extracting binding values.</strong> The last step of data processing is to extract and format binding scores in order to use plotting methods. The <code>extractBinding</code> method extracts binding scores at different locations and stores these values in the form of PlotSetArray objects and matrices. The scores are retrieved on annotations provided in a <code>gff</code> file. If one wishes to focus on peaks, their coordinates should be submitted at this step. The genome name must also be provided. For details about installing the required <code>BSgenome</code> package corresponding to the endogenous organism, see the BSgenome package documentation.</p>
<p>Please note that this steps may take a long time.</p>
<pre><code>gff=file.path(workdir,&quot;hg38_refseq_chr1.gtf&quot;)
library(BSgenome.Hsapiens.UCSC.hg38)

cs &lt;- extractBinding(cs, gff_vec=gff, genome=&quot;hg38&quot;)
</code></pre>

<p>After this step, save the workspace</p>
<pre><code>save.image(file = &quot;chipseqspike.RData&quot;)
</code></pre>

<p>To load the data:</p>
<pre><code>load(&quot;chipseqspike.RData&quot;)
</code></pre>

<h3 id="data-visualization">Data visualization.</h3>
<p><code>ChIPSeqSpike</code> offers several graphical methods for normalization diagnosis and data exploration. These choices enable one to visualize each step of the normalization through exploring intersamples differences using profiles, heatmaps, boxplots and correlation plots.</p>
<p>When performing this exercise on Uppmax, save the plots to pdf for viewing:</p>
<pre><code>pdf(&quot;filename.pdf&quot;)
## here command to produce the plot
dev.off()
</code></pre>

<h4 id="visualization-with-gene-meta-profiles">Visualization with gene meta-profiles</h4>
<p>The first step of spike-in normalized ChIP-Seq data analysis is an inter-sample comparison by meta-gene or meta-annotation profiling. The method <code>plotProfile</code> automatically plots all experiments at the start, midpoint, end and composite locations of the annotations provided to the method extractBinding in gff format. The effect of each transformation on a particular experiment can be visualized with <code>plotTransform</code>.</p>
<pre><code>## Plot spiked-in data
plotProfile(cs, legends = TRUE)

## Add profiles before transformation
plotProfile(cs, legends = TRUE, notScaled=TRUE)

## Visualize the effect of each transformation on each experiment
plotTransform(cs, legends = TRUE, separateWindows = TRUE)
</code></pre>

<h4 id="visualization-with-boxplots">Visualization with Boxplots</h4>
<p><code>boxplotSpike</code> plots boxplots of the mean values of ChIP-seq experiments on the annotations given to the <code>extractBinding</code> method. </p>
<pre><code>## Boxplot of the spiked-in data
boxplotSpike(cs, outline = FALSE)

## Boxplot of the raw data
boxplotSpike(cs,rawFile = TRUE, spiked = FALSE, outline=FALSE)

## Boxplot of all transformations
boxplotSpike(cs,rawFile = TRUE, rpmFile = TRUE, bgsubFile = TRUE, revFile = TRUE, spiked = TRUE, outline = FALSE)
</code></pre>

<h4 id="correlation-plots">Correlation plots</h4>
<p>The <code>plotCor</code> method plots the correlation between ChIP-seq experiments using heatscatter plot.</p>
<pre><code>## Log transform correlation plot of spiked data with heatscatter representation
plotCor(cs, rawFile = FALSE, rpmFile = FALSE,  bgsubFile = FALSE,  revFile = FALSE, spiked = TRUE,  main = &quot;heatscatter&quot;,  method_cor = &quot;spearman&quot;, add_contour = FALSE,  nlevels = 10,  color_contour = &quot;black&quot;, method_scale = &quot;log&quot;,  allOnPanel = TRUE, separateWindows = FALSE,  verbose = FALSE)

## Plot as above with raw data
plotCor(cs, rawFile = TRUE, rpmFile = FALSE,  bgsubFile = FALSE,  revFile = FALSE, spiked = FALSE,  main = &quot;heatscatter&quot;,  method_cor = &quot;spearman&quot;, add_contour = FALSE,  nlevels = 10,  color_contour = &quot;black&quot;, method_scale = &quot;log&quot;,  allOnPanel = TRUE, separateWindows = FALSE,  verbose = FALSE)

## Correlation table comparing all transformations
corr_matrix &lt;- plotCor(cs, rawFile = TRUE, rpmFile = TRUE, bgsubFile = TRUE, revFile = TRUE, spiked = TRUE, heatscatterplot = FALSE, verbose = TRUE)
</code></pre>

<h2 id="what-to-do-next">What to do next</h2>
<ul>
<li>use scaled <code>bigWig</code> tracks to view the signal in IGV</li>
<li>use <code>bed</code> files produced from scaled <code>bigWigs</code> to perform peak calling for instance with MACS2</li>
<li>differential binding analysis using <code>csaw</code> (more appropriate for broad marks) inputing the scaling factors obtained from scaling by <code>ChIPSeqSpike</code></li>
<li>perform the normalisation / scaling directly in <code>csaw</code>    </li>
<li>use scaled <code>bed</code> / <code>bigwig</code> for data exploration using PCA and MA plots</li>
</ul>
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
