

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Integrating scATAC-seq and scRNA-seq data &mdash; Epigenomics Workshop 2020 1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Genomic overlaps" href="../genomeOverlap/lab-genomicGoverlaps.html" />
    <link rel="prev" title="Data integration" href="../data_integration_tutorials.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Epigenomics Workshop 2020
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../precourse.html">Pre-course Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../online-classroom.html">Online classroom</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../schedule.html">Schedule</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../setup/lab-setup.html">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="../methylation_tutorials.html">DNA Methylation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chipseq_tutorials.html">ChIPseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../atacseq_tutorials.html">ATACseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quantitative_chip_tutorials.html">Advanced ChIP methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chip_downstream_tutorials.html">Downstream Processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../expdesign_data.html">Experimental Design and Data Considerations</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../data_integration_tutorials.html">Data Integration</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">scATAC-seq analysis and integraion with scRNA-seq data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#learning-outcomes">Learning outcomes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#setting-up">Setting up</a></li>
<li class="toctree-l4"><a class="reference internal" href="#analysis-of-single-cell-atac-seq-data">Analysis of single cell ATAC-seq data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#integrate-with-scrna-seq-data">Integrate with scRNA-seq data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../genomeOverlap/lab-genomicGoverlaps.html">Genomic overlaps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../unsupervised_data_integration/lab-unsupervised_data_integration.html">BONUS EXERCISE: Unsupervised data integration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../nextflow.html">Nextflow &amp; nf-core</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Epigenomics Workshop 2020</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../tutorials.html">Tutorials</a> &raquo;</li>
        
          <li><a href="../data_integration_tutorials.html">Data integration</a> &raquo;</li>
        
      <li>Integrating scATAC-seq and scRNA-seq data</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../../_sources/content/tutorials/scAtacSeq/lab-sc_atac_seq.md.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="integrating-scatac-seq-and-scrna-seq-data">
<h1>Integrating scATAC-seq and scRNA-seq data<a class="headerlink" href="#integrating-scatac-seq-and-scrna-seq-data" title="Permalink to this headline">¶</a></h1>
<div class="section" id="learning-outcomes">
<h2>Learning outcomes<a class="headerlink" href="#learning-outcomes" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Have a basic understaing of the steps for quality control, normalization, dimensionality reduction, clustering and visualization of single cell ATAC-seq data</p></li>
<li><p>Have a basic understanding of the steps to combine single cell ATAC-seq data with RNA-seq data, to better infer cell types.</p></li>
</ul>
<p><strong>Table of contents</strong></p>
<ul class="simple">
<li><p><a class="reference external" href="#setting-up">Setting up</a></p></li>
<li><p><a class="reference external" href="#analysis-of-single-cell-atac-seq-data">Analysis of single cell ATAC-seq data</a></p>
<ul>
<li><p><a class="reference external" href="#loading-scatac-seq-data">Loading scATAC-seq data</a></p></li>
<li><p><a class="reference external" href="#quality-control">Quality control</a></p></li>
<li><p><a class="reference external" href="#normalization-and-intial-dimensionality-reduction">Normalization and intial dimensionality reduction.</a></p></li>
</ul>
</li>
<li><p><a class="reference external" href="#clustering-and-further-dimensionality-reduction">Clustering and further dimensionality reduction</a></p>
<ul>
<li><p><a class="reference external" href="#motif-analysis">Motif analysis</a></p>
<ul>
<li><p><a class="reference external" href="#identifying-enriched-motifs">Identifying enriched motifs</a></p></li>
<li><p><a class="reference external" href="#motif-activity-scores">Motif activity scores</a></p></li>
</ul>
</li>
</ul>
</li>
<li><p><a class="reference external" href="#integrate-with-scrna-seq-data">Integrate with scRNA-seq data</a></p>
<ul>
<li><p><a class="reference external" href="#gene-activity-matrix">Gene activity matrix</a></p></li>
<li><p><a class="reference external" href="#label-transfer">Label transfer</a></p></li>
<li><p><a class="reference external" href="#differentially-accessible-regions--again">Differentially accessible regions, again</a></p></li>
<li><p><a class="reference external" href="#annotating-regions">Annotating regions</a></p></li>
<li><p><a class="reference external" href="#visualization">Visualization</a></p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="setting-up">
<h2>Setting up<a class="headerlink" href="#setting-up" title="Permalink to this headline">¶</a></h2>
<p>This exercise has been tested on Uppmax, using <code class="docutils literal notranslate"><span class="pre">R</span> <span class="pre">4.0.0</span></code>. On Uppmax, most packages are already installed, and can be loaded into R after the <code class="docutils literal notranslate"><span class="pre">R_packages</span></code> module has been loaded. If you are running on Uppmax, start by loading the following modules:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">module</span> <span class="n">load</span> <span class="n">bioinfo</span><span class="o">-</span><span class="n">tools</span>
<span class="n">module</span> <span class="n">load</span> <span class="n">R</span><span class="o">/</span><span class="mf">4.0</span><span class="o">.</span><span class="mi">0</span>
<span class="n">module</span> <span class="n">load</span> <span class="n">R_packages</span><span class="o">/</span><span class="mf">4.0</span><span class="o">.</span><span class="mi">0</span>
<span class="n">module</span> <span class="n">load</span> <span class="n">RStudio</span>
<span class="n">module</span> <span class="n">load</span> <span class="n">samtools</span>
</pre></div>
</div>
<p>The data files needed for this exercise are in the following Uppmax directory <code class="docutils literal notranslate"><span class="pre">/sw/courses/epigenomics/sc_atac_seq</span></code>. If you are working on Uppmax you can create soft links from your working directory to these files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="o">/</span><span class="n">sw</span><span class="o">/</span><span class="n">courses</span><span class="o">/</span><span class="n">epigenomics</span><span class="o">/</span><span class="n">sc_atac_seq</span><span class="o">/*</span> <span class="o">.</span>
</pre></div>
</div>
<p>If you run this on some other system, you can copy the files with <code class="docutils literal notranslate"><span class="pre">scp</span></code>.</p>
<p>Now, start <code class="docutils literal notranslate"><span class="pre">R</span></code> or <code class="docutils literal notranslate"><span class="pre">rstudio</span></code>, and run these commands to install some additional libraries. We need to install the latest version of Signac, which takes a while.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># Install bioconductor packages
if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)
BiocManager::install(&quot;JASPAR2020&quot;)

# Install latest version of Signac from GitHub
if (!requireNamespace(&quot;devtools&quot;, quietly = TRUE))
    install.packages(&quot;devtools&quot;)
devtools::install_github(&quot;timoast/signac&quot;, ref = &quot;develop&quot;)
</pre></div>
</div>
<p>If you want to run this exercise on your laptop, you also need to make sure that all libraries loaded below are installed, as well as all dependencies.</p>
<p><img alt="George Seurat" src="../../../_images/Seurat-Gravelines-Annonciade.jpg" /></p>
</div>
<div class="section" id="analysis-of-single-cell-atac-seq-data">
<h2>Analysis of single cell ATAC-seq data<a class="headerlink" href="#analysis-of-single-cell-atac-seq-data" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://satijalab.org/seurat/">Seurat</a> is the most widley used tool to analyze scRNA-seq data. Recently, this R package has been extended to support chromatin data, e.g. ATAC. This extension package is called <a class="reference external" href="https://satijalab.org/signac/index.html">Signac</a>. Seurat makes it possbile to integrate data from different technologies. Here, we will look at how Seurat and Signac can be used to integrate scATAC-seq and scRNA-seq data. This exercise is based on <a class="reference external" href="https://satijalab.org/signac/articles/pbmc_vignette.html">this</a> and <a class="reference external" href="https://satijalab.org/signac/articles/motif_vignette.html">this</a> tutorial, using data on human peripheral blood mononuclear cells (PBMCs) provided by 10x Genomics. We will use data that have already been pre-processed using CellRanger. The starting point is a count matrix, with the number of reads in each peak in each cell, along with some meta data.</p>
<p>We start by loading the required packages: Seurat, Signac, some annotation packages and some packages for plotting.</p>
<p><img alt="Paul Signac" src="../../../_images/Paul_signac,_la_barca_a_vela_verde,_1904,_01.jpg" /></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">Signac</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">GenomeInfoDb</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">EnsDb</span><span class="o">.</span><span class="n">Hsapiens</span><span class="o">.</span><span class="n">v75</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">patchwork</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">JASPAR2020</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">TFBSTools</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">BSgenome</span><span class="o">.</span><span class="n">Hsapiens</span><span class="o">.</span><span class="n">UCSC</span><span class="o">.</span><span class="n">hg19</span><span class="p">)</span>
<span class="nb">set</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1234</span><span class="p">)</span>
</pre></div>
</div>
<div class="section" id="loading-scatac-seq-data">
<h3>Loading scATAC-seq data<a class="headerlink" href="#loading-scatac-seq-data" title="Permalink to this headline">¶</a></h3>
<p>The ATAC-seq data, consists of four files, that are created with CellRanger.</p>
<ul class="simple">
<li><p>A count file. The rows are regions (peaks) and the colums are cells. Each entry <em>i,j</em> is the number of reads mapping to region <em>i</em> in cell <em>j</em>.</p></li>
<li><p>A meta data file, with some overall statistics for each cell</p></li>
<li><p>A fragment file, with information on all sequenced fragments (where it maps to the genome, which cell barcode is associated and how many PCR duplicates were found)</p></li>
<li><p>An index file connected to the fragment file. This is like an index file for a bam file, to make it possible to quickly find fragments for a certain genomic region, with having to search the entire file.</p></li>
</ul>
<p>The PBMC data set contains ATAC-seq data on 74836 regions in 9277 cells. The corresponding RNA-seq data covers 19089 genes in 9432 cells. To make the commands in this exercise run a bit faster, we will only analyze a set of 2000 (randomly selected) cells. If you have the time and interest, you can analyze the full data set, but commenting out the corresponding line of code.</p>
<p>Here we create a <code class="docutils literal notranslate"><span class="pre">ChromatinAssay</span></code> object from the count matrix (and a link to the fragment file). This object stores the genomic regions, count data, and also possibly gene annotations and information on sequence motifs. There are many ways to interact with a <code class="docutils literal notranslate"><span class="pre">ChromatinAssay</span></code> object, see [this] (https://satijalab.org/signac/articles/data_structures.html) vignette.</p>
<p>We then create a <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> object from the <code class="docutils literal notranslate"><span class="pre">ChromatinAssay</span></code>, together with meta data about cells and gene annotations.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">counts</span> <span class="o">&lt;-</span> <span class="n">Read10X_h5</span><span class="p">(</span><span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5&quot;</span><span class="p">)</span>
<span class="n">counts</span> <span class="o">&lt;-</span> <span class="n">counts</span><span class="p">[,</span> <span class="n">sample</span><span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">counts</span><span class="p">),</span> <span class="mi">2000</span><span class="p">)]</span> <span class="c1"># Only use 2000 cells, to make commands run faster.</span>

<span class="n">metadata</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span>
	<span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;atac_v1_pbmc_10k_singlecell.csv&quot;</span><span class="p">,</span>
	<span class="n">header</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
	<span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="mi">1</span>
<span class="p">)</span>

<span class="n">chrom_assay</span> <span class="o">&lt;-</span> <span class="n">CreateChromatinAssay</span><span class="p">(</span>
	<span class="n">counts</span> <span class="o">=</span> <span class="n">counts</span><span class="p">,</span>
	<span class="n">sep</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">),</span>
	<span class="n">genome</span> <span class="o">=</span> <span class="s1">&#39;hg19&#39;</span><span class="p">,</span>
	<span class="n">fragments</span> <span class="o">=</span> <span class="s1">&#39;atac_v1_pbmc_10k_fragments.tsv.gz&#39;</span><span class="p">,</span>
	<span class="nb">min</span><span class="o">.</span><span class="n">cells</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
	<span class="nb">min</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="mi">200</span>
<span class="p">)</span>

<span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="n">CreateSeuratObject</span><span class="p">(</span>
	<span class="n">counts</span> <span class="o">=</span> <span class="n">chrom_assay</span><span class="p">,</span>
	<span class="n">assay</span> <span class="o">=</span> <span class="s2">&quot;peaks&quot;</span><span class="p">,</span>
	<span class="n">meta</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">metadata</span>
<span class="p">)</span>

<span class="c1"># Extract gene annotations from EnsDb</span>
<span class="n">annotations</span> <span class="o">&lt;-</span> <span class="n">GetGRangesFromEnsDb</span><span class="p">(</span><span class="n">ensdb</span> <span class="o">=</span> <span class="n">EnsDb</span><span class="o">.</span><span class="n">Hsapiens</span><span class="o">.</span><span class="n">v75</span><span class="p">)</span>

<span class="c1"># Change to UCSC style since the data was mapped to hg19</span>
<span class="n">seqlevelsStyle</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s1">&#39;UCSC&#39;</span>
<span class="n">genome</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s2">&quot;hg19&quot;</span>

<span class="c1"># Add the gene information to the object</span>
<span class="n">Annotation</span><span class="p">(</span><span class="n">pbmc</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">annotations</span>

<span class="c1"># Inspect the objects we have created</span>
<span class="n">pbmc</span>
<span class="n">pbmc</span><span class="p">[[</span><span class="s1">&#39;peaks&#39;</span><span class="p">]]</span>
<span class="n">granges</span><span class="p">(</span><span class="n">pbmc</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="quality-control">
<h3>Quality control<a class="headerlink" href="#quality-control" title="Permalink to this headline">¶</a></h3>
<p>The next step is to do some quality control (QC) on the ATAC-seq data. There are several qualtiy measures to consider:</p>
<ul class="simple">
<li><p><strong>Fragment sizes</strong>, determined from paired-end sequencing data. For each cell we can quantify the fragment sizes by calculaing the ratio of nucleosome size fragments (147-294 nucleotides) to nucleosome-free fragments (&lt;147 nucleotides). Open chromatin correseponds to short DNA fragmens, so we want to remove cells with to few short DNA fragments, coming from open open chromatin.</p></li>
<li><p><strong>Enrichment at transcription start sites (TSS).</strong> Since chromatin is ofter open around TSS, we expect there to be an enrichment of framgents around these sites. This is quantified by collecting all fragments mapping around TSS to form an aggregate distribution, and comparing the highest level of fragments to the level of fragments 1000bp up and downstream of the TSS.</p></li>
<li><p><strong>Number of fragments in peaks</strong>. This represents the complexity and sequencing depth for each cell. Too many reads might indicate artefacts, such as several cells sharing a barcode.</p></li>
<li><p><strong>Fraction of reads in peaks</strong>. This represents how well that ATAC protocol has worked, and we typically reqiure at least 15% if all fragments to be in peaks.</p></li>
<li><p><strong>Reads in blacklist regions.</strong> The <a class="reference external" href="https://www.encodeproject.org">ENCODE</a> project has defined lists of <a class="reference external" href="https://github.com/Boyle-Lab/Blacklist">blacklist regions</a>. These are problematic regions (typically repeats) that often have high signals in next-generation sequencing experiments regardless of cell line or experiment. Cells with a realtively high ratio of reads mapping to blacklist regions, compared to peaks, often represent technical artifacts and should be removed.</p></li>
</ul>
<p>This talkes a while to run (around 10 minutes).</p>
<div class="highlight-{r seurat_qc} notranslate"><div class="highlight"><pre><span></span># Compute nucleosome signal score per cell
pbmc &lt;- NucleosomeSignal(object = pbmc)

# Compute TSS enrichment score per cell
pbmc &lt;- TSSEnrichment(object = pbmc, fast = FALSE)

# Add blacklist ratio and fraction of reads in peaks
pbmc$pct_reads_in_peaks &lt;- pbmc$peak_region_fragments / pbmc$passed_filters * 100
pbmc$blacklist_ratio &lt;- pbmc$blacklist_region_fragments / pbmc$peak_region_fragments

gc()

pbmc$high.tss &lt;- ifelse(pbmc$TSS.enrichment &gt; 2, &#39;High&#39;, &#39;Low&#39;)
TSSPlot(pbmc, group.by = &#39;high.tss&#39;) + NoLegend()

pbmc$nucleosome_group &lt;- ifelse(pbmc$nucleosome_signal &gt; 4, &#39;NS &gt; 4&#39;, &#39;NS &lt; 4&#39;)
FragmentHistogram(object = pbmc, group.by = &#39;nucleosome_group&#39;)

VlnPlot(
	object = pbmc,
	features = c(&#39;pct_reads_in_peaks&#39;, &#39;peak_region_fragments&#39;,
							 &#39;TSS.enrichment&#39;, &#39;blacklist_ratio&#39;, &#39;nucleosome_signal&#39;),
	pt.size = 0.1,
	ncol = 3
)
</pre></div>
</div>
<p><img alt="QC plots" src="../../../_images/qc.png" /></p>
<p>After calculating the qualtiy statistics, we apply (some rather artibtrary) cutoffs, to remove outlier cells. <em>How many cells do we have left after quality filtering in this example?</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="n">subset</span><span class="p">(</span>
	<span class="n">x</span> <span class="o">=</span> <span class="n">pbmc</span><span class="p">,</span>
	<span class="n">subset</span> <span class="o">=</span> <span class="n">peak_region_fragments</span> <span class="o">&gt;</span> <span class="mi">3000</span> <span class="o">&amp;</span>
		<span class="n">peak_region_fragments</span> <span class="o">&lt;</span> <span class="mi">20000</span> <span class="o">&amp;</span>
		<span class="n">pct_reads_in_peaks</span> <span class="o">&gt;</span> <span class="mi">15</span> <span class="o">&amp;</span>
		<span class="n">blacklist_ratio</span> <span class="o">&lt;</span> <span class="mf">0.05</span> <span class="o">&amp;</span>
		<span class="n">nucleosome_signal</span> <span class="o">&lt;</span> <span class="mi">4</span> <span class="o">&amp;</span>
		<span class="n">TSS</span><span class="o">.</span><span class="n">enrichment</span> <span class="o">&gt;</span> <span class="mi">2</span>
<span class="p">)</span>
<span class="n">pbmc</span>
</pre></div>
</div>
</div>
<div class="section" id="normalization-and-intial-dimensionality-reduction">
<h3>Normalization and intial dimensionality reduction.<a class="headerlink" href="#normalization-and-intial-dimensionality-reduction" title="Permalink to this headline">¶</a></h3>
<p><strong>Normalization:</strong> Signac performs term frequency-inverse document frequency (TF-IDF) normalization. This is a two-step normalization procedure, often used in natural language processing, that both normalizes across cells to correct for differences in sequencing depth, and across peaks to give higher values to more rare peaks.</p>
<p><strong>Feature selection:</strong> To reduce noise, we sometimes only use some features (i.e. peaks) when we cluster cells. Often this is the peaks with the strongest signal, i.e. most reasd, or the ones present in most cells. Here, we will use all peaks, but you can play around with this yourself. Try setting min.cutoff to ‘q75’ to use the top 25% peaks).</p>
<p><strong>Dimension reduction:</strong> We next run singular value decomposition (SVD) on the normalized data matrix, using the features (peaks) selected above. This returns a reduced dimension representation of the matrix, similar to the output of PCA.</p>
<p>From the singular value decomposition, we get a set of components. The first of these components often correlates with seuquencing depth, rather than any biologically meaningful signal. We can therefore remove this component in the following analysis.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="n">RunTFIDF</span><span class="p">(</span><span class="n">pbmc</span><span class="p">)</span>
<span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="n">FindTopFeatures</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="nb">min</span><span class="o">.</span><span class="n">cutoff</span> <span class="o">=</span> <span class="s1">&#39;q0&#39;</span><span class="p">)</span>
<span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="n">RunSVD</span><span class="p">(</span><span class="n">pbmc</span><span class="p">)</span>

<span class="n">DepthCor</span><span class="p">(</span><span class="n">pbmc</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="clustering-and-further-dimensionality-reduction">
<h3>Clustering and further dimensionality reduction<a class="headerlink" href="#clustering-and-further-dimensionality-reduction" title="Permalink to this headline">¶</a></h3>
<p>Now we can cluster the cells to find groups that belong to the same cell types. It is possible to plot the results from the SVD, but these often are not informative. Instead, we use the UMAP algorithm, which shows a better separation between the cell types. If you are interested, the paper describing UMAP can be found <a class="reference external" href="https://arxiv.org/abs/1802.03426">here</a></p>
<div class="highlight-{r seurat_umap} notranslate"><div class="highlight"><pre><span></span>pbmc &lt;- RunUMAP(object = pbmc, reduction = &#39;lsi&#39;, dims = 2:30)
pbmc &lt;- FindNeighbors(object = pbmc, reduction = &#39;lsi&#39;, dims = 2:30)
pbmc &lt;- FindClusters(object = pbmc, verbose = FALSE, algorithm = 3)

p1 &lt;- DimPlot(object = pbmc, label = TRUE, dims = c(2, 3), reduction = &quot;lsi&quot;) + 
	NoLegend()  + 
	ggtitle(&#39;SVD&#39;)

p2 &lt;- DimPlot(object = pbmc, label = TRUE) + 
	NoLegend() + 
	ggtitle(&#39;UMAP&#39;)

p1 | p2
</pre></div>
</div>
<p><img alt="Dimension reduction" src="../../../_images/dimension_reduction.png" /></p>
<p>This might be a good time to save your data, so you don’t have to re-run all your analysis if you have problems with Uppmax or the network. To load the data later, type <code class="docutils literal notranslate"><span class="pre">load(file=&quot;pbmc.Rda&quot;)</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">save</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s2">&quot;pbmc.Rda&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="motif-analysis">
<h3>Motif analysis<a class="headerlink" href="#motif-analysis" title="Permalink to this headline">¶</a></h3>
<p>We can also analyze motif occurence in the peaks, to see how this varies between the different cell types.</p>
<div class="section" id="identifying-enriched-motifs">
<h4>Identifying enriched motifs<a class="headerlink" href="#identifying-enriched-motifs" title="Permalink to this headline">¶</a></h4>
<p>First we will look at motifs that are enriched in set of peaks, e.g. in peaks that show differential accessibility between two cell types. <em>Do you notice anything particular about these motifs?</em></p>
<div class="highlight-{r enriched_motifs} notranslate"><div class="highlight"><pre><span></span># Get a list of motif position frequency matrices from the JASPAR database
pfm &lt;- getMatrixSet(
	x = JASPAR2020,
	opts = list(species = &quot;Homo sapiens&quot;, all_versions = FALSE)
)

# Scan the DNA sequence of each peak for the presence of each motif
motif.matrix &lt;- CreateMotifMatrix(
	features = granges(pbmc$peaks),
	pwm = pfm,
	genome = &#39;hg19&#39;,
	use.counts = FALSE
)
dim(motif.matrix)
as.matrix(motif.matrix[1:10,1:10])

# Create a new Mofif object to store the results
motif &lt;- CreateMotifObject(
	data = motif.matrix,
	pwm = pfm
)

# Add the Motif object to the assay
pbmc &lt;- SetAssayData(
	object = pbmc,
	assay = &#39;peaks&#39;,
	slot = &#39;motifs&#39;,
	new.data = motif
)
pbmc$peaks@motifs

# Calculate statistics for each peak: GC content, length, dinucleotide frequencies etc.
# These are used in the test of motif enrichment
pbmc$peaks &lt;- RegionStats(object = pbmc$peaks, genome = BSgenome.Hsapiens.UCSC.hg19)

# Find differentially accessible peaks in cluster 0 compared to cluster 1
da_peaks &lt;- FindMarkers(
	object = pbmc,
	ident.1 = &quot;0&quot;,
	ident.2 = &quot;1&quot;,
	only.pos = TRUE,
	min.pct = 0.2,
	test.use = &#39;LR&#39;,
	latent.vars = &#39;peak_region_fragments&#39;
)

# Get the top differentially accessible peaks, with lowest p-values
top.da.peak &lt;- rownames(da_peaks[da_peaks$p_val &lt; 1e-20, ])

# Find motifs enriched in these top differentially accessible peaks
enriched.motifs &lt;- FindMotifs(
	object = pbmc,
	features = top.da.peak
)
head(enriched.motifs)

# Have a look at the most enriched motifs. Do you anything particular about these motifs?
MotifPlot(
	object = pbmc$peaks,
	motifs = head(rownames(enriched.motifs))
)
</pre></div>
</div>
</div>
<div class="section" id="motif-activity-scores">
<h4>Motif activity scores<a class="headerlink" href="#motif-activity-scores" title="Permalink to this headline">¶</a></h4>
<p>We can also compute a per-cell motif activity score by running chromVAR. The motif activity score for a motif M is based on the on nr reads mapping to peaks with motif M, after normalization correction for various biases: GC content, average number of reads mapping across all cells etc. You can read more about chromVar <a class="reference external" href="https://www.nature.com/articles/nmeth.4401">here</a> Motif activity scores allow us to visualize motif activities per cell.</p>
<p>It is also possible to directly test for differential activity scores between cell types, without looking at peaks with differential binding. This tends to give similar results as performing an enrichment test on differentially accessible peaks between the cell types (shown above).</p>
<p>This takes a while to run (around 5 minutes on Uppmax).</p>
<div class="highlight-{r chromvar} notranslate"><div class="highlight"><pre><span></span># Use chromVAR to calculate the motif activities of all motifs in all cells.
pbmc &lt;- RunChromVAR(
	object = pbmc,
	genome = BSgenome.Hsapiens.UCSC.hg19,
	verbose = TRUE
)

# Look at results
pbmc$chromvar
GetAssayData(pbmc$chromvar)[1:10,1:3]

# Have a look at the activitiy of the FOS motif, which has id MA0476.1
DefaultAssay(pbmc) &lt;- &#39;chromvar&#39;
FeaturePlot(
	object = pbmc,
	features = &quot;MA0476.1&quot;,
	min.cutoff = &#39;q10&#39;,
	max.cutoff = &#39;q90&#39;,
	pt.size = 0.1
)

# Look for motifs that have differential activity between clusters 0 and 1. 
differential.activity &lt;- FindMarkers(
	object = pbmc,
	ident.1 = &#39;0&#39;,
	ident.2 = &#39;1&#39;,
	only.pos = TRUE,
	test.use = &#39;LR&#39;,
	min.pct = 0.2,
	latent.vars = &#39;nCount_peaks&#39;
)

MotifPlot(
  object = pbmc,
  motifs = head(rownames(differential.activity)),
  assay = &#39;peaks&#39;
)
</pre></div>
</div>
<p><img alt="Motif activity" src="../../../_images/motif_activity.png" /></p>
</div>
</div>
</div>
<div class="section" id="integrate-with-scrna-seq-data">
<h2>Integrate with scRNA-seq data<a class="headerlink" href="#integrate-with-scrna-seq-data" title="Permalink to this headline">¶</a></h2>
<p>In this section we will compare the scATAC-seq data to scRNA-seq data from a similar starting material (although not from exatcly the same cells). Since it is easier to figure out which cell types clusters correspond to in scRNA-seq data, we will then use the scRNA-seq data to annotate the clusters in the scATAC-seq data.</p>
<div class="section" id="gene-activity-matrix">
<h3>Gene activity matrix<a class="headerlink" href="#gene-activity-matrix" title="Permalink to this headline">¶</a></h3>
<p>Gene activity scores capture how much open chromatin there is in the promoter regions of each gene (by defualt 2000bp upstream). The assumption here is that open chromatin is a proxy for gene expression. Gene activity scores are represented as a matrix, with one row per gene and one column per cell. This makes the gene activitiy scores directly compatible with single cell RNA-seq data.</p>
<p>Calculating the gene activity scores takes around 10 minutes for 2000 cells and all genes.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>DefaultAssay(pbmc) &lt;- &#39;peaks&#39;

gene.activities &lt;- GeneActivity(pbmc)

# Add the gene activity matrix to the Seurat object as a new assay and normalize it
pbmc[[&#39;RNA&#39;]] &lt;- CreateAssayObject(counts = gene.activities)
pbmc &lt;- NormalizeData(
	object = pbmc,
	assay = &#39;RNA&#39;,
	normalization.method = &#39;LogNormalize&#39;,
	scale.factor = median(pbmc$nCount_RNA)
)
GetAssayData(pbmc$RNA)[1:10,1:3]

DefaultAssay(pbmc) &lt;- &#39;RNA&#39;
FeaturePlot(
	object = pbmc,
	features = c(&#39;MS4A1&#39;, &#39;CD3D&#39;, &#39;LEF1&#39;, &#39;NKG7&#39;, &#39;TREM1&#39;, &#39;LYZ&#39;),
	pt.size = 0.1,
	max.cutoff = &#39;q95&#39;,
	ncol = 3
)
</pre></div>
</div>
<p><img alt="Gene activity" src="../../../_images/gene_activity.png" /></p>
</div>
<div class="section" id="label-transfer">
<h3>Label transfer<a class="headerlink" href="#label-transfer" title="Permalink to this headline">¶</a></h3>
<p>Having computed the gene activity scores, we are now ready to combine the ATAC-seq data with the RNA-seq data.</p>
<p>We start by finding anchors, i.e. paris of cells, one from ATAC-seq and one from RNA-seq. This is done by first projecting bort ATAC-seq and RNA-seq data into the same space, and then find pairs of cells one from ATAC-seq and the other from RNA-seq that are mutual nearest neighbors (MNNs). These are further filtered and the reliable pairs are the used as anchors.</p>
<p>These anchors can then be used to project the ATAC-seq data onto the RNA-seq data and find the cell type annotation of the nearby cells. This was annotations from the RNA-seq data can be transferred to the ATAC-seq data. This is sometimes referred to as <strong>label transfer</strong></p>
<p><em>How well do you think this worked? Can you see if some cells types are missing or merged?</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># Load the pre-processed scRNA-seq data for PBMCs
pbmc_rna &lt;- readRDS(&quot;pbmc_10k_v3.rds&quot;)

# Find anchors
transfer.anchors &lt;- FindTransferAnchors(
	reference = pbmc_rna,
	query = pbmc,
	reduction = &#39;cca&#39;
)

predicted.labels &lt;- TransferData(
	anchorset = transfer.anchors,
	refdata = pbmc_rna$celltype,
	weight.reduction = pbmc[[&#39;lsi&#39;]],
	dims = 2:30
)

pbmc &lt;- AddMetaData(object = pbmc, metadata = predicted.labels)
gc()


plot1 &lt;- DimPlot(
	object = pbmc_rna,
	group.by = &#39;celltype&#39;,
	label = TRUE,
	repel = TRUE) + NoLegend() + ggtitle(&#39;scRNA-seq&#39;)

plot2 &lt;- DimPlot(
	object = pbmc,
	group.by = &#39;predicted.id&#39;,
	label = TRUE,
	repel = TRUE) + NoLegend() + ggtitle(&#39;scATAC-seq&#39;)

plot3 &lt;- DimPlot(
  object = pbmc, 
  label = TRUE, 
  repel = TRUE) + NoLegend() + ggtitle(&#39;scATAC-seq&#39;)

plot1 | plot2 | plot3

pbmc &lt;- RenameIdents(
	object = pbmc,
	&#39;0&#39; = &#39;CD14 Mono&#39;,
	&#39;1&#39; = &#39;CD4/8 Naive&#39;,
	&#39;2&#39; = &#39;CD4 Memory (DN T)&#39;,
	&#39;3&#39; = &#39;CD14 Mono&#39;,
	&#39;4&#39; = &#39;CD8 Effector&#39;,
	&#39;5&#39; = &#39;pre-B/pro-B&#39;,
	&#39;6&#39; = &#39;NK dim&#39;,
	&#39;7&#39; = &#39;CD16 Mono&#39;,
	&#39;8&#39; = &#39;DC&#39;,
	&#39;9&#39; = &#39;pDC&#39;)
</pre></div>
</div>
</div>
<div class="section" id="differentially-accessible-regions-again">
<h3>Differentially accessible regions, again<a class="headerlink" href="#differentially-accessible-regions-again" title="Permalink to this headline">¶</a></h3>
<p>Now, we will once again look at differentially accessible region, this time between CD4/8 Naive cells and CD14 Mono cells. (Since we now can say what cell types these clusters represent.) We can plot the signal in this region in a couple of different ways. <em>Do you understand what these plots mean?</em></p>
<div class="highlight-{r seurat_diff_peaks} notranslate"><div class="highlight"><pre><span></span># Change back to working with peaks instead of gene activities
DefaultAssay(pbmc) &lt;- &#39;peaks&#39;

# Find differentially acessible regions
da_peaks &lt;- FindMarkers(
	object = pbmc,
	ident.1 = &quot;CD4/8 Naive&quot;,
	ident.2 = &quot;CD14 Mono&quot;,
	min.pct = 0.2,
	test.use = &#39;LR&#39;,
	latent.vars = &#39;peak_region_fragments&#39;
)
head(da_peaks)

# Plot the signal for the most differentially acessable region.
plot1 &lt;- VlnPlot(
	object = pbmc,
	features = rownames(da_peaks)[1],
	pt.size = 0.1,
	idents = c(&quot;CD4/8 Naive&quot;,&quot;CD14 Mono&quot;)
)

plot2 &lt;- FeaturePlot(
	object = pbmc,
	features = rownames(da_peaks)[1],
	pt.size = 0.1
)

plot1 | plot2
</pre></div>
</div>
</div>
<div class="section" id="annotating-regions">
<h3>Annotating regions<a class="headerlink" href="#annotating-regions" title="Permalink to this headline">¶</a></h3>
<p>Genomic coordinates are difficult to interpret on their own, and often it is interesting to know which genes are near, or at, the genomic regions. Below we find the nearest gene, and the distance to it, for each region. This makes it easy to check for regions near genes of interest, and to analyze genes further, e.g though a Gene Ontology analysis.</p>
<div class="highlight-{r seurat_peak_annot} notranslate"><div class="highlight"><pre><span></span>open_cd4naive &lt;- rownames(da_peaks[da_peaks$avg_logFC &gt; 0.5, ])
open_cd14mono &lt;- rownames(da_peaks[da_peaks$avg_logFC &lt; -0.5, ])

closest_genes_cd4naive &lt;- ClosestFeature(pbmc, regions = open_cd4naive)
closest_genes_cd14mono &lt;- ClosestFeature(pbmc, regions = open_cd14mono)
head(closest_genes_cd4naive)
head(closest_genes_cd14mono)
</pre></div>
</div>
</div>
<div class="section" id="visualization">
<h3>Visualization<a class="headerlink" href="#visualization" title="Permalink to this headline">¶</a></h3>
<p>It is often informative to plot the ATAC-seq signal for a particular genomic region, and group this by cluster, cell type or some other meta data. We can either plot the the aggregate signal for a group with <code class="docutils literal notranslate"><span class="pre">coveragePlot</span></code> or the signal for individual cells, with <code class="docutils literal notranslate"><span class="pre">tilePlot</span></code>. More info about plotting genomic region can be found <a class="reference external" href="https://satijalab.org/signac/articles/visualization.html">here</a></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CoveragePlot</span><span class="p">(</span>
	<span class="nb">object</span> <span class="o">=</span> <span class="n">pbmc</span><span class="p">,</span>
	<span class="n">region</span> <span class="o">=</span> <span class="n">rownames</span><span class="p">(</span><span class="n">da_peaks</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
	<span class="n">extend</span><span class="o">.</span><span class="n">upstream</span> <span class="o">=</span> <span class="mi">40000</span><span class="p">,</span>
	<span class="n">extend</span><span class="o">.</span><span class="n">downstream</span> <span class="o">=</span> <span class="mi">40000</span><span class="p">,</span>
	<span class="n">peaks</span><span class="o">=</span><span class="n">TRUE</span>
<span class="p">)</span>

<span class="n">TilePlot</span><span class="p">(</span>
	<span class="nb">object</span> <span class="o">=</span> <span class="n">pbmc</span><span class="p">,</span>
	<span class="n">region</span> <span class="o">=</span> <span class="n">rownames</span><span class="p">(</span><span class="n">da_peaks</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
	<span class="n">idents</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;CD14 Mono&#39;</span><span class="p">,</span> <span class="s1">&#39;CD4/8 Naive&#39;</span><span class="p">),</span>
	<span class="n">extend</span><span class="o">.</span><span class="n">upstream</span> <span class="o">=</span> <span class="mi">40000</span><span class="p">,</span>
	<span class="n">extend</span><span class="o">.</span><span class="n">downstream</span> <span class="o">=</span> <span class="mi">40000</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="../genomeOverlap/lab-genomicGoverlaps.html" class="btn btn-neutral float-right" title="Genomic overlaps" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../data_integration_tutorials.html" class="btn btn-neutral float-left" title="Data integration" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, NBIS Epigenomics Workshop Team. All rights reserved..

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>