

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Detection of differential binding sites using csaw &mdash; Epigenomics Workshop 2020 1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="&lt;no title&gt;" href="../diffBind/lab-diffBinding-local.html" />
    <link rel="prev" title="ChIP-seq down-stream analysis: ChIPseeker" href="../chipseeker/lab-ChIPseeker.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home" alt="Documentation Home"> Epigenomics Workshop 2020
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../precourse.html">Pre-course Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../schedule.html">Schedule</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../setup/lab-setup.html">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="../methylation_tutorials.html">DNA Methylation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chipseq_tutorials.html">ChIPseq</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../chip_downstream_tutorials.html">Downstream Processing</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../chipseeker/lab-ChIPseeker.html">Annotation using ChIPseeker</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Detection of differential binding sites using csaw</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#requirements">Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#getting-the-data">Getting the data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#loading-data-and-preparing-the-contrast-to-test">Loading data and preparing the contrast to test</a></li>
<li class="toctree-l4"><a class="reference internal" href="#filtering-out-regions-with-very-low-coverage">Filtering out regions with very low coverage</a></li>
<li class="toctree-l4"><a class="reference internal" href="#normalisation">Normalisation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#detecting-differentially-binding-db-sites">Detecting differentially binding (DB) sites</a></li>
<li class="toctree-l4"><a class="reference internal" href="#correcting-for-multiple-testing">Correcting for multiple testing</a></li>
<li class="toctree-l4"><a class="reference internal" href="#inspecting-the-results">Inspecting the results</a></li>
<li class="toctree-l4"><a class="reference internal" href="#annotation-of-the-results">Annotation of the results</a></li>
<li class="toctree-l4"><a class="reference internal" href="#creating-the-final-object-with-results-and-annotation">Creating the final object with results and annotation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../motifs/lab-motifs.html">Identification of sequence motifs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../expdesign_data.html">Experimental Design and Data Considerations</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Epigenomics Workshop 2020</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../tutorials.html">Tutorials</a> &raquo;</li>
        
          <li><a href="../chip_downstream_tutorials.html">Downstream Analyses</a> &raquo;</li>
        
      <li>Detection of differential binding sites using csaw</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../../_sources/content/tutorials/csaw/lab-csaw.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="detection-of-differential-binding-sites-using-csaw">
<h1>Detection of differential binding sites using csaw<a class="headerlink" href="#detection-of-differential-binding-sites-using-csaw" title="Permalink to this headline">¶</a></h1>
<p>This is an alternative workflow for detection of differential binding / occupancy in ChIP-seq data. In contrast to working with reads counted within peaks detected in a peak calling step (as in the earlier example with DiffBind), this approach uses a <strong>sliding window</strong> to count reads across the genome. Each window is then tested for significant differences between libraries from different conditions, using the methods in the <code class="docutils literal notranslate"><span class="pre">edgeR</span></code> package. This package also offers an FDR control strategy more appropriate for ChIP-seq experiments than simple BH adjustment.</p>
<p>It can be used for point-source binding (TFs) as well as for broad signal (histones). However, it can only be used for cases where <strong>global occupancy levels are unchanged</strong>.</p>
<p>As this method is agnostic to signal structure, it requires careful choice of strategies for filtering and normalisation. Here, we show a very simple workflow. More details can be found in the <a class="reference external" href="http://bioconductor.org/packages/devel/bioc/vignettes/csaw/inst/doc/csawUserGuide.pdf">Csaw User Guide</a>.</p>
<div class="section" id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">R</span> <span class="pre">version</span> <span class="pre">&gt;</span> <span class="pre">3.4.2</span></code> (2017-09-28)</p></li>
<li><p><a class="reference external" href="https://cran.r-project.org/web/packages/statmod/index.html">statmod</a>, required for <code class="docutils literal notranslate"><span class="pre">csaw</span></code></p></li>
<li><p><a class="reference external" href="https://gcc.gnu.org/wiki/GFortranBinaries">gfortran</a>, required for <code class="docutils literal notranslate"><span class="pre">csaw</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">csaw</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">edgeR</span></code></p></li>
</ul>
<p>R packages required for annotation:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">org.Hs.eg.db</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">TxDb.Hsapiens.UCSC.hg19.knownGene</span></code></p></li>
</ul>
<p>Recommended:
* R-Studio to work in</p>
<p><em>Note: this exercise is to be run locally. We have not tested it on Uppmax.</em></p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>To install R packages not in Bioconductor use <code class="docutils literal notranslate"><span class="pre">install.packages</span></code> command e.g.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;https://cran.r-project.org/src/contrib/statmod_1.4.30.tar.gz&quot;</span><span class="p">,</span> <span class="n">repo</span><span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> <span class="n">type</span><span class="o">=</span><span class="s">&quot;source&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>To install Bioconductor packages:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">requireNamespace</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
    <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;csaw&quot;</span><span class="p">,</span><span class="s">&quot;edgeR&quot;</span><span class="p">,</span><span class="s">&quot;org.Hs.eg.db&quot;</span><span class="p">,</span><span class="s">&quot;TxDb.Hsapiens.UCSC.hg19.knownGene&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>To install <code class="docutils literal notranslate"><span class="pre">gfortran</span></code> follow the directions on <a class="reference external" href="https://gcc.gnu.org/wiki/GFortranBinaries">gfortran homepage</a>. Further dependencies may be required for successful installation.</p>
</div>
</div>
<div class="section" id="getting-the-data">
<h2>Getting the data<a class="headerlink" href="#getting-the-data" title="Permalink to this headline">¶</a></h2>
<p>We will examine differences in REST binding in two cell types: SKNSH and HeLa. We need to download required files. Let’s use the Box links for simplicity.</p>
<p>HeLa:</p>
<ul class="simple">
<li><p>[zip](<a class="reference external" href="https://stockholmuniversity.box.com/s/2o3lchp61kzxpil1y1snn4onjk4e0sjo">https://stockholmuniversity.box.com/s/2o3lchp61kzxpil1y1snn4onjk4e0sjo</a>)</p></li>
<li><p>[tar.gz](<a class="reference external" href="https://stockholmuniversity.box.com/s/wmx4uhgo3esuessr4f8g9kmvarm42bxz">https://stockholmuniversity.box.com/s/wmx4uhgo3esuessr4f8g9kmvarm42bxz</a>)</p></li>
</ul>
<p>SKNSH:</p>
<ul class="simple">
<li><p>[zip](<a class="reference external" href="https://stockholmuniversity.box.com/s/dkurmi5suwh3qnxnx0ysfhh5c0g2d1ti">https://stockholmuniversity.box.com/s/dkurmi5suwh3qnxnx0ysfhh5c0g2d1ti</a>)</p></li>
<li><p>[tar.gz](<a class="reference external" href="https://stockholmuniversity.box.com/s/8m3rgtakx8h8rmhnwltitqccbx7h0wy3">https://stockholmuniversity.box.com/s/8m3rgtakx8h8rmhnwltitqccbx7h0wy3</a>)</p></li>
</ul>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>You can also <cite>scp</cite> the files from rackham at <cite>##</cite></p>
</div>
<p>To extract <code class="docutils literal notranslate"><span class="pre">tar.gz</span></code> files</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tar -zxvf archive_name.tar.gz
</pre></div>
</div>
</div>
<div class="section" id="loading-data-and-preparing-the-contrast-to-test">
<h2>Loading data and preparing the contrast to test<a class="headerlink" href="#loading-data-and-preparing-the-contrast-to-test" title="Permalink to this headline">¶</a></h2>
<p>You’ll be working in <code class="docutils literal notranslate"><span class="pre">R</span></code> for the remaining part of the exercise. Modify the paths to folders with respective data to match your local setup:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">dir.sknsh</span> <span class="o">=</span> <span class="s">&quot;/path/to/data/sknsh&quot;</span>
<span class="n">dir.hela</span> <span class="o">=</span> <span class="s">&quot;/path/to/data/hela&quot;</span>

<span class="n">hela.1</span><span class="o">=</span><span class="nf">file.path</span><span class="p">(</span><span class="n">dir.hela</span><span class="p">,</span><span class="s">&quot;ENCFF000PED.chr12.rmdup.sort.bam&quot;</span><span class="p">)</span>
<span class="n">hela.2</span><span class="o">=</span><span class="nf">file.path</span><span class="p">(</span><span class="n">dir.hela</span><span class="p">,</span><span class="s">&quot;ENCFF000PEE.chr12.rmdup.sort.bam&quot;</span><span class="p">)</span>
<span class="n">sknsh.1</span><span class="o">=</span><span class="nf">file.path</span><span class="p">(</span><span class="n">dir.sknsh</span><span class="p">,</span><span class="s">&quot;ENCFF000RAG.chr12.rmdup.sort.bam&quot;</span><span class="p">)</span>
<span class="n">sknsh.2</span><span class="o">=</span><span class="nf">file.path</span><span class="p">(</span><span class="n">dir.sknsh</span><span class="p">,</span><span class="s">&quot;ENCFF000RAH.chr12.rmdup.sort.bam&quot;</span><span class="p">)</span>

<span class="n">bam.files</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="n">hela.1</span><span class="p">,</span><span class="n">hela.2</span><span class="p">,</span><span class="n">sknsh.1</span><span class="p">,</span><span class="n">sknsh.2</span><span class="p">)</span>
</pre></div>
</div>
<p>We need to provide the information about the design of the experiment using <code class="docutils literal notranslate"><span class="pre">model.matrix</span></code> function:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">grouping</span> <span class="o">&lt;-</span> <span class="nf">factor</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;hela&#39;</span><span class="p">,</span> <span class="s">&#39;hela&#39;</span><span class="p">,</span> <span class="s">&#39;sknsh&#39;</span><span class="p">,</span> <span class="s">&#39;sknsh&#39;</span><span class="p">))</span>
<span class="n">design</span> <span class="o">&lt;-</span> <span class="nf">model.matrix</span><span class="p">(</span><span class="o">~</span><span class="m">0</span> <span class="o">+</span> <span class="n">grouping</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">design</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">levels</span><span class="p">(</span><span class="n">grouping</span><span class="p">)</span>
</pre></div>
</div>
<p>The design should look like this:
.. code-block:: R</p>
<blockquote>
<div><dl class="simple">
<dt>&gt; design</dt><dd><p>hela sknsh</p>
</dd>
</dl>
<p>1    1     0
2    1     0
3    0     1
4    0     1
attr(,”assign”)
[1] 1 1
attr(,”contrasts”)
attr(,”contrasts”)$grouping
[1] “contr.treatment”</p>
</div></blockquote>
<p>We prepare the information on contrast to be tested using <code class="docutils literal notranslate"><span class="pre">makeContrasts</span></code> function from package <code class="docutils literal notranslate"><span class="pre">limma</span></code>. This is not the only way to do so, and examples are given in <code class="docutils literal notranslate"><span class="pre">csaw</span></code> and <code class="docutils literal notranslate"><span class="pre">edgeR</span></code> manuals. In this case we want to test for the differences in REST binding in HeLa vs. SKNSH cell lines:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">edgeR</span><span class="p">)</span>
<span class="n">contrast</span> <span class="o">&lt;-</span> <span class="nf">makeContrasts</span><span class="p">(</span><span class="n">hela</span> <span class="o">-</span> <span class="n">sknsh</span><span class="p">,</span> <span class="n">levels</span><span class="o">=</span><span class="n">design</span><span class="p">)</span>
</pre></div>
</div>
<p>Now we are ready to load data and create an object with counted reads:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">csaw</span><span class="p">)</span>
<span class="n">data</span> <span class="o">&lt;-</span> <span class="nf">windowCounts</span><span class="p">(</span><span class="n">bam.files</span><span class="p">,</span> <span class="n">ext</span><span class="o">=</span><span class="m">100</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="m">10</span><span class="p">)</span>
</pre></div>
</div>
<p>Parameters for file loading can be modified (examples in the <code class="docutils literal notranslate"><span class="pre">csaw</span></code> User Guide), depending on how the data was processed. Here we explicitely input the value for fragment length as we have this information from the cross correlation analysis performed earlier (####ChIP-seq data processing tutorial####). It is 100 for Hela and 95 &amp; 115 for sknsh.</p>
<p>We can inspect the resulting <code class="docutils literal notranslate"><span class="pre">data</span></code> object, e.g.:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span><span class="o">$</span><span class="n">totals</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">1637778</span> <span class="m">2009932</span> <span class="m">2714033</span> <span class="m">4180463</span>
</pre></div>
</div>
</div>
<div class="section" id="filtering-out-regions-with-very-low-coverage">
<h2>Filtering out regions with very low coverage<a class="headerlink" href="#filtering-out-regions-with-very-low-coverage" title="Permalink to this headline">¶</a></h2>
<p>The next step is to filter out uninformative regions, i.e. windows with low read count, which represent background. There are many strategies to do it, depending on the biology of the experiment, IP efficiency and data processing. Here, we filter out lowest 99.9% of the windows, retaining the 0.1% windows with highest signal. The rationale is that for TF experiments only 0.1% of the genome can be bound, hence the remaining must represent background.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">keep</span> <span class="o">&lt;-</span> <span class="nf">filterWindows</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">type</span><span class="o">=</span><span class="s">&quot;proportion&quot;</span><span class="p">)</span><span class="o">$</span><span class="n">filter</span> <span class="o">&gt;</span> <span class="m">0.999</span>
<span class="n">data.filt</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="p">[</span><span class="n">keep</span><span class="p">,]</span>
</pre></div>
</div>
<p>To investigate the effectiveness of our filtering strategy:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="nf">summary</span><span class="p">(</span><span class="n">keep</span><span class="p">)</span>
   <span class="n">Mode</span>   <span class="kc">FALSE</span>    <span class="kc">TRUE</span>
<span class="n">logical</span>  <span class="m">145558</span>    <span class="m">9850</span>
</pre></div>
</div>
</div>
<div class="section" id="normalisation">
<h2>Normalisation<a class="headerlink" href="#normalisation" title="Permalink to this headline">¶</a></h2>
<p>Assigning reads into larger bins for normalisation:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">binned</span> <span class="o">&lt;-</span> <span class="nf">windowCounts</span><span class="p">(</span><span class="n">bam.files</span><span class="p">,</span> <span class="n">bin</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="m">10000</span><span class="p">)</span>
</pre></div>
</div>
<p>Calculating normalization factors:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">data.filt</span> <span class="o">&lt;-</span> <span class="nf">normOffsets</span><span class="p">(</span><span class="n">binned</span><span class="p">,</span> <span class="n">se.out</span><span class="o">=</span><span class="n">data.filt</span><span class="p">)</span>
</pre></div>
</div>
<p>Inspecting the normalisation factors:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data.filt</span><span class="o">$</span><span class="n">norm.factors</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">0.9727458</span> <span class="m">1.0718693</span> <span class="m">0.9279702</span> <span class="m">1.0335341</span>
</pre></div>
</div>
</div>
<div class="section" id="detecting-differentially-binding-db-sites">
<h2>Detecting differentially binding (DB) sites<a class="headerlink" href="#detecting-differentially-binding-db-sites" title="Permalink to this headline">¶</a></h2>
<p>Detecting DB windows:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">data.filt.calc</span> <span class="o">&lt;-</span> <span class="nf">asDGEList</span><span class="p">(</span><span class="n">data.filt</span><span class="p">)</span>
<span class="n">data.filt.calc</span> <span class="o">&lt;-</span> <span class="nf">estimateDisp</span><span class="p">(</span><span class="n">data.filt.calc</span><span class="p">,</span> <span class="n">design</span><span class="p">)</span>
<span class="n">fit</span> <span class="o">&lt;-</span> <span class="nf">glmQLFit</span><span class="p">(</span><span class="n">data.filt.calc</span><span class="p">,</span> <span class="n">design</span><span class="p">,</span> <span class="n">robust</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
<span class="n">results</span> <span class="o">&lt;-</span> <span class="nf">glmQLFTest</span><span class="p">(</span><span class="n">fit</span><span class="p">,</span> <span class="n">contrast</span><span class="o">=</span><span class="n">contrast</span><span class="p">)</span>
</pre></div>
</div>
<p>Inspecting the results table:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="nf">head</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">table</span><span class="p">)</span>
     <span class="n">logFC</span>   <span class="n">logCPM</span>         <span class="bp">F</span>       <span class="n">PValue</span>
<span class="m">1</span> <span class="m">7.239404</span> <span class="m">2.165639</span> <span class="m">17.229173</span> <span class="m">3.327018e-05</span>
<span class="m">2</span> <span class="m">5.244217</span> <span class="m">2.783211</span>  <span class="m">9.484909</span> <span class="m">2.074540e-03</span>
<span class="m">3</span> <span class="m">3.023888</span> <span class="m">2.755437</span>  <span class="m">4.721852</span> <span class="m">2.979352e-02</span>
<span class="m">4</span> <span class="m">2.050617</span> <span class="m">2.612401</span>  <span class="m">2.684560</span> <span class="m">1.013412e-01</span>
<span class="m">5</span> <span class="m">1.827703</span> <span class="m">2.459979</span>  <span class="m">2.459072</span> <span class="m">1.168638e-01</span>
<span class="m">6</span> <span class="m">4.336717</span> <span class="m">2.052296</span> <span class="m">14.330442</span> <span class="m">1.538194e-04</span>
</pre></div>
</div>
</div>
<div class="section" id="correcting-for-multiple-testing">
<h2>Correcting for multiple testing<a class="headerlink" href="#correcting-for-multiple-testing" title="Permalink to this headline">¶</a></h2>
<p>First we merge adjacent DB windows into longer clusters. Windows that are less than <code class="docutils literal notranslate"><span class="pre">tol</span></code> apart are considered to be adjacent and are grouped into the same cluster. The chosen <code class="docutils literal notranslate"><span class="pre">tol</span></code>
represents the minimum distance at which two binding events are treated as separate sites.
Large values (500 - 1000 bp) reduce redundancy and favor a region-based interpretation of
the results, while smaller values (&lt; 200 bp) allow resolution of individual binding sites.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">merged</span> <span class="o">&lt;-</span> <span class="nf">mergeWindows</span><span class="p">(</span><span class="nf">rowRanges</span><span class="p">(</span><span class="n">data.filt</span><span class="p">),</span> <span class="n">tol</span><span class="o">=</span><span class="m">1000L</span><span class="p">)</span>
</pre></div>
</div>
<p>Next, we apply the multiple testing correction to obtain FDR. We combine p-values across clustered tests using Simes??? method to control the cluster FDR.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">table.combined</span> <span class="o">&lt;-</span> <span class="nf">combineTests</span><span class="p">(</span><span class="n">merged</span><span class="o">$</span><span class="n">id</span><span class="p">,</span> <span class="n">results</span><span class="o">$</span><span class="n">table</span><span class="p">)</span>
</pre></div>
</div>
<p>The resulting <code class="docutils literal notranslate"><span class="pre">table.combined</span></code> object contains FDR for each cluster:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="nf">head</span><span class="p">(</span><span class="n">table.combined</span><span class="p">)</span>
  <span class="n">nWindows</span> <span class="n">logFC.up</span> <span class="n">logFC.down</span>       <span class="n">PValue</span>          <span class="n">FDR</span>
<span class="m">1</span>        <span class="m">7</span>        <span class="m">0</span>          <span class="m">7</span> <span class="m">2.328912e-04</span> <span class="m">0.0040397108</span>
<span class="m">2</span>        <span class="m">3</span>        <span class="m">0</span>          <span class="m">3</span> <span class="m">6.989334e-06</span> <span class="m">0.0004822892</span>
<span class="m">3</span>        <span class="m">3</span>        <span class="m">0</span>          <span class="m">3</span> <span class="m">1.948039e-04</span> <span class="m">0.0036799249</span>
<span class="m">4</span>        <span class="m">5</span>        <span class="m">0</span>          <span class="m">5</span> <span class="m">4.108169e-05</span> <span class="m">0.0011680754</span>
<span class="m">5</span>        <span class="m">3</span>        <span class="m">0</span>          <span class="m">3</span> <span class="m">6.674578e-05</span> <span class="m">0.0017204192</span>
<span class="m">6</span>        <span class="m">5</span>        <span class="m">0</span>          <span class="m">5</span> <span class="m">1.880546e-04</span> <span class="m">0.0036207355</span>
  <span class="n">direction</span>
<span class="m">1</span>      <span class="n">down</span>
<span class="m">2</span>      <span class="n">down</span>
<span class="m">3</span>      <span class="n">down</span>
<span class="m">4</span>      <span class="n">down</span>
<span class="m">5</span>      <span class="n">down</span>
<span class="m">6</span>      <span class="n">down</span>
</pre></div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">nWindows</span></code> - the total number of windows in each cluster;</p></li>
<li><p>fields <code class="docutils literal notranslate"><span class="pre">*.up</span></code> and <code class="docutils literal notranslate"><span class="pre">*.down</span></code> - for each log-FC column in <code class="docutils literal notranslate"><span class="pre">results$table</span></code>; contain the number of</p></li>
</ul>
<p>windows with log-FCs above 0.5 or below -0.5, respectively;
* <code class="docutils literal notranslate"><span class="pre">PValue</span></code> - the combined p value;
* <code class="docutils literal notranslate"><span class="pre">FDR</span></code> - the q-value corresponding to the combined p value;
* <code class="docutils literal notranslate"><span class="pre">direction</span></code> - the dominant direction of change for windows in each cluster.</p>
<p>Each combined p value represents evidence against the global null hypothesis,
i.e., all individual nulls are true in each cluster. This may be more relevant than examining each
test individually when multiple tests in a cluster represent parts of the same underlying event, i.e.,
genomic regions consisting of clusters of windows. The BH method is then applied to control the
FDR across all clusters.</p>
</div>
<div class="section" id="inspecting-the-results">
<h2>Inspecting the results<a class="headerlink" href="#inspecting-the-results" title="Permalink to this headline">¶</a></h2>
<p>We select statistically significant DB events at FDR 0.05:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">is.sig.region</span> <span class="o">&lt;-</span> <span class="n">table.combined</span><span class="o">$</span><span class="n">FDR</span> <span class="o">&lt;=</span> <span class="m">0.05</span>
<span class="nf">table</span><span class="p">(</span><span class="n">table.combined</span><span class="o">$</span><span class="n">direction</span><span class="p">[</span><span class="n">is.sig.region</span><span class="p">])</span>
</pre></div>
</div>
<p>How many regions were detected as differentialy bound?</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">down</span>   <span class="n">up</span>
<span class="m">201</span>  <span class="m">231</span>
</pre></div>
</div>
<p>out of</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="nf">length</span><span class="p">(</span><span class="n">table.combined</span><span class="o">$</span><span class="n">FDR</span><span class="p">)</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">2758</span>
</pre></div>
</div>
<p>We can also obtain information on the best window in each cluster:</p>
<p>We can inspect congruency of the replicates on MDS. We subsample counts for faster calculations:</p>
</div>
<div class="section" id="annotation-of-the-results">
<h2>Annotation of the results<a class="headerlink" href="#annotation-of-the-results" title="Permalink to this headline">¶</a></h2>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="p">)</span>

<span class="n">anno</span> <span class="o">&lt;-</span> <span class="nf">detailRanges</span><span class="p">(</span><span class="n">merged</span><span class="o">$</span><span class="n">region</span><span class="p">,</span> <span class="n">txdb</span><span class="o">=</span><span class="n">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="p">,</span>
<span class="n">orgdb</span><span class="o">=</span><span class="n">org.Hs.eg.db</span><span class="p">,</span> <span class="n">promoter</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">3000</span><span class="p">,</span> <span class="m">1000</span><span class="p">),</span> <span class="n">dist</span><span class="o">=</span><span class="m">5000</span><span class="p">)</span>

<span class="n">merged</span><span class="o">$</span><span class="n">region</span><span class="o">$</span><span class="n">overlap</span> <span class="o">&lt;-</span> <span class="n">anno</span><span class="o">$</span><span class="n">overlap</span>
<span class="n">merged</span><span class="o">$</span><span class="n">region</span><span class="o">$</span><span class="n">left</span> <span class="o">&lt;-</span> <span class="n">anno</span><span class="o">$</span><span class="n">left</span>
<span class="n">merged</span><span class="o">$</span><span class="n">region</span><span class="o">$</span><span class="n">right</span> <span class="o">&lt;-</span> <span class="n">anno</span><span class="o">$</span><span class="n">right</span>
</pre></div>
</div>
</div>
<div class="section" id="creating-the-final-object-with-results-and-annotation">
<h2>Creating the final object with results and annotation<a class="headerlink" href="#creating-the-final-object-with-results-and-annotation" title="Permalink to this headline">¶</a></h2>
<p>Now we bring it all together:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">all.results</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="n">merged</span><span class="o">$</span><span class="n">region</span><span class="p">)[,</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">],</span> <span class="n">table.combined</span><span class="p">,</span> <span class="n">anno</span><span class="p">)</span>
</pre></div>
</div>
<p>All significant regions are in:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">sig</span><span class="o">=</span><span class="n">all.results</span><span class="p">[</span><span class="n">all.results</span><span class="o">$</span><span class="n">FDR</span><span class="o">&lt;</span><span class="m">0.05</span><span class="p">,]</span>
</pre></div>
</div>
<p>To view the top of the <code class="docutils literal notranslate"><span class="pre">all.results</span></code> table:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">all.results</span> <span class="o">&lt;-</span> <span class="n">all.results</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">all.results</span><span class="o">$</span><span class="n">PValue</span><span class="p">),]</span>

<span class="o">&gt;</span> <span class="nf">head</span><span class="p">(</span><span class="n">all.results</span><span class="p">)</span>
     <span class="n">seqnames</span>     <span class="n">start</span>       <span class="n">end</span> <span class="n">nWindows</span> <span class="n">logFC.up</span>
<span class="m">1726</span>     <span class="n">chr2</span>  <span class="m">25642751</span>  <span class="m">25642760</span>        <span class="m">1</span>        <span class="m">1</span>
<span class="m">822</span>      <span class="n">chr1</span> <span class="m">143647051</span> <span class="m">143647060</span>        <span class="m">1</span>        <span class="m">0</span>
<span class="m">876</span>      <span class="n">chr1</span> <span class="m">149785201</span> <span class="m">149785210</span>        <span class="m">1</span>        <span class="m">0</span>
<span class="m">386</span>      <span class="n">chr1</span>  <span class="m">40530701</span>  <span class="m">40530710</span>        <span class="m">1</span>        <span class="m">1</span>
<span class="m">2519</span>     <span class="n">chr2</span> <span class="m">199778551</span> <span class="m">199778560</span>        <span class="m">1</span>        <span class="m">1</span>
<span class="m">1613</span>     <span class="n">chr2</span>   <span class="m">8683951</span>   <span class="m">8683960</span>        <span class="m">1</span>        <span class="m">0</span>
     <span class="n">logFC.down</span>       <span class="n">PValue</span>          <span class="n">FDR</span> <span class="n">direction</span>
<span class="m">1726</span>          <span class="m">0</span> <span class="m">7.875683e-07</span> <span class="m">0.0004407602</span>        <span class="n">up</span>
<span class="m">822</span>           <span class="m">1</span> <span class="m">1.197351e-06</span> <span class="m">0.0004407602</span>      <span class="n">down</span>
<span class="m">876</span>           <span class="m">1</span> <span class="m">1.197351e-06</span> <span class="m">0.0004407602</span>      <span class="n">down</span>
<span class="m">386</span>           <span class="m">0</span> <span class="m">1.574877e-06</span> <span class="m">0.0004407602</span>        <span class="n">up</span>
<span class="m">2519</span>          <span class="m">0</span> <span class="m">1.574877e-06</span> <span class="m">0.0004407602</span>        <span class="n">up</span>
<span class="m">1613</span>          <span class="m">1</span> <span class="m">2.198223e-06</span> <span class="m">0.0004407602</span>      <span class="n">down</span>
                          <span class="n">overlap</span>                  <span class="n">left</span>
<span class="m">1726</span>                     <span class="n">DTNB</span><span class="o">|</span><span class="n">I</span><span class="o">|-</span>     <span class="n">DTNB</span><span class="o">|</span><span class="m">18-19</span><span class="o">|-</span><span class="p">[</span><span class="m">347</span><span class="p">]</span>
<span class="m">822</span>
<span class="m">876</span>  <span class="n">HIST2H2BF</span><span class="o">|</span><span class="m">0</span><span class="o">|-</span><span class="p">,</span><span class="n">HIST2H3D</span><span class="o">|</span><span class="m">0-1</span><span class="o">|-</span> <span class="n">HIST2H2BF</span><span class="o">|</span><span class="m">1-2</span><span class="o">|-</span><span class="p">[</span><span class="m">1273</span><span class="p">]</span>
<span class="m">386</span>                      <span class="n">CAP1</span><span class="o">|</span><span class="n">I</span><span class="o">|+</span>      <span class="n">CAP1</span><span class="o">|</span><span class="m">5-11</span><span class="o">|+</span><span class="p">[</span><span class="m">470</span><span class="p">]</span>
<span class="m">2519</span>
<span class="m">1613</span>
                     <span class="n">right</span>
<span class="m">1726</span>
<span class="m">822</span>  <span class="o">&lt;</span><span class="m">100286793</span><span class="o">&gt;|</span><span class="m">10</span><span class="o">|-</span><span class="p">[</span><span class="m">579</span><span class="p">]</span>
<span class="m">876</span>
<span class="m">386</span>     <span class="n">CAP1</span><span class="o">|</span><span class="m">12-14</span><span class="o">|+</span><span class="p">[</span><span class="m">1177</span><span class="p">]</span>
<span class="m">2519</span>
<span class="m">1613</span>
</pre></div>
</div>
<p>We of course discourage ranking the results by p value ;-).</p>
<p>Now you are ready to save the results as a table, inspect further and generate a compelling scientific hypothesis.
You can also compare the outcome with results obtained from peak-based couting approach.</p>
<p>One final note: In this example we have used preprocessed bam files, i.e. reads mapped to the regions of spurious high signal in ChIP-seq (i.e. the ENCODE “blacklisted regions”) were removed, as were the so called <strong>duplicated reads</strong> - reads mapped to the same genomic positions. While filtering out the blacklisted regions is always recommended, <strong>removal of duplicated reads is not recommended</strong> for DB analysis, as they may represent true signal. As always, your mileage may vary, depending on the project, so exploring several options is essential for obtaining meaningful results.</p>
<dl class="footnote brackets">
<dt class="label" id="id1"><span class="brackets">1</span></dt>
<dd><p>en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</p>
</dd>
</dl>
<dl class="footnote brackets">
<dt class="label" id="id2"><span class="brackets">1</span></dt>
<dd><p>parallel  stats4    stats     graphics  grDevices utils     datasets</p>
</dd>
<dt class="label" id="id3"><span class="brackets">8</span></dt>
<dd><p>methods   base</p>
</dd>
</dl>
<dl class="footnote brackets">
<dt class="label" id="id4"><span class="brackets">1</span></dt>
<dd><p>TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2</p>
</dd>
<dt class="label" id="id5"><span class="brackets">2</span></dt>
<dd><p>GenomicFeatures_1.30.0</p>
</dd>
<dt class="label" id="id6"><span class="brackets">3</span></dt>
<dd><p>org.Hs.eg.db_3.5.0</p>
</dd>
<dt class="label" id="id7"><span class="brackets">4</span></dt>
<dd><p>AnnotationDbi_1.40.0</p>
</dd>
<dt class="label" id="id8"><span class="brackets">5</span></dt>
<dd><p>csaw_1.12.0</p>
</dd>
<dt class="label" id="id9"><span class="brackets">6</span></dt>
<dd><p>BiocParallel_1.12.0</p>
</dd>
<dt class="label" id="id10"><span class="brackets">7</span></dt>
<dd><p>SummarizedExperiment_1.8.0</p>
</dd>
<dt class="label" id="id11"><span class="brackets">8</span></dt>
<dd><p>DelayedArray_0.4.1</p>
</dd>
<dt class="label" id="id12"><span class="brackets">9</span></dt>
<dd><p>matrixStats_0.52.2</p>
</dd>
<dt class="label" id="id13"><span class="brackets">10</span></dt>
<dd><p>Biobase_2.38.0</p>
</dd>
<dt class="label" id="id14"><span class="brackets">11</span></dt>
<dd><p>GenomicRanges_1.30.0</p>
</dd>
<dt class="label" id="id15"><span class="brackets">12</span></dt>
<dd><p>GenomeInfoDb_1.14.0</p>
</dd>
<dt class="label" id="id16"><span class="brackets">13</span></dt>
<dd><p>IRanges_2.12.0</p>
</dd>
<dt class="label" id="id17"><span class="brackets">14</span></dt>
<dd><p>S4Vectors_0.16.0</p>
</dd>
<dt class="label" id="id18"><span class="brackets">15</span></dt>
<dd><p>BiocGenerics_0.24.0</p>
</dd>
<dt class="label" id="id19"><span class="brackets">16</span></dt>
<dd><p>edgeR_3.20.1</p>
</dd>
<dt class="label" id="id20"><span class="brackets">17</span></dt>
<dd><p>limma_3.34.1</p>
</dd>
</dl>
<dl class="footnote brackets">
<dt class="label" id="id21"><span class="brackets">1</span></dt>
<dd><p>Rcpp_0.12.13             compiler_3.4.2</p>
</dd>
<dt class="label" id="id22"><span class="brackets">3</span></dt>
<dd><p>XVector_0.18.0           prettyunits_1.0.2</p>
</dd>
<dt class="label" id="id23"><span class="brackets">5</span></dt>
<dd><p>bitops_1.0-6             tools_3.4.2</p>
</dd>
<dt class="label" id="id24"><span class="brackets">7</span></dt>
<dd><p>zlibbioc_1.24.0          progress_1.1.2</p>
</dd>
<dt class="label" id="id25"><span class="brackets">9</span></dt>
<dd><p>statmod_1.4.30           biomaRt_2.34.0</p>
</dd>
<dt class="label" id="id26"><span class="brackets">11</span></dt>
<dd><p>digest_0.6.12            bit_1.1-12</p>
</dd>
<dt class="label" id="id27"><span class="brackets">13</span></dt>
<dd><p>RSQLite_2.0              memoise_1.1.0</p>
</dd>
<dt class="label" id="id28"><span class="brackets">15</span></dt>
<dd><p>tibble_1.3.4             lattice_0.20-35</p>
</dd>
<dt class="label" id="id29"><span class="brackets">17</span></dt>
<dd><p>pkgconfig_2.0.1          rlang_0.1.4</p>
</dd>
<dt class="label" id="id30"><span class="brackets">19</span></dt>
<dd><p>Matrix_1.2-12            DBI_0.7</p>
</dd>
<dt class="label" id="id31"><span class="brackets">21</span></dt>
<dd><p>GenomeInfoDbData_0.99.1  rtracklayer_1.38.0</p>
</dd>
<dt class="label" id="id32"><span class="brackets">23</span></dt>
<dd><p>stringr_1.2.0            Biostrings_2.46.0</p>
</dd>
<dt class="label" id="id33"><span class="brackets">25</span></dt>
<dd><p>locfit_1.5-9.1           bit64_0.9-7</p>
</dd>
<dt class="label" id="id34"><span class="brackets">27</span></dt>
<dd><p>grid_3.4.2               R6_2.2.2</p>
</dd>
<dt class="label" id="id35"><span class="brackets">29</span></dt>
<dd><p>XML_3.98-1.9             RMySQL_0.10.13</p>
</dd>
<dt class="label" id="id36"><span class="brackets">31</span></dt>
<dd><p>magrittr_1.5             Rhtslib_1.10.0</p>
</dd>
<dt class="label" id="id37"><span class="brackets">33</span></dt>
<dd><p>blob_1.1.0               splines_3.4.2</p>
</dd>
<dt class="label" id="id38"><span class="brackets">35</span></dt>
<dd><p>GenomicAlignments_1.14.1 Rsamtools_1.30.0</p>
</dd>
<dt class="label" id="id39"><span class="brackets">37</span></dt>
<dd><p>assertthat_0.2.0         stringi_1.1.6</p>
</dd>
<dt class="label" id="id40"><span class="brackets">39</span></dt>
<dd><p>RCurl_1.95-4.8</p>
</dd>
</dl>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../diffBind/lab-diffBinding-local.html" class="btn btn-neutral float-right" title="&lt;no title&gt;" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../chipseeker/lab-ChIPseeker.html" class="btn btn-neutral float-left" title="ChIP-seq down-stream analysis: ChIPseeker" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, NBIS Epigenomics Workshop Team. All rights reserved.

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>