<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>Signal visualisation with deepTools - Epigenomics 2020</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../../css/theme.css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "Signal visualisation with deepTools";
    var mkdocs_page_input_path = "tutorials/lab-vis.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js" defer></script>
  <script src="../../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> Epigenomics 2020</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../welcome/">Welcome</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../about/">About</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../precourse/">Precourse Information</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../schedule/">Schedule</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../">Tutorials</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">Epigenomics 2020</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
    
    <li>Signal visualisation with deepTools</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h3 id="learning-outcomes">Learning outcomes</h3>
<p>Using <code>deepTools</code>
- to visualise ChIP signal in relation to annotated TSS</p>
<h2 id="signal-visualisation-with-deeptools">Signal visualisation with deepTools</h2>
<p>One more thing that may come useful when analysing ChIP-seq data is visualising ChIP signal in relation to annotated transcription start sites (TSS), here on chromosomes 1 and 2. To do so we will:
* convert bedgraph to bigWig using UCSC utilities
* calculate scores per genome regions using among others the bigWig file
* plot a heatmap of scores associated with genomic regions</p>
<p>In case you have logged out Uppmax:</p>
<pre><code class="bash">
ssh -Y &lt;username&gt;@rackham.uppmax.uu.se
interactive -A g2018030 -p core -n 4 --reservation=g2018030_WED
source ~/chipseq_env.sh

</code></pre>

<p>Assuming the same files structure as in the main data processing tutorial, create a separate directory in <code>~/chipseq/analysis</code> and navigate to it. Copy the files needed for this exercise.</p>
<pre><code class="bash">cd ~/chipseq/analysis/
mkdir ~/chipseq/analysis/vis
cd ~/chipseq/analysis/vis

cp ../../hg19/chrom.sizes.hg19 chrom.sizes.hg19
cp ../bam_preproc/ENCFF000PED.chr12.cov.norm1x.bedgraph ./
</code></pre>

<p>To calculate scores per genome with deepTools <a href="http://deeptools.readthedocs.org/en/latest/content/tools/computeMatrix.html">computeMatrix</a> we need <a href="https://genome.ucsc.edu/goldenpath/help/bigWig.html">bigWig</a> file that we can obtain by converting bedgraph using UCSC utilities:</p>
<pre><code class="bash">module load ucsc-utilities/v287

bedGraphToBigWig ENCFF000PED.chr12.cov.norm1x.bedgraph chrom.sizes.hg19 hela_1.bw

module unload ucsc-utilities/v287
</code></pre>

<p>We can now compute the matrix of scores for visualisation using <a href="http://deeptools.readthedocs.org/en/latest/content/tools/computeMatrix.html">computeMatrix</a>. This tool calculates scores per genome regions and prepares an intermediate file that can be used with <code>plotHeatmap</code> and <code>plotProfiles</code>. Typically, the genome regions are genes, but any other regions defined in a BED file can be used. <code>computeMatrix</code> accepts multiple score files (bigWig format) and multiple regions files (BED format). This tool can also be used to filter and sort regions according to their score.</p>
<p>We will need a BED file with positions of TSS that we can copy to the working directory before running computeMatrix e.g.</p>
<pre><code class="bash">module load deepTools/2.5.1

cp /sw/share/compstore/courses/ngsintro/chipseq/hg19/refGene_hg19_TSS_chr12_sorted_corr.bed ./

computeMatrix reference-point -S hela_1.bw \
-R refGene_hg19_TSS_chr12_sorted_corr.bed -b 5000 -a 5000 \
--outFileName matrix.tss.dat --outFileNameMatrix matrix.tss.txt \
--referencePoint=TSS --numberOfProcessors=max
</code></pre>

<p>We can now create a heatmap for scores associated with genomic regions, i.e. plot the binding profile around TSS</p>
<pre><code class="bash">
plotHeatmap --matrixFile matrix.tss.dat \
--outFileName tss.hela_1.pdf \
--sortRegions descend --sortUsing mean

</code></pre>

<p>Have a look at the <code>tss.hela_1.pdf</code>. What do you think?</p>
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
