---
title: "Welcome"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.path="session-intro-files/figures/")
knitr::opts_chunk$set(fig.pos = 'H')
knitr::opts_chunk$set(cache.path = "tmp")
knitr::opts_chunk$set(cache = FALSE)


```


### Epigenomics data analysis: from bulk to single cell
- 23th - 27th, November, 2020
- online


Online workshop open for PhD students, postdocs, researchers and other employees within Swedish academia. This course is organised by the National Bioinformatics Infrastructure Sweden (NBIS). This workshop builds on our previous [ChIPseq data analysis workshop](https://nbisweden.github.io/workshop-archive/workshop-ChIP-seq/2018-11-07) to introduce best practice bioinformatics methods for processing, analyses and integration of epigenomics data. The online teaching includes lectures, programming tutorials and interactive group sessions.


<br/><br/>

### Getting to know each other
<br/>

##### Teaching team

- [Agata Smialowska][agata], NBIS
- [Olga Dethlefsen][olga], NBIS
- [Jakub Orzechowski Westholm][jakub], NBIS
- [Vincent van Hoef][vincent], NBIS
- [Jessica Nordlund][jessica], SNP&SEQ Technology Platform
- [Simon Els√§sser][simon], KI
- [Carmen Navarro Luzon][carmen], KI
- [Phil Ewels][phil], National Genomic Infrastructure
- [Harshil Patel][harshil], The Francis Crick Institute

[olga]: https://nbis.se/about/staff/olga-dethlefsen/
[agata]: https://nbis.se/about/staff/agata-smialowska/
[jakub]: https://nbis.se/about/staff/jakub-orzechowski-westholm
[vincent]: https://nbis.se/about/staff/vincent-van-hoef
[phil]: https://www.kth.se/profile/phile?l=en
[harshil]: https://www.researchgate.net/profile/Harshil_Patel12
[simon]: https://www.scilifelab.se/researchers/simon-elsasser/
[carmen]: https://staff.ki.se/people/carnav
[jessica]: https://katalog.uu.se/profile/?id=N8-1393

<br/><br/>

##### Email us
edu.epigenomics@nbis.se
<br/><br/>

#### What about you? 

<br/><br/>
```{r, echo=F, message=FALSE, warning=FALSE}

rm(list=ls())
library(dplyr)
library(ggplot2)
library(RColorBrewer)
library(ggpubr)
library(scales)

data.inp <- read.csv("background.csv")

#head(data.inp)

```


```{r, echo=F, message=F, warning=F, fig.align="center"}

colornames <- brewer.pal(n=4, "Set1")

data.inp$R <- factor(data.inp$R, levels = c("None", "Beginner", "Intermediate", "Advanced"))
data.inp$Unix <- factor(data.inp$Unix, levels = c("None", "Beginner", "Intermediate", "Advanced"))
data.inp$Position <- factor(data.inp$Position)
data.inp$University <- factor(data.inp$University)

bar.width <- 0.7
g1 <- ggplot(data.inp, aes(y=Position)) + geom_bar(fill=colornames[2], alpha=0.5) + ylab("") + theme_classic2() + scale_x_continuous(breaks= pretty_breaks()) + ggtitle("Position")
g2 <- ggplot(data.inp, aes(y=University)) + geom_bar(fill=colornames[3], alpha=0.5) + ylab("") + theme_classic2()  + scale_x_continuous(breaks= pretty_breaks()) + ggtitle("University")
g3 <- ggplot(data.inp, aes(y=R)) + geom_bar(fill=colornames[4], alpha=0.5) + ylab("") + ylab("")+ theme_classic2() + scale_x_continuous(breaks= pretty_breaks()) + ggtitle("R")
g4 <- ggplot(data.inp, aes(y=Unix)) + geom_bar(alpha = 0.5) + ylab("") + theme_classic2() + scale_x_continuous(breaks= pretty_breaks()) + ggtitle("Unix")


ggarrange(g1, g2, ncol=2, widths = c(1, 2))

#g3
#g4

```

<br/><br/>
```{r, echo=F, message=F, warning=F, fig.align="center"}
ggarrange(g3, g4, ncol=2)
```


<br/><br/>

### Teaching approach
- we try to focus on active learning with us talking less and you doing more
- we try to focus on fundamentals, thinking that if you get the basics right you will have no problems reading up the more advanced things
- we believe is self-organization, we will let you choose how to study ([Online Classroom](https://nbis-workshop-epigenomics.readthedocs.io/en/latest/content/online-classroom.html))
- **be brave**: help us to build a **stimulating learning environment** by asking questions in public on Zoom and Zulip. 

<br/><br/>

### Educational research project
- a side project
- not much is really known what teaching approaches really do work best during short-term courses
- one thing that is known is that we, humans, need some time to process and learn new information
- can we somehow learn better when using different methods?
- A big thank you if you've sent your informed consent form. Not too late if you haven't. Either way, the course is the same for everyone! 
- [Informed Consent Form](Informed-consent-form.pdf)

<br/><br/>

### Certificate requirements
- active participation during the week
- online presence in live streams and group sessions
- completing "Test yourself" session: opens daily ca. 15.00
- please email us your address that we can send the certificate to

<br/><br/>

### Course website:
https://nbis-workshop-epigenomics.readthedocs.io/en/latest/index.html




