

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>ChIP-seq with exogenous chromatin spike &mdash; Epigenomics Workshop 2020 1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Downstream Analyses" href="../chip_downstream_tutorials.html" />
    <link rel="prev" title="Analysis of CUT&amp;Tag H3K27me3 dataset" href="cut-and-tag-data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Epigenomics Workshop 2020
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../precourse.html">Pre-course Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../online-classroom.html">Online classroom</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../schedule.html">Schedule</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../setup/lab-setup.html">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="../methylation_tutorials.html">DNA Methylation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chipseq_tutorials.html">ChIPseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../atacseq_tutorials.html">ATACseq</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../quantitative_chip_tutorials.html">Advanced ChIP methods</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="setup-quant-chip.html">Setting up</a></li>
<li class="toctree-l3"><a class="reference internal" href="cut-and-run-data.html">CUT&amp;RUN and ChIP data for CTCF</a></li>
<li class="toctree-l3"><a class="reference internal" href="cut-and-tag-data.html">CUT&amp;Tag H3K27me3 data</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Spike-in</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#requirements">Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#data">Data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#data-preparation">Data preparation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fingerprint-plots">Fingerprint plots</a></li>
<li class="toctree-l4"><a class="reference internal" href="#disclaimer">Disclaimer</a></li>
<li class="toctree-l4"><a class="reference internal" href="#using-chipseqspike-for-chipseq-signal-scaling">Using <code class="docutils literal notranslate"><span class="pre">ChIPSeqSpike</span></code> for ChIPseq signal scaling</a></li>
<li class="toctree-l4"><a class="reference internal" href="#what-to-do-next">What to do next</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../chip_downstream_tutorials.html">Downstream Processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../expdesign_data.html">Experimental Design and Data Considerations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_integration_tutorials.html">Data Integration</a></li>
<li class="toctree-l2"><a class="reference internal" href="../nextflow.html">Nextflow &amp; nf-core</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Epigenomics Workshop 2020</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../tutorials.html">Tutorials</a> &raquo;</li>
        
          <li><a href="../quantitative_chip_tutorials.html">Advanced ChIP methods</a> &raquo;</li>
        
      <li>ChIP-seq with exogenous chromatin spike</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../../_sources/content/tutorials/quantitativeChip/lab-exospike.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="chip-seq-with-exogenous-chromatin-spike">
<h1>ChIP-seq with exogenous chromatin spike<a class="headerlink" href="#chip-seq-with-exogenous-chromatin-spike" title="Permalink to this headline">¶</a></h1>
<p>This tutorial is included from <a class="reference external" href="https://github.com/NBISweden/workshop-epigenomics/blob/master/depreciated/labs/lab-exospike.md">previous workshop</a>. Thanks!</p>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#requirements" id="id1">Requirements</a></p>
<ul>
<li><p><a class="reference internal" href="#uppmax" id="id2">Uppmax</a></p></li>
<li><p><a class="reference internal" href="#local" id="id3">Local</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#data" id="id4">Data</a></p></li>
<li><p><a class="reference internal" href="#data-preparation" id="id5">Data preparation</a></p></li>
<li><p><a class="reference internal" href="#fingerprint-plots" id="id6">Fingerprint plots</a></p></li>
<li><p><a class="reference internal" href="#disclaimer" id="id7">Disclaimer</a></p></li>
<li><p><a class="reference internal" href="#using-chipseqspike-for-chipseq-signal-scaling" id="id8">Using <code class="docutils literal notranslate"><span class="pre">ChIPSeqSpike</span></code> for ChIPseq signal scaling</a></p>
<ul>
<li><p><a class="reference internal" href="#files-and-directories" id="id9">Files and directories</a></p></li>
<li><p><a class="reference internal" href="#scaling-of-signal-to-exogenous-chromatin-spike" id="id10">Scaling of signal to exogenous chromatin spike</a></p></li>
<li><p><a class="reference internal" href="#data-visualization" id="id11">Data visualization</a></p>
<ul>
<li><p><a class="reference internal" href="#visualization-with-gene-meta-profiles" id="id12">Visualization with gene meta-profiles</a></p></li>
<li><p><a class="reference internal" href="#visualization-with-boxplots" id="id13">Visualization with Boxplots</a></p></li>
<li><p><a class="reference internal" href="#correlation-plots" id="id14">Correlation plots</a></p></li>
</ul>
</li>
</ul>
</li>
<li><p><a class="reference internal" href="#what-to-do-next" id="id15">What to do next</a></p></li>
</ul>
</div>
<div class="section" id="requirements">
<h2><a class="toc-backref" href="#id1">Requirements</a><a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<div class="section" id="uppmax">
<h3><a class="toc-backref" href="#id2">Uppmax</a><a class="headerlink" href="#uppmax" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p>R 3.5.0 (2018-04-23) or newer</p></li>
<li><p>Bioconductor packages:</p>
<blockquote>
<div><ul class="simple">
<li><p>ChIPSeqSpike</p></li>
<li><p>BSgenome.Hsapiens.UCSC.hg38</p></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="local">
<h3><a class="toc-backref" href="#id3">Local</a><a class="headerlink" href="#local" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p>Bioconductor packages:</p>
<blockquote>
<div><ul class="simple">
<li><p>ChIPSeqSpike</p></li>
<li><p>BSgenome.Hsapiens.UCSC.hg38</p></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>The lab was tested on R 4.0.0. (R run from the console, not in R Studio). The packages were installed from Bioconductor version: Release (3.12).</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">requireNamespace</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
<span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>

<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;ChIPSeqSpike&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;BSgenome.Hsapiens.UCSC.hg38&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="data">
<h2><a class="toc-backref" href="#id4">Data</a><a class="headerlink" href="#data" title="Permalink to this headline">¶</a></h2>
<p>We will use ChIP-seq of H3K79me2 from Orlando et al, 2014 (“Quantitative ChIP-Seq Normalization Reveals Global Modulation of the Epigenome”).</p>
<p>The histone H3 lysine-79 dimethyl (H3K79me2) modification is catalyzed by the DOT1L protein and is associated with the release of paused RNA Polymerase II and licensing of transcriptional elongation. This modification is typically deposited within the 5′ regions of genes.</p>
<p>The experimental design was as follows:</p>
<ul class="simple">
<li><p>Jurkat cells were treated with selective DOT1L inhibitor EPZ5676 to globally deplete H3K79me2;</p></li>
<li><p>chromatin from treated and untreated cells was mixed in different proportions (here we use two conditions: 100 % treated and 50-50% treated + untreated) to mimic a global change in abundance of H3K79me2</p></li>
<li><p>chromatin from <em>Drosophila</em> S2 cells was added in a 1:2 ratio (1 S2 cell per 2 Jurkat cells) which provided a constant “reference” amount of H3K79me2 per human cell.</p></li>
</ul>
<p>GEO accession is <code class="docutils literal notranslate"><span class="pre">GSE60104</span></code>.
ENA accession is <code class="docutils literal notranslate"><span class="pre">PRJNA257491</span></code>.</p>
<p>files in the dataset:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sample              GEO accession  SRA accession

Jurkat_K79_100%_R1  GSM1465008     SRR1536561
Jurkat_K79_100%_R2  GSM1464998     SRR1536551
Jurkat_K79_50%_R1   GSM1465006     SRR1536559
Jurkat_WCE_100%_R1  GSM1511469     SRR1584493
Jurkat_WCE_100%_R2  GSM1511474     SRR1584498
Jurkat_WCE_50%_R1   GSM1511467     SRR1584491
</pre></div>
</div>
</div>
<div class="section" id="data-preparation">
<h2><a class="toc-backref" href="#id5">Data preparation</a><a class="headerlink" href="#data-preparation" title="Permalink to this headline">¶</a></h2>
<p>All data processing steps were already performed.</p>
<p>Raw fastq reads were filtered so that low quality bases and adapters were removed. Reads were mapped to the composite reference which consisted of <strong>hg38</strong> and <strong>dm6</strong> using <code class="docutils literal notranslate"><span class="pre">bowtie</span></code>. Only reads with one best alignemnt were retained. Alignments were split by reference using <cite>samtools</cite> and were subset to <strong>chromosome 1</strong> for hg38 and <strong>chromosome 2L</strong> for dm6.</p>
<p>Quality metrics were computed for each bam split by reference genome.</p>
<p>Fixed-step bigWig files were generated as follows:</p>
<ul class="simple">
<li><p>genome coverage of data in <cite>bam</cite> files at 1 bp resolution was calculated using <cite>bedtools</cite></p></li>
<li><p>covareage was converted to fixed step <cite>wig</cite> files using <a class="reference external" href="https://gist.github.com/svigneau/8846527/bedgraph_to_wig.pl">https://gist.github.com/svigneau/8846527/bedgraph_to_wig.pl</a> using step size 100</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">wig</span></code> was converted to <code class="docutils literal notranslate"><span class="pre">bigWig</span></code> using UCSC toolkit using chrom.sizes for <cite>hg38</cite> downloaded from UCSC genome browser</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>bedtools genomecov -bga -ibam <span class="k">in</span>.bam &gt;out.bg

bedgraph_to_wig.pl --bedgraph out.bg --wig out.wig --step <span class="m">100</span>

wigToBigWig out.wig chrom.sizes final.bw
</pre></div>
</div>
<p>All files necessary to execute the code in <code class="docutils literal notranslate"><span class="pre">R</span></code> can be copied from Rackham from:</p>
<p><code class="docutils literal notranslate"><span class="pre">/sw/courses/epigenomics/quantitative_chip_simon/exospike.tar.gz</span></code></p>
<p>After copying the files please decompress the archive and note the path to folder <code class="docutils literal notranslate"><span class="pre">/chip_exo_spike</span></code> on your local system.</p>
<p>For local setup you can copy and extract files:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>scp &lt;user&gt;@rackham.uppmax.uu.se:/sw/courses/epigenomics/quantitative_chip_simon/exospike.tar.gz .

tar -xzf exospike.tar.gz
</pre></div>
</div>
</div>
<div class="section" id="fingerprint-plots">
<h2><a class="toc-backref" href="#id6">Fingerprint plots</a><a class="headerlink" href="#fingerprint-plots" title="Permalink to this headline">¶</a></h2>
<p>all reads mapped to <code class="docutils literal notranslate"><span class="pre">hg38</span></code> (i.e. not subset):</p>
<a class="reference external image-reference" href="Figures/exo-hg38-fingerprint.png"><img alt="" src="../../../_images/exo-hg38-fingerprint.png" /></a>
<p>all reads mapped to <code class="docutils literal notranslate"><span class="pre">dm6</span></code> (i.e. not subset):</p>
<a class="reference external image-reference" href="Figures/exo-dm6-fingerprint.png"><img alt="" src="../../../_images/exo-dm6-fingerprint.png" /></a>
</div>
<div class="section" id="disclaimer">
<h2><a class="toc-backref" href="#id7">Disclaimer</a><a class="headerlink" href="#disclaimer" title="Permalink to this headline">¶</a></h2>
<p>Please be aware that this is an experimental code, and as such does not represent any golden standard for analyses of this type. This is my exploration of the topic of using exogenous chromatic spike for ChIP-seq. I will aim to keep updating it with further steps of the analysis, once I get there.</p>
</div>
<div class="section" id="using-chipseqspike-for-chipseq-signal-scaling">
<h2><a class="toc-backref" href="#id8">Using <code class="docutils literal notranslate"><span class="pre">ChIPSeqSpike</span></code> for ChIPseq signal scaling</a><a class="headerlink" href="#using-chipseqspike-for-chipseq-signal-scaling" title="Permalink to this headline">¶</a></h2>
<p>This workflow is based on this <a class="reference external" href="https://github.com/descostesn/BiocNYC-ChIPSeqSpike">repository</a>.</p>
<p>The scaling procedure works on computers with <em>non-Windows</em> operating systems. This includes Uppmax, so you can use <code class="docutils literal notranslate"><span class="pre">salloc</span></code> command to book a node and follow the workflow remotely.</p>
<div class="section" id="files-and-directories">
<h3><a class="toc-backref" href="#id9">Files and directories</a><a class="headerlink" href="#files-and-directories" title="Permalink to this headline">¶</a></h3>
<p>In <code class="docutils literal notranslate"><span class="pre">R</span></code>:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">workdir</span><span class="o">=</span><span class="s">&quot;/path/to/chip_exo_spike&quot;</span>
<span class="nf">setwd</span><span class="p">(</span><span class="n">workdir</span><span class="p">)</span>

<span class="n">bam_path</span><span class="o">=</span><span class="nf">file.path</span><span class="p">(</span><span class="n">workdir</span><span class="p">,</span><span class="s">&quot;bam&quot;</span><span class="p">)</span>
<span class="n">bw_path</span><span class="o">=</span><span class="nf">file.path</span><span class="p">(</span><span class="n">workdir</span><span class="p">,</span><span class="s">&quot;tracks&quot;</span><span class="p">)</span>
<span class="n">exp_data</span><span class="o">=</span><span class="nf">file.path</span><span class="p">(</span><span class="n">workdir</span><span class="p">,</span><span class="s">&quot;exp_data.csv&quot;</span><span class="p">)</span>

<span class="c1">#you will have to copy the initial bigwig tracks to the output folder at a later stage</span>
<span class="c1">#output_folder=file.path(workdir,&quot;results&quot;)</span>
<span class="c1">#dir.create(output_folder)</span>

<span class="c1">#so until the package code is fixed:</span>
<span class="n">output_folder</span><span class="o">=</span><span class="n">bw_path</span>
</pre></div>
</div>
<p>You can inspect the file <code class="docutils literal notranslate"><span class="pre">exp_data.csv</span></code> to familiarize yourself with the structure:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">info</span><span class="o">=</span><span class="nf">read.table</span><span class="p">(</span><span class="n">exp_data</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;,&quot;</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">info</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="scaling-of-signal-to-exogenous-chromatin-spike">
<h3><a class="toc-backref" href="#id10">Scaling of signal to exogenous chromatin spike</a><a class="headerlink" href="#scaling-of-signal-to-exogenous-chromatin-spike" title="Permalink to this headline">¶</a></h3>
<p>Load the library and create the object:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">ChIPSeqSpike</span><span class="p">)</span>
<span class="n">cs</span> <span class="o">&lt;-</span> <span class="nf">spikeDataset</span><span class="p">(</span><span class="n">exp_data</span><span class="p">,</span> <span class="n">bam_path</span><span class="p">,</span> <span class="n">bw_path</span><span class="p">)</span>
</pre></div>
</div>
<p>Calculate the size factors based on numbers of mapped reads:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span> <span class="o">&lt;-</span> <span class="nf">estimateScalingFactors</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="nf">spikeSummary</span><span class="p">(</span><span class="n">cs</span><span class="p">)</span>
                <span class="n">endoScalFact</span> <span class="n">exoScalFact</span> <span class="n">endoCount</span> <span class="n">exoCount</span>
<span class="n">H3K79me2_0</span>         <span class="m">0.5367522</span>   <span class="m">1.0216143</span>   <span class="m">1863057</span>   <span class="m">978843</span>
<span class="n">input</span>              <span class="m">1.1604563</span>          <span class="kc">NA</span>    <span class="m">861730</span>       <span class="kc">NA</span>
<span class="n">H3K79me2_50</span>        <span class="m">0.6604427</span>   <span class="m">0.7663511</span>   <span class="m">1514136</span>  <span class="m">1304885</span>
<span class="n">input</span>              <span class="m">2.9039209</span>          <span class="kc">NA</span>    <span class="m">344362</span>       <span class="kc">NA</span>
<span class="n">H3K79me2_100_r1</span>    <span class="m">1.5994012</span>   <span class="m">0.3687641</span>    <span class="m">625234</span>  <span class="m">2711761</span>
<span class="n">input</span>              <span class="m">2.5008003</span>          <span class="kc">NA</span>    <span class="m">399872</span>       <span class="kc">NA</span>
<span class="n">H3K79me2_100_r2</span>    <span class="m">2.6171433</span>   <span class="m">0.6153835</span>    <span class="m">382096</span>  <span class="m">1625003</span>
<span class="n">input</span>              <span class="m">7.7456934</span>          <span class="kc">NA</span>    <span class="m">129104</span>       <span class="kc">NA</span>
</pre></div>
</div>
<p><strong>RPM scaling.</strong> The first normalization applied to the data is the ‘Reads Per Million’ (RPM) mapped reads. The method ‘scaling’ is used to achieve this normalization using default parameters.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span> <span class="o">&lt;-</span> <span class="nf">scaling</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">outputFolder</span> <span class="o">=</span> <span class="n">output_folder</span><span class="p">)</span>
</pre></div>
</div>
<p>You are supposed to obtain files <code class="docutils literal notranslate"><span class="pre">*-RPM.bw</span></code> after this step.</p>
<p><strong>Input subtraction.</strong> This step is to subtract background (from input samples) from signal. The <cite>inputSubtraction</cite> method simply subtracts scores of the input DNA experiment from the corresponding ones.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span> <span class="o">&lt;-</span> <span class="nf">inputSubtraction</span><span class="p">(</span><span class="n">cs</span><span class="p">)</span>
</pre></div>
</div>
<p>You are supposed to obtain <code class="docutils literal notranslate"><span class="pre">*-RPM-BGSub.bw</span></code> after this step.</p>
<p><strong>RPM scaling reversal.</strong> After RPM and input subtraction normalization, the RPM normalization is reversed in order for the data to be normalized by the exogenous scaling factors.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span><span class="o">&lt;-</span> <span class="nf">scaling</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">reverse</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">*-RPM-BGSub-reverted.bw</span></code> files after this step.</p>
<p><strong>Exogenous Scaling.</strong> Finally, exogenous scaling factors are applied to the data.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span> <span class="o">&lt;-</span> <span class="nf">scaling</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">type</span> <span class="o">=</span> <span class="s">&quot;exo&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The end result: <code class="docutils literal notranslate"><span class="pre">*-RPM-BGSub-reverted-spiked.bw</span></code> files after this step.</p>
<p><strong>Extracting binding values.</strong> The last step of data processing is to extract and format binding scores in order to use plotting methods. The <cite>extractBinding</cite> method extracts binding scores at different locations and stores these values in the form of PlotSetArray objects and matrices. The scores are retrieved on annotations provided in a <cite>gff</cite> file. If one wishes to focus on peaks, their coordinates should be submitted at this step. The genome name must also be provided. For details about installing the required <cite>BSgenome</cite> package corresponding to the endogenous organism, see the BSgenome package documentation.</p>
<p>Please note that this steps may take a long time.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">gff</span><span class="o">=</span><span class="nf">file.path</span><span class="p">(</span><span class="n">workdir</span><span class="p">,</span><span class="s">&quot;hg38_refseq_chr1.gtf&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">BSgenome.Hsapiens.UCSC.hg38</span><span class="p">)</span>

<span class="n">cs</span> <span class="o">&lt;-</span> <span class="nf">extractBinding</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">gff_vec</span><span class="o">=</span><span class="n">gff</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s">&quot;hg38&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>After this step, save the workspace</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">save.image</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="s">&quot;chipseqspike.RData&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>To load the data:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">load</span><span class="p">(</span><span class="s">&quot;chipseqspike.RData&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="data-visualization">
<h3><a class="toc-backref" href="#id11">Data visualization</a><a class="headerlink" href="#data-visualization" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">ChIPSeqSpike</span></code> offers several graphical methods for normalization diagnosis and data exploration. These choices enable one to visualize each step of the normalization through exploring intersamples differences using profiles, heatmaps, boxplots and correlation plots.</p>
<p>When performing this exercise on Uppmax, save the plots to pdf for viewing:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">pdf</span><span class="p">(</span><span class="s">&quot;filename.pdf&quot;</span><span class="p">)</span>
<span class="c1">## here command to produce the plot</span>
<span class="nf">dev.off</span><span class="p">()</span>
</pre></div>
</div>
<div class="section" id="visualization-with-gene-meta-profiles">
<h4><a class="toc-backref" href="#id12">Visualization with gene meta-profiles</a><a class="headerlink" href="#visualization-with-gene-meta-profiles" title="Permalink to this headline">¶</a></h4>
<p>The first step of spike-in normalized ChIP-Seq data analysis is an inter-sample comparison by meta-gene or meta-annotation profiling. The method <code class="docutils literal notranslate"><span class="pre">plotProfile</span></code> automatically plots all experiments at the start, midpoint, end and composite locations of the annotations provided to the method extractBinding in gff format. The effect of each transformation on a particular experiment can be visualized with <cite>plotTransform</cite>.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">## Plot spiked-in data</span>
<span class="nf">plotProfile</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">legends</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>

<span class="c1">## Add profiles before transformation</span>
<span class="nf">plotProfile</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">legends</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">notScaled</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>

<span class="c1">## Visualize the effect of each transformation on each experiment</span>
<span class="nf">plotTransform</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">legends</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">separateWindows</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="visualization-with-boxplots">
<h4><a class="toc-backref" href="#id13">Visualization with Boxplots</a><a class="headerlink" href="#visualization-with-boxplots" title="Permalink to this headline">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">boxplotSpike</span></code> plots boxplots of the mean values of ChIP-seq experiments on the annotations given to the <code class="docutils literal notranslate"><span class="pre">extractBinding</span></code> method.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">## Boxplot of the spiked-in data</span>
<span class="nf">boxplotSpike</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">outline</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>

<span class="c1">## Boxplot of the raw data</span>
<span class="nf">boxplotSpike</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span><span class="n">rawFile</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">spiked</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">outline</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>

<span class="c1">## Boxplot of all transformations</span>
<span class="nf">boxplotSpike</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span><span class="n">rawFile</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">rpmFile</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">bgsubFile</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">revFile</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">spiked</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">outline</span> <span class="o">=</span>     <span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="correlation-plots">
<h4><a class="toc-backref" href="#id14">Correlation plots</a><a class="headerlink" href="#correlation-plots" title="Permalink to this headline">¶</a></h4>
<p>The <code class="docutils literal notranslate"><span class="pre">plotCor</span></code> method plots the correlation between ChIP-seq experiments using heatscatter plot.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">## Log transform correlation plot of spiked data with heatscatter representation</span>
<span class="nf">plotCor</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">rawFile</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">rpmFile</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>  <span class="n">bgsubFile</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>  <span class="n">revFile</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">spiked</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>  <span class="n">main</span> <span class="o">=</span>     <span class="s">&quot;heatscatter&quot;</span><span class="p">,</span>  <span class="n">method_cor</span> <span class="o">=</span> <span class="s">&quot;spearman&quot;</span><span class="p">,</span> <span class="n">add_contour</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>  <span class="n">nlevels</span> <span class="o">=</span> <span class="m">10</span><span class="p">,</span>  <span class="n">color_contour</span> <span class="o">=</span> <span class="s">&quot;black&quot;</span><span class="p">,</span>     <span class="n">method_scale</span> <span class="o">=</span> <span class="s">&quot;log&quot;</span><span class="p">,</span>  <span class="n">allOnPanel</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">separateWindows</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>  <span class="n">verbose</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>

<span class="c1">## Plot as above with raw data</span>
<span class="nf">plotCor</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">rawFile</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">rpmFile</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>  <span class="n">bgsubFile</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>  <span class="n">revFile</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">spiked</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>  <span class="n">main</span> <span class="o">=</span>     <span class="s">&quot;heatscatter&quot;</span><span class="p">,</span>  <span class="n">method_cor</span> <span class="o">=</span> <span class="s">&quot;spearman&quot;</span><span class="p">,</span> <span class="n">add_contour</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>  <span class="n">nlevels</span> <span class="o">=</span> <span class="m">10</span><span class="p">,</span>  <span class="n">color_contour</span> <span class="o">=</span> <span class="s">&quot;black&quot;</span><span class="p">,</span>     <span class="n">method_scale</span> <span class="o">=</span> <span class="s">&quot;log&quot;</span><span class="p">,</span>  <span class="n">allOnPanel</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">separateWindows</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>  <span class="n">verbose</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>

<span class="c1">## Correlation table comparing all transformations</span>
<span class="n">corr_matrix</span> <span class="o">&lt;-</span> <span class="nf">plotCor</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">rawFile</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">rpmFile</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">bgsubFile</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">revFile</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">spiked</span> <span class="o">=</span>     <span class="kc">TRUE</span><span class="p">,</span> <span class="n">heatscatterplot</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="what-to-do-next">
<h2><a class="toc-backref" href="#id15">What to do next</a><a class="headerlink" href="#what-to-do-next" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>use scaled <code class="docutils literal notranslate"><span class="pre">bigWig</span></code> tracks to view the signal in IGV.</p></li>
<li><p>use <cite>bed</cite> files produced from scaled <code class="docutils literal notranslate"><span class="pre">bigWigs</span></code> to perform peak calling for instance with MACS2</p></li>
<li><p>differential binding analysis using <code class="docutils literal notranslate"><span class="pre">csaw</span></code> (more appropriate for broad marks) inputing the scaling factors obtained from scaling by <code class="docutils literal notranslate"><span class="pre">ChIPSeqSpike</span></code>.</p></li>
<li><p>perform the normalisation / scaling directly in <code class="docutils literal notranslate"><span class="pre">csaw</span></code>.</p></li>
<li><p>use scaled <code class="docutils literal notranslate"><span class="pre">bed</span></code> / <code class="docutils literal notranslate"><span class="pre">bigwig</span></code> for data exploration using PCA and MA plots.</p></li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="../chip_downstream_tutorials.html" class="btn btn-neutral float-right" title="Downstream Analyses" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="cut-and-tag-data.html" class="btn btn-neutral float-left" title="Analysis of CUT&amp;Tag H3K27me3 dataset" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, NBIS Epigenomics Workshop Team. All rights reserved..

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>