

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Setup &mdash; Epigenomics Workshop 2020 1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="In-situ methods: CUT&amp;RUN and ChIP data for CTCF" href="cut-and-run-data.html" />
    <link rel="prev" title="Advanced ChIP methods" href="../quantitative_chip_tutorials.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Epigenomics Workshop 2020
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../precourse.html">Pre-course Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../online-classroom.html">Online classroom</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../schedule.html">Schedule</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../setup/lab-setup.html">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="../methylation_tutorials.html">DNA Methylation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chipseq_tutorials.html">ChIPseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../atacseq_tutorials.html">ATACseq</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../quantitative_chip_tutorials.html">Advanced ChIP methods</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Setting up</a></li>
<li class="toctree-l3"><a class="reference internal" href="cut-and-run-data.html">CUT&amp;RUN and ChIP data for CTCF</a></li>
<li class="toctree-l3"><a class="reference internal" href="cut-and-tag-data.html">CUT&amp;Tag H3K27me3 data</a></li>
<li class="toctree-l3"><a class="reference internal" href="lab-exospike.html">Spike-in</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../chip_downstream_tutorials.html">Downstream Processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../expdesign_data.html">Experimental Design and Data Considerations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_integration_tutorials.html">Data Integration</a></li>
<li class="toctree-l2"><a class="reference internal" href="../nextflow.html">Nextflow &amp; nf-core</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Epigenomics Workshop 2020</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../tutorials.html">Tutorials</a> &raquo;</li>
        
          <li><a href="../quantitative_chip_tutorials.html">Advanced ChIP methods</a> &raquo;</li>
        
      <li>Setup</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../../_sources/content/tutorials/quantitativeChip/setup-quant-chip.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="setup">
<h1>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h1>
<p>In these tutorials we will be using the following software:</p>
<ul class="simple">
<li><p><strong>deepTools</strong>. Analysis suite for sequencing data. You can read about the installation <a class="reference external" href="https://deeptools.readthedocs.io/en/develop/content/installation.html">in their documentation</a>.</p></li>
<li><p><strong>IGV</strong>. <a class="reference external" href="http://software.broadinstitute.org/software/igv/">Integrative Genomics Viewer</a>.</p></li>
<li><p><strong>seqplots</strong>. R package for the analysis of bigWig files. It can be run as a shiny app and can be installed from Bioconductor. Check <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/seqplots.html">seqplots Bioconductor web</a> for more details.</p></li>
<li><p><strong>intervene</strong>. Easy to use R tool that allows to plot overlap between bed files. It can be installed through bioconda and pip (for me it worked using pip, but not conda though). Bedtools is a dependency. See more detailed info <a class="reference external" href="https://intervene.readthedocs.io/en/latest/install.html">in the intervene documentation site</a>.</p></li>
</ul>
<p>Additionally, other Bioconductor R packages will be needed, mostly as dependencies: <code class="code docutils literal notranslate"><span class="pre">GenomicRanges</span></code>, <code class="code docutils literal notranslate"><span class="pre">rtracklayer</span></code>.</p>
<p>Bioconductor packages can be installed in the usual way:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">requireNamespace</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
  <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>

<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;seqplots&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;GenomicRanges&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;rtracklayer&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can run things from Uppmax as well as on your local laptop. If you choose to run examples on Uppmax, the relevant module load calls are:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load deepTools
module load IGV
module load R_packages
</pre></div>
</div>
<p><code class="code docutils literal notranslate"><span class="pre">R_packages</span></code> contains all the mentioned R packages and their dependencies: <code class="code docutils literal notranslate"><span class="pre">GenomicRanges</span></code>, <code class="code docutils literal notranslate"><span class="pre">rtracklayer</span></code>, <code class="code docutils literal notranslate"><span class="pre">seqplots</span></code>. Be sure to have the corresponding modules loaded (and <code class="code docutils literal notranslate"><span class="pre">bioinfo-tools</span></code> before anything) before running them.</p>
<p>However, I recommend to download the bigWig and peak files and look at them locally. Most of the computationally demanding work has been pre-processed and it’s probably easier to look at the files on
IGV locally rather than remote. <code class="docutils literal notranslate"><span class="pre">seqplots</span></code> is also easier to run locally.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Computationally demanding steps have been precalculated and resulting plots are shown. Some of them can optionally be run again (such as <code class="docutils literal notranslate"><span class="pre">deepTools</span></code> computations). In those cases it will be noted within the corresponding section.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="cut-and-run-data.html" class="btn btn-neutral float-right" title="In-situ methods: CUT&amp;RUN and ChIP data for CTCF" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../quantitative_chip_tutorials.html" class="btn btn-neutral float-left" title="Advanced ChIP methods" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, NBIS Epigenomics Workshop Team. All rights reserved..

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>