

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Signal visualisation &mdash; Epigenomics Workshop 2020 1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="ATAC-seq" href="../atacseq_tutorials.html" />
    <link rel="prev" title="Alternative QC using ChIPQC" href="../chipseqQC/lab-chipqc.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Epigenomics Workshop 2020
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../precourse.html">Pre-course Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../online-classroom.html">Online classroom</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../schedule.html">Schedule</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../setup/lab-setup.html">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="../methylation_tutorials.html">DNA Methylation</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../chipseq_tutorials.html">ChIPseq</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../chipseqProc/lab-chipseq-processing.html">ChIPseq: processing and narrow peak detection</a></li>
<li class="toctree-l3"><a class="reference internal" href="../chipseqBroadPeaks/lab-broadpeaks.html">ChIPseq: broad peaks detection</a></li>
<li class="toctree-l3"><a class="reference internal" href="../chipseqQC/lab-chipqc.html">ChIPseq: alternative QC</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Signal visualisation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#signal-visualisation-with-deeptools">Signal visualisation with deepTools</a></li>
<li class="toctree-l4"><a class="reference internal" href="#appendix">Appendix</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../atacseq_tutorials.html">ATACseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quantitative_chip_tutorials.html">Advanced ChIP methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chip_downstream_tutorials.html">Downstream Processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../expdesign_data.html">Experimental Design and Data Considerations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_integration_tutorials.html">Data Integration</a></li>
<li class="toctree-l2"><a class="reference internal" href="../nextflow.html">Nextflow &amp; nf-core</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Epigenomics Workshop 2020</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../tutorials.html">Tutorials</a> &raquo;</li>
        
          <li><a href="../chipseq_tutorials.html">ChIP-seq</a> &raquo;</li>
        
      <li>Signal visualisation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../../_sources/content/tutorials/visDeepTools/lab-vis-deepTools.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="signal-visualisation">
<h1>Signal visualisation<a class="headerlink" href="#signal-visualisation" title="Permalink to this headline">¶</a></h1>
<p><strong>Learning outcomes</strong></p>
<p>Using <code class="docutils literal notranslate"><span class="pre">deepTools</span></code>
- to visualise ChIP signal in relation to annotated TSS</p>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#signal-visualisation-with-deeptools" id="id3">Signal visualisation with deepTools</a></p></li>
<li><p><a class="reference internal" href="#appendix" id="id4">Appendix</a></p></li>
</ul>
</div>
<div class="section" id="signal-visualisation-with-deeptools">
<h2><a class="toc-backref" href="#id3">Signal visualisation with deepTools</a><a class="headerlink" href="#signal-visualisation-with-deeptools" title="Permalink to this headline">¶</a></h2>
<p>One more thing that may come useful when analysing ChIP-seq data is visualising ChIP signal in relation to annotated features.</p>
<p>One such kind of features relevant for TFs are transcription start sites (TSS). In this exersice we use annotations for chromosomes 1 and 2. To do so we will:</p>
<ul class="simple">
<li><p>convert <code class="docutils literal notranslate"><span class="pre">bedgraph</span></code> to <code class="docutils literal notranslate"><span class="pre">bigWig</span></code> using <code class="docutils literal notranslate"><span class="pre">UCSC</span> <span class="pre">utilities</span></code></p></li>
<li><p>calculate scores per genome regions using among others the <code class="docutils literal notranslate"><span class="pre">bigWig</span></code> file</p></li>
<li><p>plot a heatmap of scores associated with genomic regions</p></li>
</ul>
<p>Assuming the same files structure as in the main data processing tutorial, create a separate directory in <code class="docutils literal notranslate"><span class="pre">~/chipseq/analysis</span></code> and navigate to it. Copy the files needed for this exercise.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> ~/chipseq/analysis/
mkdir vis
<span class="nb">cd</span> vis

cp ../../hg19/chrom.sizes.hg19 chrom.sizes.hg19
cp ../bam_preproc/ENCFF000PED.chr12.cov.norm1x.bedgraph ./
</pre></div>
</div>
<p>To calculate scores per genome with <code class="docutils literal notranslate"><span class="pre">deepTools</span></code> <a class="reference external" href="http://deeptools.readthedocs.org/en/latest/content/tools/computeMatrix.html">computeMatrix</a> we need <a class="reference external" href="https://genome.ucsc.edu/goldenpath/help/bigWig.html">bigWig</a> file that we can obtain by converting bedgraph using <code class="docutils literal notranslate"><span class="pre">UCSC</span> <span class="pre">utilities</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load ucsc-utilities/v398

bedGraphToBigWig ENCFF000PED.chr12.cov.norm1x.bedgraph chrom.sizes.hg19 hela_1.bw

module unload ucsc-utilities
</pre></div>
</div>
<p>We can now compute the matrix of scores for visualisation using <a class="reference external" href="http://deeptools.readthedocs.org/en/latest/content/tools/computeMatrix.html">computeMatrix</a>. This tool calculates scores per genome regions and prepares an intermediate file that can be used with <code class="docutils literal notranslate"><span class="pre">plotHeatmap</span></code> and <code class="docutils literal notranslate"><span class="pre">plotProfiles</span></code>. Typically, the genome regions are genes, but any other regions defined in a BED file can be used. <code class="docutils literal notranslate"><span class="pre">computeMatrix</span></code> accepts multiple score files (bigWig format) and multiple regions files (BED format). This tool can also be used to filter and sort regions according to their score.</p>
<p>We will need a <code class="docutils literal notranslate"><span class="pre">BED</span></code> file with positions of TSS that we can copy to the working directory before running <code class="docutils literal notranslate"><span class="pre">computeMatrix</span></code> e.g.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module load deepTools/3.3.2

cp ../../hg19/refGene_hg19_TSS_chr12_sorted_corr.bed ./

computeMatrix reference-point -S hela_rep1.bw <span class="se">\</span>
-R refGene_hg19_TSS_chr12_sorted_corr.bed -b <span class="m">5000</span> -a <span class="m">5000</span> <span class="se">\</span>
--outFileName matrix.tss.dat --outFileNameMatrix matrix.tss.txt <span class="se">\</span>
--referencePoint<span class="o">=</span>TSS -p <span class="m">5</span>
</pre></div>
</div>
<p>We can now create a heatmap for scores associated with genomic regions, i.e. plot the binding profile around TSS</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>plotHeatmap --matrixFile matrix.tss.dat <span class="se">\</span>
--outFileName tss.hela_1.pdf <span class="se">\</span>
--sortRegions descend --sortUsing mean
</pre></div>
</div>
<p>Have a look at the <code class="docutils literal notranslate"><span class="pre">tss.hela_rep1.pdf</span></code>. What do you think?</p>
<p>This is a very basic plot. We can add on to it, for example we can cluster genes based on the signal profile around TSS. For more possibilities please check <a class="reference external" href="https://deeptools.readthedocs.io/en/develop/content/tools/plotHeatmap.html">plotHetmap</a>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>plotHeatmap --matrixFile matrix.tss.dat <span class="se">\</span>
--outFileName tss.hela_rep1_k3_.pdf <span class="se">\</span>
--sortRegions descend --sortUsing mean <span class="se">\</span>
--kmeans <span class="m">3</span>
</pre></div>
</div>
<p><span class="raw-html"><br /></span></p>
</div>
<div class="section" id="appendix">
<h2><a class="toc-backref" href="#id4">Appendix</a><a class="headerlink" href="#appendix" title="Permalink to this headline">¶</a></h2>
<p>The plots generated in this tutorial.</p>
<table class="colwidths-given docutils align-default" id="id2">
<caption><span class="caption-text">Figure 1. Signal profiles and heatmaps centered on TSS for REST ChIP-seq in HeLa, replicate 1.</span><a class="headerlink" href="#id2" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>non-clustered</p></th>
<th class="head"><p>clustered (kmeans, k=3)</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><a class="reference internal image-reference" href="../../../_images/tss.hela_1.png"><img alt="../../../_images/tss.hela_1.png" src="../../../_images/tss.hela_1.png" style="width: 200px;" /></a>
</td>
<td><a class="reference internal image-reference" href="../../../_images/tss.hela_rep1_k3.png"><img alt="../../../_images/tss.hela_rep1_k3.png" src="../../../_images/tss.hela_rep1_k3.png" style="width: 210px;" /></a>
</td>
</tr>
</tbody>
</table>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="../atacseq_tutorials.html" class="btn btn-neutral float-right" title="ATAC-seq" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../chipseqQC/lab-chipqc.html" class="btn btn-neutral float-left" title="Alternative QC using ChIPQC" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, NBIS Epigenomics Workshop Team. All rights reserved..

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>