<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>Public data - Epigenomics 2020</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../../css/theme.css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "Public data";
    var mkdocs_page_input_path = "tutorials/lab-public-resources.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js" defer></script>
  <script src="../../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> Epigenomics 2020</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../welcome/">Welcome</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../about/">About</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../precourse/">Precourse Information</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../schedule/">Schedule</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../">Tutorials</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">Epigenomics 2020</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
    
    <li>Public data</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="public-chip-seq-resources">Public ChIP-seq resources</h1>
<p>Aim of this exercise</p>
<ul>
<li>
<p>To see what public datasets are out there</p>
</li>
<li>
<p>To be able to find and download public ChIP-seq data in useful formats</p>
</li>
</ul>
<p>This exercise is mostly run in a web browser, so it’s easiest to run it on you local computer.</p>
<h2 id="1-enocde">1. ENOCDE</h2>
<p><a href="https://www.encodeproject.org">https://www.encodeproject.org</a></p>
<p>Besides ENCODE data, the ENCODE data portal also contains data from the Roadmap Epigenome project, as well as the modENCODE &amp; modERN projects (for fly and worm).</p>
<p>To explore this data repostioty, go to the <a href="https://www.encodeproject.org">encode website</a> and select <em>Data</em> and then <em>Search</em>. Say that we want to see all data sets for the histone mark H3K27me3. Start by by typing "H3K27me3" in the search box in the top right corner. <strong>How many results do you see?</strong> This refers to everything in the encode data base: experiments, series of experiments, publications etc.</p>
<p>You can select subsets of the results from the panel on the left. Select <em>Experiment</em> to only see experiments. <strong>How many results do you see now? Are all these ChIP-seq experiments?</strong></p>
<p>Now, let’s make a finer selection: Select only released experiments (from <em>Experiment Status</em>), and then only experiments using the GRCh38 genome. <strong>How many experiments do we have now?</strong></p>
<p>Perhaps we are only interested in data from the brain. Under <em>Organ</em>, click on <em>See more</em>, and the select <em>brain</em>. <strong>How many experiments do we have now?</strong></p>
<p>You can see a list of all experiments to the right. Click on the first one, and open the page in a new browser window or tab. This will take you to a page describing this experiment, and what protocols and analysis pipelines were used. If you scroll down and select the tab <em>File details</em> you will se a list of all files that are available for download.</p>
<p><strong>Do you know what these files are?</strong> Try downloading some if you want to. But since some of these files are large, remember to remove them when you are done looking at them.</p>
<p>Now have seen how to find and download a single ChIP-seq data set. If we instead want to download all H3K27me3 data from the human brain mapped to the GRCh38 genome, go back to the ENCODE search page, where you had selected the relevant experiments. Then click the button <em>Download</em>. This will download a file to your computer.</p>
<p>Open this file. It contains URLs to all data files for the selected experiments. If we want to download e.g. only the bed files with peaks that are stable in both replicates of each experiment, we need to do some extra steps.</p>
<p>First, we download the meta data table. The URL is the first line in the file you just downloaded:</p>
<pre><code>wget &quot;https://www.encodeproject.org/metadata/searchTerm=H3K27me3&amp;type=Experiment&amp;status=released&amp;assembly=GRCh38&amp;organ_slims=brain/metadata.tsv&quot;
</code></pre>

<p>You can open this file, e.g. in excel to have a look. We want to select all lines corresponding to bed files with stable/replicated peaks, using the GRCh38 genome, and save these in a new file <code>metadata_peak_files.tsv</code>:</p>
<pre><code>grep bed.gz metadata.tsv | grep &quot;stable\|replicated&quot; | grep GRCh38 &gt; metadata_peak_files.tsv
</code></pre>

<p>From this file we can now get the URLs, which are in column 37:</p>
<pre><code>cut -d$'\t' -f37 metadata_peak_files.tsv &gt; metadata_peak_files_urls.txt
</code></pre>

<p>Finally, we can now download all these files:</p>
<pre><code>wget -i metadata_peak_files_urls.txt
</code></pre>

<p>This will download load all peak files. These will still have non-informative names, e.g. <code>ENCFF591RMN.bed.gz</code>. To see which experiment each file corresponds to, look in <code>metadata_peak_files.tsv</code></p>
<h3 id="roadmap-epigenomics">(Roadmap epigenomics)</h3>
<p>There are several ways to download data from the Roadmap Epigenomics web site. But since these data sets are also available through ENCODE, it's probably easiest to use the ENCODE web site. In case you want to have a look, these pages also host the Roadmap Epigenomics data:</p>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/roadmap/epigenomics/">https://www.ncbi.nlm.nih.gov/geo/roadmap/epigenomics/</a></p>
<p><a href="http://genboree.org/EdaccData/Release-9/">http://genboree.org/EdaccData/Release-9/</a></p>
<h2 id="2-cistrome">2. Cistrome</h2>
<p><a href="http://cistrome.org/db/#/">http://cistrome.org/db/#/</a></p>
<p>Cistrome is another database where ChIP-seq data has been collected and processed uniformly. This is a good complement to the big projects (ENCODE &amp; Roadmap Epigenomics) since it collects data from many smaller studies.</p>
<p>As an example, we will look for data on the three human transcription factors <em>Grhl1</em>, <em>Grhl2</em> and <em>Grhl3</em>. Grhl stands for “Grainy head like”, which means that these proteins are similar to the <em>Grainy head</em> protein first found in fruit fly. In human there are 3 Grhl homologs: <em>Grhl1</em>, <em>Grhl2</em> and <em>Grhl3</em>. They are involved in development and would healing, and have been implicated in hearing loss and cancer.</p>
<p>To see which ChIP-seq data sets are available for the Grhl proteins, go to <a href="http://cistrome.org/db/#/">http://cistrome.org/db/#/</a>, type "Grhl" in the search box and click on <em>Search</em>. We can then refine the search further by selecting <em>Homo sapiens</em> under <em>Species</em>. <strong>Which Grhl proteins do we find data for?</strong></p>
<p>There is also an option to filter data on quality measures. To try this, click on <em>Options</em> and then <em>Samples passing peak quality controls</em>. <strong>Which Grhl proteins do we still have data for after this filtering?</strong></p>
<p>Now, select the first data set in the list. It will be highlighted in blue. Scroll down to see details about this sample. Select the tab <em>QC reports</em>. <strong>Can you make sense of this information? Does this look like an experiment that worked?</strong></p>
<p>In the cistrome database, motif finding programs were run on all transcription factor data sets. Select the tab <em>QC motifs</em>, and have a look at some of the top motifs. <strong>Do they look similar the known Grhl site? Can you find what the Grhl site is?</strong> Hint: Go to JASPAR and search for "Grhl".</p>
<p>It’s also possible to download files from cistrome. For each experiment, bigwig files with the read coverage signal and bed files with peaks are available. Note that batch download is not available for manually selected batches, just for e.g. all human transcription factor data or all mouse chromatin data etc.</p>
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
