# Epigenomics Data Analysis: from Bulk to Single Cell

Online workshop open for PhD students, postdocs, researchers and other employees within Swedish academia. This course is organised by the National Bioinformatics Infrastructure Sweden (NBIS).

This workshop builds on our previous [ChIP-seq data analysis workshop](https://nbisweden.github.io/workshop-archive/workshop-ChIP-seq/2018-11-07/) to introduce the best practice bioinformatics methods for processing, analyses and integration of epigenomics data. The online teaching includes lectures, programming tutorials and interactive group sessions.

<br/>
## <img border="0" src="icons/calendar-date.png" width="40" height="40"> Date
23rd - 27th November 2020

<br/>
## <img border="0" src="icons/important-dates.png" width="40" height="40"> Important dates

- Application opens: August, 1st, 2020
- Application deadline: November, 1st, 2020
- Confirmation to accepted students: November, 7th, 2020
- Course days: 23rd - 27th, November, 2020


<br/>
## <img border="0" src="icons/content.png" width="40" height="40"> Workshop content
This workshop is designed to introduce the best practice bioinformatics methods for processing, analyses and integration of epigenomics and functional genomics data.

Topis covered include:

* **ChIP-seq** and **ATAC-seq**: peak calling, peak independent / dependent quality metrics, differential binding and differential accessibility analysis; DNA motif enrichment;

* Data processing and analyses for differential **DNA methylation** with Illumina EPIC arrays and Bisulfite-seq;
* **Functional analysis**, including finding nearest genes and custom features, GO terms and Reactome pathways enrichment;
* **Integrative visualisations** of epigenomics datasets;
* RNA-seq integration for identification of targets of transcription factors;
* Basic multi-omics exploration and integration;
* Introduction to analysis of single cell functional genomics data (ATAC-seq);
* **Nf-cores pipelines**: Methylseq, ChIP-seq, ATAC-seq

<br/>
## <img border="0" src="icons/entrance.png" width="40" height="40"> Entry requirements

Required to be able to follow the tutorials:

-  BYOL, bring your own laptop with R and RStudio installed
- Basic knowledge in Linux
- Basic R programming experience

Makes learning experience easier:

- Experience working on the SNIC center Uppmax or another HPC
- Previous experience with NGS data analyses
- Completing NBIS workshops “Introduction to Bioinformatics using NGS data” and “R Programming Foundations for Life Scientists” or equivalent


<br/>
## <img border="0" src="icons/application.png" width="40" height="40"> Application
- Application is now open [here](https://forms.gle/zBorhLmT3ZNBn7bm8).
- Application deadline: 1st November, 2020

<br/>
## <img border="0" src="icons/selection.png" width="40" height="40"> Selection criteria
 - Due to limited capacity the workshop can accommodate maximum of 30 participants. If we receive more applications, participants will be selected based on several criteria. Selection criteria include correct entry requirements, motivation to attend the course as well as gender and geographical balance.
 - NBIS prioritises academic participants, especially PhD students, affiliated with Swedish universities. We warmly welcome international and/or non-academic participants when we have seats available and the requirements criteria are met.

<br/>
## <img border="0" src="icons/fees.png" width="40" height="40"> Fees
- This workshop is free of charge; however a no-show fee of 2000 SEK may apply.

<br/>
## <img border="0" src="icons/diploma.png" width="40" height="40"> Certificate & University credits

We will issue a course certificate upon a successful course completion, assessed by active participation in all course sessions (lectures, computer practicals, group discussion etc.).

Unfortunately we are **not** able to warrant any university credits (högskolepoäng). Many universities however, recognize the attendance in our courses and award 1.5 HPs, corresponding to a 40 h of studying. It is up to participants to clarify and arrange credit transfer with the relevant university department.


<br/>
## <img border="0" src="icons/team.png" width="40" height="40"> Teaching team

- [Agata Smialowska][agata] <<agata.smialowska@nbis.se>>
- [Olga Dethlefsen][olga] <<olga.dethlefsen@nbis.se>>
- [Jakub Orzechowski Westholm][jakub] <<jakub.westholm@scilifelab.se>>
- [Vincent van Hoef][vincent] <<vincent.vanhoef@nbis.se>>
- [Phil Ewels][phil] <<phil.ewels@scilifelab.se>>
- [Harshil Patel][harshil] <<Harshil.Patel@crick.ac.uk>>

[agata]: https://nbis.se/about/staff/agata-smialowska/
[olga]: https://nbis.se/about/staff/olga-dethlefsen/
[jakub]: https://nbis.se/about/staff/jakub-orzechowski-westholm/
[vincent]: https://nbis.se/about/staff/vincent-van-hoef
[phil]: phil.ewels@scilifelab.se
[harshil]: https://www.researchgate.net/profile/Harshil_Patel12

<br/>
## Contact us
- edu.epigenomics@nbis.se




<!--

## Travel Info

Science for Life Laboratory (SciLifeLab) is located in Stockholm suburb Solna (Tomtebodavägen 23A, 171 65 Solna).

We are at number 23a on [Karolinska Solna Campus Map](https://nbisweden.github.io/workshop-archive/workshop-ChIP-seq/2018-11-07/files/karta_campus_solna_16_11_14.pdf).

The closest bus stop is called ***Karolinska Institutet Biomedicum*** (search for public transport options [here](https://sl.se/en/)).

Enter the SciLifeLab / Karolinska Institutet Science Park building. After entering, turn left and pass through the glass door to find the rooms *Air* and *Fire*, where the workshop takes place. You can ask for help in the reception if you cannot find the way.


 -->
