# Precourse Material

There are few things to do ***before*** the course starts. Please read carefully and follow the instructions so we can have a good start to the course. Contact us in case anything in unclear.


## Computational Resources

During the course we will be using Uppsala University’s high performance computer cluster (Uppmax) as well as run scripts locally on laptops using R and RStudio.

To be able to follow exercises we ask you to

* configure access to [Uppmax](https://uppmax.uu.se/);

* install [R](https://cran.r-project.org/) and [RStudio](https://rstudio.com/) on your laptop;

* install [Integrative Genomics Viewer](https://software.broadinstitute.org/software/igv/) on your laptop.


### Uppmax

Computational resources are provided by SNIC / [Uppmax](https://uppmax.uu.se/). To be able to use them

* create a user account (only if you do not already have one);

* associate your user account with the course project.


#### Create user account & request membership in the course project (for those new to Uppmax)

This involves few steps. Briefly,

* registering at [SUPR](https://supr.snic.se/);

* accepting the User Agreement;

* becoming a member in the project ( ***XXXX***);

* applying for an account at Uppmax.


To go through the steps follow the instructions on Uppmax website. While at this keep this information handy:

* Cluster name: rackham

* Project ID:  ***XXXX***


#### Request membership in the course project (for those already having Uppmax account)


* log in to SUPR;

* under Projects: Requesting Membership in Projects, request membership in ***XXXX***



#### Check configuration (everyone)

After you complete setting-up

* login to `rackham.uppmax.uu.se`

* type `id` in the command line
    
* copy the output of the command and email back (to the course organisers at edu.epigenomics@nbis.se) 

[A guide on how to log-in for the first time]((http://www.uppmax.uu.se/support/user-guides/guide–first-login-to-uppmax/)


### R and RStudio

We will also be using the latest version of `R` and `RStudio` locally. Both of these work on computers running Linux, Windows and Macintosh operating systems. `RStudio` is a set of tools as well as an editor that facilitates the use of `R` (R ICE). Over the last years it has become a very popular tool and in many ways become a *de-facto* standard for working with `R.`

Note that on same operative systems it will be easier to install and run `R` and `RStudio` if you are administrator of your own computer and hence are allowed to install software on your machine. If you do not have these privileges please ask your system administrator to install the latest version of `R` and `RStudio`.


### IGV

To install, follow the instructions on the [IGV website](https://software.broadinstitute.org/software/igv/).


## Further optional preparations

For those of you wanting to start ahead and/or brush up on various skills before the course.


### Computer skills

* [Unix](http://www.ee.surrey.ac.uk/Teaching/Unix/): especially the first three chapters.
 
* DataCamp free [R tutorial](https://learn.datacamp.com/courses/free-introduction-to-r). A nice self learn tutorial to `R`, introducing many central concepts to `R`.

* [A short introduction to R](https://cran.r-project.org/doc/contrib/Torfs+Brauer-Short-R-Intro.pdf). A very short introduction to using `R`.

### ChIP-seq

* [Video introduction to ChIP-seq data analysis](https://www.youtube.com/watch?v=zwuUveGgmS0) by Dr. Carl Hermann, University of Heidelberg.
    
* [ChIP-seq and beyond: new and improved methodologies to detect and characterize protein-DNA interactions](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3591838/).
    
* [Q&A: ChIP-seq technologies and the study of gene regulation](https://bmcbiol.biomedcentral.com/articles/10.1186/1741-7007-8-56).



