# Lab Tutorials


## Setting up

* [Setup](tutorials/lab-setup.md)



## Day 1: DNA methylation


## Day 2: Workflow for ChIP-seq data processing

* [ChIP-seq data processing and QC](tutorials/lab-chipseq-processing.md) (Agata)

* [Visualisation](tutorials/lab-vis.md) (?)

* [Alternative way to QC in R](tutorials/lab-chipqc.md) (Agata)


## Day 3: Downstream analyses


* [Main differential binding and functional annotations workflow](tutorials/lab-diffBinding-local.md) (Olga) [local version]

* [Main differential binding and functional annotations workflow](tutorials/lab-diffBinding-remote.md) (Olga) [Uppmax version]

* [ChIPseeker package for ChIP profiling, peaks annotations, visualisation and functional annotations](tutorials/lab-ChIPseeker.md) (Olga) [local]

* [Motifs finding](tutorials/lab-motifs.md) (Jakub) [Uppmax]
    
* [Calling broad peaks](tutorials/lab-broadpeaks.md) (Agata) [local]

* [Alternative differential binding analyses using csaw](tutorials/) (Agata) [local]



## Day 4: ATAC-seq, special cases and curiosities

Introduction to practicals IV:

* ATAC-seq

* scATAC-seq

* [Scaling normalisation for ChIP-seq with exogenous chromatin spike](tutorials/lab-exospike.md) (Agata) [local]
   
* [Finding data in public resources](tutorials/lab-public-resources.md) (Jakub) [local]





## Day 5: Omics integration















<!--
How to link:

Let's link a lecture [here](lectures/Meyer_Liu_bias_nihms680120.pdf)

Let's link a tutorial [here](tutorials/day2/lab-chipseq-proc/chipseq-proc.md)
 -->

