

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>ChIP-seq down-stream analysis: ChIPseeker &mdash; Epigenomics Workshop 2020 1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="ChIP-seq down-stream analysis" href="../diffBind/lab-diffBinding-local.html" />
    <link rel="prev" title="Downstream Analyses" href="../chip_downstream_tutorials.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Epigenomics Workshop 2020
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../precourse.html">Pre-course Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../online-classroom.html">Online classroom</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../schedule.html">Schedule</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../setup/lab-setup.html">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="../methylation_tutorials.html">DNA Methylation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../chipseq_tutorials.html">ChIPseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../atacseq_tutorials.html">ATACseq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quantitative_chip_tutorials.html">Advanced ChIP methods</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../chip_downstream_tutorials.html">Downstream Processing</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Annotation using ChIPseeker</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l4"><a class="reference internal" href="#data-methods">Data &amp; Methods</a></li>
<li class="toctree-l4"><a class="reference internal" href="#setting-up">Setting-up</a></li>
<li class="toctree-l4"><a class="reference internal" href="#chip-profile">ChIP profile</a></li>
<li class="toctree-l4"><a class="reference internal" href="#peaks-annotation">Peaks Annotation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#functional-analysis">Functional analysis</a></li>
<li class="toctree-l4"><a class="reference internal" href="#concluding-remarks-and-next-steps">Concluding remarks and next steps</a></li>
<li class="toctree-l4"><a class="reference internal" href="#appendix">Appendix</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../diffBind/lab-diffBinding-local.html">Detection of differential binding sites using DiffBind local and annotation with ChIPpeakAnno</a></li>
<li class="toctree-l3"><a class="reference internal" href="../diffBind/lab-diffBinding-remote.html">Detection of differential binding sites using DiffBind remote</a></li>
<li class="toctree-l3"><a class="reference internal" href="../motifs/lab-motifs.html">Identification of sequence motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../csaw/lab-csaw.html">Detection of differential binding sites using csaw</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../expdesign_data.html">Experimental Design and Data Considerations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_integration_tutorials.html">Data Integration</a></li>
<li class="toctree-l2"><a class="reference internal" href="../nextflow.html">Nextflow &amp; nf-core</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Epigenomics Workshop 2020</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../tutorials.html">Tutorials</a> &raquo;</li>
        
          <li><a href="../chip_downstream_tutorials.html">Downstream Analyses</a> &raquo;</li>
        
      <li>ChIP-seq down-stream analysis: ChIPseeker</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../../_sources/content/tutorials/chipseeker/lab-ChIPseeker.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="chip-seq-down-stream-analysis-chipseeker">
<h1>ChIP-seq down-stream analysis: ChIPseeker<a class="headerlink" href="#chip-seq-down-stream-analysis-chipseeker" title="Permalink to this headline">¶</a></h1>
<p><strong>Learning outcomes</strong></p>
<p>Using <code class="docutils literal notranslate"><span class="pre">ChIPseeker</span></code> package
- to profile ChIP signal by genomics location and by ChIP binding to TSS regions
- to annotate peaks, visualise and compare annotations
- to run and compare functional enrichment</p>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#introduction" id="id3">Introduction</a></p></li>
<li><p><a class="reference internal" href="#data-methods" id="id4">Data &amp; Methods</a></p></li>
<li><p><a class="reference internal" href="#setting-up" id="id5">Setting-up</a></p></li>
<li><p><a class="reference internal" href="#chip-profile" id="id6">ChIP profile</a></p>
<ul>
<li><p><a class="reference internal" href="#chip-peaks-coverage-plot" id="id7">ChIP peaks coverage plot</a></p></li>
<li><p><a class="reference internal" href="#profile-of-chip-peaks-binding-to-tss-regions" id="id8">Profile of ChIP peaks binding to TSS regions</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#peaks-annotation" id="id9">Peaks Annotation</a></p></li>
<li><p><a class="reference internal" href="#functional-analysis" id="id10">Functional analysis</a></p></li>
<li><p><a class="reference internal" href="#concluding-remarks-and-next-steps" id="id11">Concluding remarks and next steps</a></p></li>
<li><p><a class="reference internal" href="#appendix" id="id12">Appendix</a></p></li>
</ul>
</div>
<div class="section" id="introduction">
<h2><a class="toc-backref" href="#id3">Introduction</a><a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>In this tutorial we use another package, <code class="docutils literal notranslate"><span class="pre">ChIPseeker</span></code>, to have a look at the ChIP profiles, annotate peaks and visualise annotations as well as to run functional enrichment. In a way, <code class="docutils literal notranslate"><span class="pre">ChIPseeker</span></code> can be seen as an alternative and newer workflow to <code class="docutils literal notranslate"><span class="pre">ChIPpeakAnno</span></code> (introduced in #####). It also offers additional functionality, e.g. especially when it comes to visualising ChIP profiles and comparing functional annotations.</p>
<p><em>It supports annotating ChIP peaks and provides functions to visualize ChIP peaks coverage over chromosomes and profiles of peaks binding to TSS regions. Comparison of ChIP peak profiles and annotation are also supported. Moreover, it supports evaluating significant overlap among ChIP-seq datasets. Currently, ChIPseeker contains 17,000 bed file information from GEO database. These datasets can be downloaded and compare with user’s own data to explore significant overlap datasets for inferring co-regulation or transcription factor complex for further investigation.</em></p>
</div>
<div class="section" id="data-methods">
<h2><a class="toc-backref" href="#id4">Data &amp; Methods</a><a class="headerlink" href="#data-methods" title="Permalink to this headline">¶</a></h2>
<p>We will build upon the main labs, using the same dataset and results from <code class="docutils literal notranslate"><span class="pre">DiffBind</span></code> analyses that we have saved under <code class="docutils literal notranslate"><span class="pre">DiffBind.RData</span></code>. The tutorial is based on the <a class="reference external" href="https://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html">ChIPseeker package tutorial</a> so feel free to have this open alongside to read and experiment more.</p>
<p><span class="raw-html"><br /></span></p>
</div>
<div class="section" id="setting-up">
<h2><a class="toc-backref" href="#id5">Setting-up</a><a class="headerlink" href="#setting-up" title="Permalink to this headline">¶</a></h2>
<p>If you have not done it already, install <code class="docutils literal notranslate"><span class="pre">R</span></code> and <code class="docutils literal notranslate"><span class="pre">R-Studio</span></code>. Refer back to ####[pre-course](../precourse)### preparations for instructions.</p>
<p>You can continue working in the <code class="docutils literal notranslate"><span class="pre">diffBind</span></code> directory. We need access to <code class="docutils literal notranslate"><span class="pre">diffBind.RData</span></code> object &amp; some libraries</p>
<p>In an <a href="#id1"><span class="problematic" id="id2">``</span></a>R``session:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load libraries (install if needed)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">DiffBind</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ChIPseeker</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ReactomePA</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">clusterProfiler</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">biomaRt</span><span class="p">)</span>

<span class="nf">library</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="p">)</span>
<span class="n">txdb</span> <span class="o">&lt;-</span> <span class="n">TxDb.Hsapiens.UCSC.hg19.knownGene</span>
</pre></div>
</div>
<p><span class="raw-html"><br /></span></p>
</div>
<div class="section" id="chip-profile">
<h2><a class="toc-backref" href="#id6">ChIP profile</a><a class="headerlink" href="#chip-profile" title="Permalink to this headline">¶</a></h2>
<div class="section" id="chip-peaks-coverage-plot">
<h3><a class="toc-backref" href="#id7">ChIP peaks coverage plot</a><a class="headerlink" href="#chip-peaks-coverage-plot" title="Permalink to this headline">¶</a></h3>
<p>After peak calling one may want to visualise distribution of peaks locations over the whole genome. Function <code class="docutils literal notranslate"><span class="pre">covplot</span></code> calculates coverage of peaks regions over chromosomes.</p>
<p>Let’s use data saved in <code class="docutils literal notranslate"><span class="pre">DiffBind.RData</span></code> objects. From this object we can easily extract peaks called for all our libraries as well as consensus peakset. In principle, we could also use <code class="docutils literal notranslate"><span class="pre">ChIPseeker</span></code> on raw <code class="docutils literal notranslate"><span class="pre">.BED</span></code> files.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Let&#39;s start fresh removing all objects from R environment</span>
<span class="nf">rm</span><span class="p">(</span><span class="n">list</span> <span class="o">=</span> <span class="nf">ls</span><span class="p">())</span>

<span class="c1"># loading diffBind.RData</span>
<span class="nf">load</span><span class="p">(</span><span class="s">&quot;diffBind.RData&quot;</span><span class="p">)</span>

<span class="c1"># Do you remember what objects we have saved in the diffBind.RData</span>
<span class="nf">ls</span><span class="p">()</span>

<span class="c1"># res.cnt3 object was the final one containing consensus peaks and differential binding results</span>

<span class="c1"># viewing all samples</span>
<span class="nf">dba.show</span><span class="p">(</span><span class="n">res.cnt3</span><span class="p">)</span>

<span class="c1"># this should show you our 8 libraries</span>
<span class="o">&gt;</span> <span class="nf">dba.show</span><span class="p">(</span><span class="n">res.cnt3</span><span class="p">)</span>
          <span class="n">ID</span> <span class="n">Tissue</span> <span class="n">Factor</span> <span class="n">Replicate</span> <span class="n">Caller</span> <span class="n">Intervals</span> <span class="n">FRiP</span>
<span class="m">1</span> <span class="n">REST_chip1</span>   <span class="n">HeLa</span>   <span class="n">REST</span>         <span class="m">1</span> <span class="n">counts</span>      <span class="m">6518</span> <span class="m">0.11</span>
<span class="m">2</span> <span class="n">REST_chip2</span>   <span class="n">HeLa</span>   <span class="n">REST</span>         <span class="m">2</span> <span class="n">counts</span>      <span class="m">6518</span> <span class="m">0.08</span>
<span class="m">3</span> <span class="n">REST_chip3</span> <span class="n">neural</span>   <span class="n">REST</span>         <span class="m">1</span> <span class="n">counts</span>      <span class="m">6518</span> <span class="m">0.07</span>
<span class="m">4</span> <span class="n">REST_chip4</span> <span class="n">neural</span>   <span class="n">REST</span>         <span class="m">2</span> <span class="n">counts</span>      <span class="m">6518</span> <span class="m">0.09</span>
<span class="m">5</span> <span class="n">REST_chip5</span>  <span class="n">HepG2</span>   <span class="n">REST</span>         <span class="m">1</span> <span class="n">counts</span>      <span class="m">6518</span> <span class="m">0.08</span>
<span class="m">6</span> <span class="n">REST_chip6</span>  <span class="n">HepG2</span>   <span class="n">REST</span>         <span class="m">2</span> <span class="n">counts</span>      <span class="m">6518</span> <span class="m">0.06</span>
<span class="m">7</span> <span class="n">REST_chip7</span>  <span class="n">sknsh</span>   <span class="n">REST</span>         <span class="m">1</span> <span class="n">counts</span>      <span class="m">6518</span> <span class="m">0.10</span>
<span class="m">8</span> <span class="n">REST_chip8</span>  <span class="n">sknsh</span>   <span class="n">REST</span>         <span class="m">2</span> <span class="n">counts</span>      <span class="m">6518</span> <span class="m">0.06</span>
</pre></div>
</div>
<p>To plot peaks over genomic locations we need to extract from <code class="docutils literal notranslate"><span class="pre">res.cnt3</span></code> peaks of interest, e.g. consensus peaks or present in a single replicate etc. Here, we will focus on peaks present in HeLa replicates.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># extracting consensus peak set with 6518 peaks</span>
<span class="n">peaks.consensus</span> <span class="o">&lt;-</span> <span class="nf">dba.peakset</span><span class="p">(</span><span class="n">res.cnt3</span><span class="p">,</span> <span class="n">bRetrieve</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>

<span class="c1"># extracting HeLA peaks</span>
<span class="n">peaks.HeLa_rep1</span> <span class="o">&lt;-</span> <span class="n">peaks.consensus</span><span class="p">[</span><span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span><span class="o">==</span><span class="m">1</span><span class="p">]</span> <span class="c1"># peaks called in rep 1</span>
<span class="n">peaks.HeLa_rep2</span> <span class="o">&lt;-</span> <span class="n">peaks.consensus</span><span class="p">[</span><span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">2</span><span class="p">]</span><span class="o">==</span><span class="m">1</span><span class="p">]</span> <span class="c1"># peaks called in rep 2</span>

<span class="c1"># adding an unified affinity scores column (re-formatting data)</span>
<span class="n">peaks.HeLa_rep1</span><span class="o">$</span><span class="n">Score</span> <span class="o">&lt;-</span> <span class="n">peaks.HeLa_rep1</span><span class="o">$</span><span class="n">REST_chip1</span>
<span class="n">peaks.HeLa_rep2</span><span class="o">$</span><span class="n">Score</span> <span class="o">&lt;-</span> <span class="n">peaks.HeLa_rep2</span><span class="o">$</span><span class="n">REST_chip2</span>

<span class="c1"># plotting coverage for replicate 1, using affinity scores as a weight for peaks height</span>
<span class="nf">covplot</span><span class="p">(</span><span class="n">peaks.HeLa_rep1</span><span class="p">,</span> <span class="n">weightCol</span> <span class="o">=</span> <span class="s">&quot;Score&quot;</span><span class="p">)</span>

<span class="c1"># zooming in to a selected region is also possible</span>
<span class="nf">covplot</span><span class="p">(</span><span class="n">peaks.HeLa_rep1</span><span class="p">,</span> <span class="n">weightCol</span> <span class="o">=</span> <span class="s">&quot;Score&quot;</span><span class="p">,</span> <span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">1e07</span><span class="p">))</span>
</pre></div>
</div>
<p>We can also compare peaks across replicates. This should give us visual assessment of variability between replicates: peaks locations and strength should match in an idealistic scenario</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># creating genomicRangesList object holding replicates 1 and 2</span>
<span class="n">grL.HeLa</span> <span class="o">&lt;-</span> <span class="nf">GenomicRangesList</span><span class="p">(</span><span class="n">HeLa_rep1</span><span class="o">=</span><span class="n">peaks.HeLa_rep1</span><span class="p">,</span> <span class="n">HeLa_rep2</span><span class="o">=</span><span class="n">peaks.HeLa_rep2</span><span class="p">)</span>

<span class="c1"># plotting using affinity scores as a weight for peaks height</span>
<span class="nf">covplot</span><span class="p">(</span><span class="n">grL.HeLa</span><span class="p">,</span> <span class="n">weightCol</span> <span class="o">=</span> <span class="s">&quot;Score&quot;</span><span class="p">)</span>

<span class="c1"># zooming in</span>
<span class="nf">covplot</span><span class="p">(</span><span class="n">grL.HeLa</span><span class="p">,</span> <span class="n">weightCol</span> <span class="o">=</span> <span class="s">&quot;Score&quot;</span><span class="p">,</span> <span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">1e07</span><span class="p">))</span>
</pre></div>
</div>
<p>What do you think?
- are these peaks reproducible?
- which pair of replicates is most consistent, HeLA, neural, HepG2 or sknsh?
- why is it good to always look at the data instead of simply trusting the output of the summary statistics, after all, we do rely on <code class="docutils literal notranslate"><span class="pre">diffBind</span></code> to call peaks being consistent?</p>
</div>
<div class="section" id="profile-of-chip-peaks-binding-to-tss-regions">
<h3><a class="toc-backref" href="#id8">Profile of ChIP peaks binding to TSS regions</a><a class="headerlink" href="#profile-of-chip-peaks-binding-to-tss-regions" title="Permalink to this headline">¶</a></h3>
<p>For calculating the profile of ChIP peaks binding to TSS regions, we need to prepare the TSS regions, which are defined as the flanking sequence of the TSS sites. Then we can align the peaks that are mapping to these regions, and generate the tagMatrix used for plotting.</p>
<p>Here, we will select peaks present per cell type, i.e. found in two replicates. We will also create tagMatrix list to enable groups comparisons across cell lines.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># extracting peaks for each cell line present across replicates</span>
<span class="n">peaks.HeLa</span> <span class="o">&lt;-</span> <span class="n">peaks.consensus</span><span class="p">[</span><span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span><span class="o">==</span><span class="m">1</span> <span class="o">&amp;</span> <span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">2</span><span class="p">]</span><span class="o">==</span><span class="m">1</span><span class="p">]</span>
<span class="n">peaks.neural</span> <span class="o">&lt;-</span> <span class="n">peaks.consensus</span><span class="p">[</span><span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">3</span><span class="p">]</span><span class="o">==</span><span class="m">1</span> <span class="o">&amp;</span> <span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">4</span><span class="p">]</span><span class="o">==</span><span class="m">1</span><span class="p">]</span>
<span class="n">peaks.HepG2</span> <span class="o">&lt;-</span> <span class="n">peaks.consensus</span><span class="p">[</span><span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">5</span><span class="p">]</span><span class="o">==</span><span class="m">1</span> <span class="o">&amp;</span> <span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">6</span><span class="p">]</span><span class="o">==</span><span class="m">1</span><span class="p">]</span>
<span class="n">peaks.sknsh</span> <span class="o">&lt;-</span> <span class="n">peaks.consensus</span><span class="p">[</span><span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">7</span><span class="p">]</span><span class="o">==</span><span class="m">1</span> <span class="o">&amp;</span> <span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">8</span><span class="p">]</span><span class="o">==</span><span class="m">1</span><span class="p">]</span>

<span class="c1"># getting TSS regions</span>
<span class="n">promoter</span> <span class="o">&lt;-</span> <span class="nf">getPromoters</span><span class="p">(</span><span class="n">TxDb</span><span class="o">=</span><span class="n">txdb</span><span class="p">,</span> <span class="n">upstream</span><span class="o">=</span><span class="m">3000</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="m">3000</span><span class="p">)</span>

<span class="c1"># calculating tagMatrix</span>
<span class="n">tagMatrix.1</span> <span class="o">&lt;-</span> <span class="nf">getTagMatrix</span><span class="p">(</span><span class="n">peaks.HeLa</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="n">promoter</span><span class="p">)</span>
<span class="n">tagMatrix.2</span> <span class="o">&lt;-</span> <span class="nf">getTagMatrix</span><span class="p">(</span><span class="n">peaks.neural</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="n">promoter</span><span class="p">)</span>
<span class="n">tagMatrix.3</span> <span class="o">&lt;-</span> <span class="nf">getTagMatrix</span><span class="p">(</span><span class="n">peaks.HepG2</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="n">promoter</span><span class="p">)</span>
<span class="n">tagMatrix.4</span> <span class="o">&lt;-</span> <span class="nf">getTagMatrix</span><span class="p">(</span><span class="n">peaks.sknsh</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="n">promoter</span><span class="p">)</span>

<span class="c1"># preparing tagMatrix list to enable cell lines comparisions</span>
<span class="n">tagMatrixList</span> <span class="o">&lt;-</span> <span class="nf">list</span><span class="p">(</span><span class="n">HeLa</span><span class="o">=</span><span class="n">tagMatrix.1</span><span class="p">,</span> <span class="n">neural</span><span class="o">=</span><span class="n">tagMatrix.2</span><span class="p">,</span> <span class="n">HepG2</span><span class="o">=</span><span class="n">tagMatrix.3</span><span class="p">,</span> <span class="n">sknsh</span><span class="o">=</span><span class="n">tagMatrix.4</span><span class="p">)</span>

<span class="c1"># plotting tagMatrix heatmaps for each cell line</span>
<span class="nf">tagHeatmap</span><span class="p">(</span><span class="n">tagMatrixList</span><span class="p">,</span> <span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-3000</span><span class="p">,</span> <span class="m">3000</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span>

<span class="c1"># plotting average profile of ChIP peaks among different cell lines</span>
<span class="nf">plotAvgProf</span><span class="p">(</span><span class="n">tagMatrixList</span><span class="p">,</span> <span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-3000</span><span class="p">,</span> <span class="m">3000</span><span class="p">))</span>
</pre></div>
</div>
<p><span class="raw-html"><br /></span></p>
</div>
</div>
<div class="section" id="peaks-annotation">
<h2><a class="toc-backref" href="#id9">Peaks Annotation</a><a class="headerlink" href="#peaks-annotation" title="Permalink to this headline">¶</a></h2>
<p><strong>Peak annotations is performed by ``annotatePeak`` function</strong>. Here, we can define TSS region, by default set to -3kb to 3kb. The output of annotatePeak is <code class="docutils literal notranslate"><span class="pre">csAnno</span></code> object than we can convert to <code class="docutils literal notranslate"><span class="pre">GRanges</span></code> with <code class="docutils literal notranslate"><span class="pre">as.GRanges()</span></code> function or to data frame with <code class="docutils literal notranslate"><span class="pre">as.data.frame()</span></code> function.</p>
<p>Similar to annotations with <code class="docutils literal notranslate"><span class="pre">ChIPpeakAnno</span></code> we will need <code class="docutils literal notranslate"><span class="pre">TxDB</span></code> object containing annotations, transcript-related features of a particular genome. We can use Bioconductor packages providing annotations for various model organisms. It may be however <strong>good to know that one can also prepare their own TxDb object</strong> by retrieving information from UCSC or BioMart using <code class="docutils literal notranslate"><span class="pre">GenomicFeature</span></code> package. Here, we will use <code class="docutils literal notranslate"><span class="pre">TxDb.Hsapiens.UCSC.hg19.knownGene</span></code> annotations provided by Bioconductor.</p>
<p>Some <strong>annotations may overlap</strong> and by default ChIPseeker annotates peaks with the priority: promoter, 5’ UTR, 3’ UTR, exon, intron, downstreamn, intergenic, where downstream is defined as the downstream of gene end. This priority can be changed with <code class="docutils literal notranslate"><span class="pre">genomicAnnotationPriority</span></code> parameter.</p>
<p>While annotating peaks we can include optional parameter <code class="docutils literal notranslate"><span class="pre">annoDb</span></code> containig further genome wide annotation data. If added, this will add SYMBOL, GENENAME, ENSEMBL/ENTREZID to the peaks annotations. Again, we will use Bioconductor <code class="docutils literal notranslate"><span class="pre">org.Hs.eg.db</span></code> for human genome wide annotation data.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># extracting all consensus peaks (repeating commands for clarity)</span>
<span class="n">peaks.consensus</span> <span class="o">&lt;-</span> <span class="nf">dba.peakset</span><span class="p">(</span><span class="n">res.cnt3</span><span class="p">,</span> <span class="n">bRetrieve</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>

<span class="c1"># extracting peaks for each cell line present across replicates (repeating commands for clarity)</span>
<span class="n">peaks.HeLa</span> <span class="o">&lt;-</span> <span class="n">peaks.consensus</span><span class="p">[</span><span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span><span class="o">==</span><span class="m">1</span> <span class="o">&amp;</span> <span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">2</span><span class="p">]</span><span class="o">==</span><span class="m">1</span><span class="p">]</span>
<span class="n">peaks.neural</span> <span class="o">&lt;-</span> <span class="n">peaks.consensus</span><span class="p">[</span><span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">3</span><span class="p">]</span><span class="o">==</span><span class="m">1</span> <span class="o">&amp;</span> <span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">4</span><span class="p">]</span><span class="o">==</span><span class="m">1</span><span class="p">]</span>
<span class="n">peaks.HepG2</span> <span class="o">&lt;-</span> <span class="n">peaks.consensus</span><span class="p">[</span><span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">5</span><span class="p">]</span><span class="o">==</span><span class="m">1</span> <span class="o">&amp;</span> <span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">6</span><span class="p">]</span><span class="o">==</span><span class="m">1</span><span class="p">]</span>
<span class="n">peaks.sknsh</span> <span class="o">&lt;-</span> <span class="n">peaks.consensus</span><span class="p">[</span><span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">7</span><span class="p">]</span><span class="o">==</span><span class="m">1</span> <span class="o">&amp;</span> <span class="n">res.cnt3</span><span class="o">$</span><span class="n">called</span><span class="p">[,</span><span class="m">8</span><span class="p">]</span><span class="o">==</span><span class="m">1</span><span class="p">]</span>

<span class="c1"># annotating peaks</span>
<span class="n">peaks.HeLa_ann</span> <span class="o">&lt;-</span> <span class="nf">annotatePeak</span><span class="p">(</span><span class="n">peaks.HeLa</span><span class="p">,</span> <span class="n">tssRegion</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-3000</span><span class="p">,</span> <span class="m">3000</span><span class="p">),</span> <span class="n">TxDb</span><span class="o">=</span><span class="n">txdb</span><span class="p">,</span> <span class="n">annoDb</span><span class="o">=</span><span class="s">&quot;org.Hs.eg.db&quot;</span><span class="p">)</span>
<span class="n">peaks.neural_ann</span> <span class="o">&lt;-</span> <span class="nf">annotatePeak</span><span class="p">(</span><span class="n">peaks.neural</span><span class="p">,</span> <span class="n">tssRegion</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-3000</span><span class="p">,</span> <span class="m">3000</span><span class="p">),</span> <span class="n">TxDb</span><span class="o">=</span><span class="n">txdb</span><span class="p">,</span> <span class="n">annoDb</span><span class="o">=</span><span class="s">&quot;org.Hs.eg.db&quot;</span><span class="p">)</span>
<span class="n">peaks.HepG2_ann</span> <span class="o">&lt;-</span> <span class="nf">annotatePeak</span><span class="p">(</span><span class="n">peaks.HepG2</span><span class="p">,</span> <span class="n">tssRegion</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-3000</span><span class="p">,</span> <span class="m">3000</span><span class="p">),</span> <span class="n">TxDb</span><span class="o">=</span><span class="n">txdb</span><span class="p">,</span> <span class="n">annoDb</span><span class="o">=</span><span class="s">&quot;org.Hs.eg.db&quot;</span><span class="p">)</span>
<span class="n">peaks.sknsh_ann</span> <span class="o">&lt;-</span> <span class="nf">annotatePeak</span><span class="p">(</span><span class="n">peaks.sknsh</span><span class="p">,</span> <span class="n">tssRegion</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-3000</span><span class="p">,</span> <span class="m">3000</span><span class="p">),</span> <span class="n">TxDb</span><span class="o">=</span><span class="n">txdb</span><span class="p">,</span> <span class="n">annoDb</span><span class="o">=</span><span class="s">&quot;org.Hs.eg.db&quot;</span><span class="p">)</span>

<span class="c1"># previewing annotations summary for HeLa peaks</span>
<span class="n">peaks.HeLa_ann</span>

<span class="o">&gt;</span> <span class="n">peaks.HeLa_ann</span>
<span class="n">Annotated</span> <span class="n">peaks</span> <span class="n">generated</span> <span class="n">by</span> <span class="n">ChIPseeker</span>
<span class="m">1096</span><span class="o">/</span><span class="m">1096</span>  <span class="n">peaks</span> <span class="n">were</span> <span class="n">annotated</span>
<span class="n">Genomic</span> <span class="n">Annotation</span> <span class="n">Summary</span><span class="o">:</span>
              <span class="n">Feature</span>  <span class="n">Frequency</span>
<span class="m">9</span>    <span class="nf">Promoter </span><span class="p">(</span><span class="o">&lt;=</span><span class="m">1</span><span class="n">kb</span><span class="p">)</span> <span class="m">18.2481752</span>
<span class="m">10</span>   <span class="nf">Promoter </span><span class="p">(</span><span class="m">1-2</span><span class="n">kb</span><span class="p">)</span>  <span class="m">5.0182482</span>
<span class="m">11</span>   <span class="nf">Promoter </span><span class="p">(</span><span class="m">2-3</span><span class="n">kb</span><span class="p">)</span>  <span class="m">3.2846715</span>
<span class="m">4</span>              <span class="m">5</span><span class="s">&#39; UTR  0.1824818</span>
<span class="s">3              3&#39;</span> <span class="n">UTR</span>  <span class="m">1.7335766</span>
<span class="m">1</span>            <span class="m">1</span><span class="n">st</span> <span class="n">Exon</span>  <span class="m">0.1824818</span>
<span class="m">7</span>          <span class="n">Other</span> <span class="n">Exon</span>  <span class="m">2.6459854</span>
<span class="m">2</span>          <span class="m">1</span><span class="n">st</span> <span class="n">Intron</span> <span class="m">10.3102190</span>
<span class="m">8</span>        <span class="n">Other</span> <span class="n">Intron</span> <span class="m">20.6204380</span>
<span class="m">6</span>  <span class="nf">Downstream </span><span class="p">(</span><span class="o">&lt;=</span><span class="m">300</span><span class="p">)</span>  <span class="m">1.0036496</span>
<span class="m">5</span>   <span class="n">Distal</span> <span class="n">Intergenic</span> <span class="m">36.7700730</span>

<span class="c1"># previewing peaks annotations for HeLa peaks</span>
<span class="nf">head</span><span class="p">(</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="n">peaks.HeLa_ann</span><span class="p">))</span>
</pre></div>
</div>
<p>We find our genomic annotations in _annotation_ column. Plots, pie and barplot, are supported to visualise these annotations.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># creating barplot for HeLa peaks genomics annotations</span>
<span class="nf">plotAnnoBar</span><span class="p">(</span><span class="n">peaks.HeLa_ann</span><span class="p">)</span>

<span class="c1"># creating vennpie plot</span>
<span class="nf">vennpie</span><span class="p">(</span><span class="n">peaks.HeLa_ann</span><span class="p">)</span>

<span class="c1"># creating upsetplot showing overlapping annotations</span>
<span class="nf">upsetplot</span><span class="p">(</span><span class="n">peaks.HeLa_ann</span><span class="p">)</span>
</pre></div>
</div>
<p>We can also use <code class="docutils literal notranslate"><span class="pre">plotAnnoBar</span></code> to compare annotations between different datasets, here cell lines. For that, we just need to create a list containing peaks annotations of datasets to compare.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># creating list holding annotations for different cell lines</span>
<span class="n">list.annotations</span> <span class="o">&lt;-</span> <span class="nf">list</span><span class="p">(</span><span class="n">HeLa</span><span class="o">=</span><span class="n">peaks.HeLa_ann</span><span class="p">,</span> <span class="n">neural</span><span class="o">=</span><span class="n">peaks.neural_ann</span><span class="p">,</span> <span class="n">HepG2</span><span class="o">=</span><span class="n">peaks.HepG2_ann</span><span class="p">,</span> <span class="n">sknskh</span><span class="o">=</span><span class="n">peaks.sknsh_ann</span><span class="p">)</span>

<span class="c1"># creating barplot for HeLa, neural, HepG2 and sknsh peaks genomic annotations</span>
<span class="nf">plotAnnoBar</span><span class="p">(</span><span class="n">list.annotations</span><span class="p">)</span>
</pre></div>
</div>
<p>Finally, we can also visualise distribution of TF-binding loci relative to TSS, for single annotation set or using annotations list for comparisons.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># plotting distance to TSS for HeLa peaks</span>
<span class="nf">plotDistToTSS</span><span class="p">(</span><span class="n">peaks.HeLa_ann</span><span class="p">)</span>

<span class="c1"># plotting distance to TSS for all cell lines in our annotation list</span>
<span class="nf">plotDistToTSS</span><span class="p">(</span><span class="n">list.annotations</span><span class="p">)</span>
</pre></div>
</div>
<p>What do you think?
- would you expect such distribution of features?
- do these distributions differ between cell-lines?</p>
<p><span class="raw-html"><br /></span></p>
</div>
<div class="section" id="functional-analysis">
<h2><a class="toc-backref" href="#id10">Functional analysis</a><a class="headerlink" href="#functional-analysis" title="Permalink to this headline">¶</a></h2>
<p>Having obtained annotations to nearest genes, we can perform <strong>functional enrichment analysis to identify predominant biological themes</strong> among these genes by incorporating biological knowledge provided by biological ontologies, incl. GO (Gene Ontology, Ashburner et al. 2000), KEGG (Kyoto Encyclopedia of Genes and Genomes, Kanehisa et al. 2004), DO (Disease Ontology, Schriml et al. 2011) or Reactome (Croft et al. 2013).</p>
<p>Here, we can also use <code class="docutils literal notranslate"><span class="pre">seq2gene</span></code> function for linking genomic regions to genes in a <strong>many-to-many mapping</strong>. This function consider host gene (exon/intron), promoter region and flanking gene from intergenic region that may undergo control via cis-regulation.</p>
<p>One can <strong>build on</strong> using ChIPseeker for functional enrichment and annotation as there are several packages by the same author to identify biological themes, i.e. <code class="docutils literal notranslate"><span class="pre">ReactomePA</span></code> for reactome pathways enrichment, <code class="docutils literal notranslate"><span class="pre">DOSE</span></code> for Disease Ontology, <code class="docutils literal notranslate"><span class="pre">clusterProfiler</span></code> for Gene Ontology and KEGG enrichment analysis. Especially <a class="reference external" href="http://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html">clustserProfiler</a>
comes handy when <strong>visualising and comparing</strong> biological themes, also when comparing functions derived from other omics technologies for integrative analyses.</p>
<p>Here, we will experiment with few functions only. We will search for enriched reactome pathways using genes annotated to peaks by nearest location and allowing for many-to-many mapping. We will also learn how to compare functional annotations between peak sets using GO terms as an example.</p>
<p>We will start by defying our genes background, i.e. genes on chromosome 1 and 2. For this we can use functions from <code class="docutils literal notranslate"><span class="pre">BioMart</span></code></p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># defining chromosomes</span>
<span class="n">chrom</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">)</span>

<span class="c1"># defining source</span>
<span class="n">ensembl</span><span class="o">=</span><span class="nf">useMart</span><span class="p">(</span><span class="s">&quot;ensembl&quot;</span><span class="p">)</span>
<span class="n">ensembl</span> <span class="o">=</span> <span class="nf">useDataset</span><span class="p">(</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span><span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">)</span>

<span class="c1"># running query: extracting ENTREZID for genes on chromosome 1 and 2</span>
<span class="n">genes.chr1chr2</span> <span class="o">&lt;-</span> <span class="nf">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span> <span class="s">&quot;entrezgene&quot;</span><span class="p">,</span>
        <span class="n">filters</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;chromosome_name&quot;</span><span class="p">),</span>
        <span class="n">values</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span> <span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">)</span>

<span class="c1"># reformatting output to character string (as required later on by clusterProfiler functions)</span>
<span class="n">genes.universe</span> <span class="o">&lt;-</span> <span class="nf">as.character</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.matrix</span><span class="p">(</span><span class="n">genes.chr1chr2</span><span class="p">)))</span>
</pre></div>
</div>
<p>Reactome pathway enrichment of genes defined as a) nearest feature to the peaks and b) allowing for many-to-many mapping</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># a: selecting annotated peaks for functional enrichment in object</span>
<span class="n">data.peaks_ann</span> <span class="o">&lt;-</span> <span class="n">peaks.neural_ann</span>

<span class="c1"># a: fining enriched Reactome pathways using chromosome 1 and 2 genes as a background</span>
<span class="n">pathway.reac1</span> <span class="o">&lt;-</span> <span class="nf">enrichPathway</span><span class="p">(</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="n">data.peaks_ann</span><span class="p">)</span><span class="o">$</span><span class="n">geneId</span><span class="p">,</span> <span class="n">universe</span> <span class="o">=</span> <span class="n">genes.universe</span><span class="p">)</span>

<span class="c1"># a: previewing enriched Reactome pathways</span>
<span class="nf">head</span><span class="p">(</span><span class="n">pathway.reac1</span><span class="p">)</span>


<span class="c1"># b: selecting peaks</span>
<span class="n">data.peaks</span> <span class="o">&lt;-</span> <span class="n">peaks.neural</span>

<span class="c1"># b: running seq2gene function for many-to-many mapping based on sequence regions (note: no prior peaks annotations here, many-to-many mapping is done from the sequence)</span>
<span class="n">genes.m2m</span> <span class="o">&lt;-</span> <span class="nf">seq2gene</span><span class="p">(</span><span class="n">data.peaks</span><span class="p">,</span> <span class="n">tssRegion</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">-3000</span><span class="p">,</span> <span class="m">3000</span><span class="p">),</span> <span class="n">flankDistance</span> <span class="o">=</span> <span class="m">3000</span><span class="p">,</span> <span class="n">TxDb</span><span class="o">=</span><span class="n">txdb</span><span class="p">)</span>

<span class="c1"># b: finding enriched Reactome pathways given many to many mapping and chromosome 1 and 2 genes as a background</span>
<span class="n">pathway.reac2</span> <span class="o">&lt;-</span> <span class="nf">enrichPathway</span><span class="p">(</span><span class="n">genes.m2m</span><span class="p">,</span> <span class="n">universe</span> <span class="o">=</span> <span class="n">genes.universe</span><span class="p">)</span>

<span class="c1"># b: creating dotplot to visualise enrichment results</span>
<span class="nf">dotplot</span><span class="p">(</span><span class="n">pathway.reac2</span><span class="p">)</span>
</pre></div>
</div>
<p>Let’s search for enriched GO terms, and let’s see how we can do it for all the peak sets together so we can easily compare the results on a <code class="docutils literal notranslate"><span class="pre">dotplot</span></code>. Also, let’s learn how to simplify the output of GO terms using <code class="docutils literal notranslate"><span class="pre">simplify</span></code> function, useful in cases where lots of GO terms turn-up to be significant and it becomes difficult to interpret results. <code class="docutils literal notranslate"><span class="pre">simply</span></code> function removes redundant GO terms obtained from <code class="docutils literal notranslate"><span class="pre">encrichGO</span></code> calling internally <code class="docutils literal notranslate"><span class="pre">GoSemSim</span></code> function to calculate similarities among GO terms and removes those highly similar terms by keeping one representative term.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># creating a gene list with ENTREZID ideas extracted from our annotation list, containing annotated peaks for all four cell lines</span>
<span class="n">list.genes</span> <span class="o">=</span> <span class="nf">lapply</span><span class="p">(</span><span class="n">list.annotations</span><span class="p">,</span> <span class="nf">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">$</span><span class="n">geneId</span><span class="p">)</span>
<span class="nf">names</span><span class="p">(</span><span class="n">list.genes</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">sub</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">,</span> <span class="s">&quot;\n&quot;</span><span class="p">,</span> <span class="nf">names</span><span class="p">(</span><span class="n">list.genes</span><span class="p">))</span>

<span class="c1"># running enrichedGO function to find enriched MF correlation_libraries_normalised on the gene list</span>

<span class="n">compMF</span> <span class="o">&lt;-</span> <span class="nf">compareCluster</span><span class="p">(</span><span class="n">geneCluster</span> <span class="o">=</span> <span class="n">list.genes</span><span class="p">,</span>
                       <span class="n">fun</span>           <span class="o">=</span> <span class="s">&quot;enrichGO&quot;</span><span class="p">,</span>
                       <span class="n">pvalueCutoff</span>  <span class="o">=</span> <span class="m">0.05</span><span class="p">,</span>
                       <span class="n">pAdjustMethod</span> <span class="o">=</span> <span class="s">&quot;BH&quot;</span><span class="p">,</span>
                       <span class="n">OrgDb</span><span class="o">=</span><span class="s">&#39;org.Hs.eg.db&#39;</span><span class="p">,</span>
                       <span class="n">ont</span><span class="o">=</span><span class="s">&quot;MF&quot;</span><span class="p">)</span>

<span class="c1"># comparing results on a dotplot</span>
<span class="nf">dotplot</span><span class="p">(</span><span class="n">compMF</span><span class="p">)</span>

<span class="c1"># simplifying results although here we do not have problems with too many GO terms</span>
<span class="n">compMF.flr</span> <span class="o">&lt;-</span> <span class="nf">simplify</span><span class="p">(</span><span class="n">compMF</span><span class="p">,</span> <span class="n">cutoff</span> <span class="o">=</span> <span class="m">0.7</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s">&quot;p.adjust&quot;</span><span class="p">,</span> <span class="n">select_fun</span> <span class="o">=</span> <span class="n">min</span><span class="p">,</span> <span class="n">measure</span> <span class="o">=</span> <span class="s">&quot;Wang&quot;</span><span class="p">,</span> <span class="n">semData</span> <span class="o">=</span> <span class="kc">NULL</span><span class="p">)</span>

<span class="c1"># creating a dotplot on reduced GO terms</span>
<span class="nf">dotplot</span><span class="p">(</span><span class="n">compMF.flr</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="concluding-remarks-and-next-steps">
<h2><a class="toc-backref" href="#id11">Concluding remarks and next steps</a><a class="headerlink" href="#concluding-remarks-and-next-steps" title="Permalink to this headline">¶</a></h2>
<p>There are different flavours to functional annotations, and what and how functional annotations should be done is context dependent, i.e. they should be adjusted given available data and biological question being asked. There are many methods out there, all relying on the available annotations and databases, being constantly improved and developed. As a rule of thumb to understand the results and be able to draw biological conclusions, it may be good to think about i) the statistical test behind the method, ii) what is compared against what (i.e. genes vs. background) and which databases are being used (i.e. Reactome, GO, DO, KEGG).</p>
<p>For more examples on what can be done in terms on functional annotations, we recommend reading tutorials on <a class="reference external" href="http://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html#reduce-redundancy-of-enriched-go-terms">clusterProfiler</a> and <a class="reference external" href="https://bioconductor.org/packages/release/bioc/vignettes/DOSE/inst/doc/DOSE.html">DOSE</a>, where you can further learn about semantic similarity analysis, disease enrichment analysis, GSEA analysis and much more.</p>
<p><span class="raw-html"><br /></span></p>
</div>
<div class="section" id="appendix">
<h2><a class="toc-backref" href="#id12">Appendix</a><a class="headerlink" href="#appendix" title="Permalink to this headline">¶</a></h2>
<p>Links to figures generated during this session are shown below.</p>
<p><strong>ChIP profiling</strong></p>
<a class="reference internal image-reference" href="../../../_images/fig-covplot-1.png"><img alt="../../../_images/fig-covplot-1.png" src="../../../_images/fig-covplot-1.png" style="width: 600px;" /></a>
<p>Fig: Coverage plot for HeLa replicate 1 peaks</p>
<hr class="docutils" />
<a class="reference internal image-reference" href="../../../_images/fig-covplot-cmp-1.png"><img alt="../../../_images/fig-covplot-cmp-1.png" src="../../../_images/fig-covplot-cmp-1.png" style="width: 600px;" /></a>
<p>Fig: Coverage plot for HeLa replicate 1 and 2 peaks</p>
<hr class="docutils" />
<a class="reference internal image-reference" href="../../../_images/fig-covplot-cmp-2.png"><img alt="../../../_images/fig-covplot-cmp-2.png" src="../../../_images/fig-covplot-cmp-2.png" style="width: 600px;" /></a>
<p>Fig: Coverage plot for HeLa replicate 1 and 2 peaks (selected region)</p>
<hr class="docutils" />
<a class="reference internal image-reference" href="../../../_images/fig-TSS-heatmap-1.png"><img alt="../../../_images/fig-TSS-heatmap-1.png" src="../../../_images/fig-TSS-heatmap-1.png" style="width: 600px;" /></a>
<p>Fig: Heatmap of ChIP peaks in different cell lines</p>
<hr class="docutils" />
<a class="reference internal image-reference" href="../../../_images/fig-TSS-avg-profile-1.png"><img alt="../../../_images/fig-TSS-avg-profile-1.png" src="../../../_images/fig-TSS-avg-profile-1.png" style="width: 600px;" /></a>
<p>Fig: Average profile of ChIP peaks in different cell lines</p>
<hr class="docutils" />
<p><strong>Peak annotation</strong></p>
<a class="reference internal image-reference" href="../../../_images/fig-anno-vennpie-1.png"><img alt="../../../_images/fig-anno-vennpie-1.png" src="../../../_images/fig-anno-vennpie-1.png" style="width: 600px;" /></a>
<p>Fig: Genomic annotations (vennpie)</p>
<hr class="docutils" />
<a class="reference internal image-reference" href="../../../_images/fig-anno-barplot-1.png"><img alt="../../../_images/fig-anno-barplot-1.png" src="../../../_images/fig-anno-barplot-1.png" style="width: 600px;" /></a>
<p>Fig: Genomic annotations (barplot)</p>
<hr class="docutils" />
<a class="reference internal image-reference" href="../../../_images/fig-anno-upsetplot-1.png"><img alt="../../../_images/fig-anno-upsetplot-1.png" src="../../../_images/fig-anno-upsetplot-1.png" style="width: 600px;" /></a>
<p>Fig: Annotations overlap with UpSetPlot</p>
<hr class="docutils" />
<a class="reference internal image-reference" href="../../../_images/fig-anno-barplot-cmp-1.png"><img alt="../../../_images/fig-anno-barplot-cmp-1.png" src="../../../_images/fig-anno-barplot-cmp-1.png" style="width: 600px;" /></a>
<p>Fig: Genomics locations: dataset comparisons with barplot</p>
<hr class="docutils" />
<a class="reference internal image-reference" href="../../../_images/fig-anno-dist-to-tss-cmp-1.png"><img alt="../../../_images/fig-anno-dist-to-tss-cmp-1.png" src="../../../_images/fig-anno-dist-to-tss-cmp-1.png" style="width: 600px;" /></a>
<p>Fig: Distance to TSS: datasets comparisons</p>
<hr class="docutils" />
<p><strong>Functional annotation</strong></p>
<a class="reference internal image-reference" href="../../../_images/fig-func-reactome-1.png"><img alt="../../../_images/fig-func-reactome-1.png" src="../../../_images/fig-func-reactome-1.png" style="width: 600px;" /></a>
<p>Fig: Dotplot for enriched reactome pathways for neuronal peaks (many-to-many mapping)</p>
<hr class="docutils" />
<a class="reference internal image-reference" href="../../../_images/fig-func-go-1.png"><img alt="../../../_images/fig-func-go-1.png" src="../../../_images/fig-func-go-1.png" style="width: 600px;" /></a>
<p>Fig: Comparison of enriched GO MF terms (mapping to the nearest gene)</p>
<hr class="docutils" />
<a class="reference internal image-reference" href="../../../_images/fig-func-go-simp-1.png"><img alt="../../../_images/fig-func-go-simp-1.png" src="../../../_images/fig-func-go-simp-1.png" style="width: 600px;" /></a>
<p>Fig: Comparison of enriched GO MF terms (mapping to the nearest gene) after reducing redundant terms</p>
<hr class="docutils" />
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="../diffBind/lab-diffBinding-local.html" class="btn btn-neutral float-right" title="ChIP-seq down-stream analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../chip_downstream_tutorials.html" class="btn btn-neutral float-left" title="Downstream Analyses" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, NBIS Epigenomics Workshop Team. All rights reserved..

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>